// openclaw/dist/api-BASCRZud.js
import { d as readStringValue } from "./string-coerce-BUSzWgUA.js";
import { r as normalizeProviderId } from "./provider-id-Dts8rixb.js";
import { n as resolveProviderEndpoint } from "./provider-attribution-DhRYxXov.js";
import "./provider-tools-DnkYXOur.js";
import { n as getModelProviderHint } from "./provider-model-shared-pqO3tDFY.js";
import "./text-runtime-3VPC1sGh.js";
import "./model-definitions-Cx-cbPjg.js";
import "./provider-catalog-DZAL7B0-.js";
import "./onboard-Shos_xdp.js";
import "./provider-models-kos12Imz.js";
//#region extensions/xai/api.ts
function isXaiNativeEndpoint(baseUrl) {
	return typeof baseUrl === "string" && resolveProviderEndpoint(baseUrl).endpointClass === "xai-native";
}
function isXaiModelHint(modelId) {
	return getModelProviderHint(modelId) === "x-ai";
}
function shouldUseXaiResponsesTransport(params) {
	if (params.api !== "openai-completions") return false;
	if (isXaiNativeEndpoint(params.baseUrl)) return true;
	return normalizeProviderId(params.provider) === "xai" && !params.baseUrl;
}
function shouldContributeXaiCompat(params) {
	if (params.model.api !== "openai-completions") return false;
	return isXaiNativeEndpoint(params.model.baseUrl) || isXaiModelHint(params.modelId);
}
function resolveXaiTransport(params) {
	if (!shouldUseXaiResponsesTransport(params)) return;
	return {
		api: "openai-responses",
		baseUrl: readStringValue(params.baseUrl)
	};
}
//#endregion
export { resolveXaiTransport as n, shouldContributeXaiCompat as r, isXaiModelHint as t };
