// openclaw/dist/bootstrap-registry-nPyF75ul.js
import { s as normalizeOptionalString } from "./string-coerce-BUSzWgUA.js";
import { i as listChannelCatalogEntries } from "./ids-ZUx9SnUD.js";
import { i as getBundledChannelSetupSecrets, n as getBundledChannelSecrets, r as getBundledChannelSetupPlugin, t as getBundledChannelPlugin } from "./bundled-CdLjswoh.js";
//#region src/channels/plugins/bundled-ids.ts
let bundledChannelPluginIds = null;
function listBundledChannelPluginIds() {
	bundledChannelPluginIds ??= listChannelCatalogEntries({ origin: "bundled" }).map((entry) => entry.pluginId).toSorted((left, right) => left.localeCompare(right));
	return [...bundledChannelPluginIds];
}
//#endregion
//#region src/channels/plugins/bootstrap-registry.ts
let cachedBootstrapPlugins = null;
function mergePluginSection(runtimeValue, setupValue) {
	if (runtimeValue && setupValue && typeof runtimeValue === "object" && typeof setupValue === "object") {
		const merged = { ...runtimeValue };
		for (const [key, value] of Object.entries(setupValue)) if (value !== void 0) merged[key] = value;
		return { ...merged };
	}
	return setupValue ?? runtimeValue;
}
function mergeBootstrapPlugin(runtimePlugin, setupPlugin) {
	return {
		...runtimePlugin,
		...setupPlugin,
		meta: mergePluginSection(runtimePlugin.meta, setupPlugin.meta),
		capabilities: mergePluginSection(runtimePlugin.capabilities, setupPlugin.capabilities),
		commands: mergePluginSection(runtimePlugin.commands, setupPlugin.commands),
		doctor: mergePluginSection(runtimePlugin.doctor, setupPlugin.doctor),
		reload: mergePluginSection(runtimePlugin.reload, setupPlugin.reload),
		config: mergePluginSection(runtimePlugin.config, setupPlugin.config),
		setup: mergePluginSection(runtimePlugin.setup, setupPlugin.setup),
		messaging: mergePluginSection(runtimePlugin.messaging, setupPlugin.messaging),
		actions: mergePluginSection(runtimePlugin.actions, setupPlugin.actions),
		secrets: mergePluginSection(runtimePlugin.secrets, setupPlugin.secrets)
	};
}
function buildBootstrapPlugins() {
	return {
		sortedIds: listBundledChannelPluginIds(),
		byId: /* @__PURE__ */ new Map(),
		secretsById: /* @__PURE__ */ new Map(),
		missingIds: /* @__PURE__ */ new Set()
	};
}
function getBootstrapPlugins() {
	cachedBootstrapPlugins ??= buildBootstrapPlugins();
	return cachedBootstrapPlugins;
}
function listBootstrapChannelPluginIds() {
	return getBootstrapPlugins().sortedIds;
}
function* iterateBootstrapChannelPlugins() {
	for (const id of listBootstrapChannelPluginIds()) {
		const plugin = getBootstrapChannelPlugin(id);
		if (plugin) yield plugin;
	}
}
function getBootstrapChannelPlugin(id) {
	const resolvedId = normalizeOptionalString(id) ?? "";
	if (!resolvedId) return;
	const registry = getBootstrapPlugins();
	const cached = registry.byId.get(resolvedId);
	if (cached) return cached;
	if (registry.missingIds.has(resolvedId)) return;
	const runtimePlugin = getBundledChannelPlugin(resolvedId);
	const setupPlugin = getBundledChannelSetupPlugin(resolvedId);
	const merged = runtimePlugin && setupPlugin ? mergeBootstrapPlugin(runtimePlugin, setupPlugin) : setupPlugin ?? runtimePlugin;
	if (!merged) {
		registry.missingIds.add(resolvedId);
		return;
	}
	registry.byId.set(resolvedId, merged);
	return merged;
}
function getBootstrapChannelSecrets(id) {
	const resolvedId = normalizeOptionalString(id) ?? "";
	if (!resolvedId) return;
	const registry = getBootstrapPlugins();
	const cached = registry.secretsById.get(resolvedId);
	if (cached) return cached;
	if (registry.secretsById.has(resolvedId)) return;
	const merged = mergePluginSection(getBundledChannelSecrets(resolvedId), getBundledChannelSetupSecrets(resolvedId));
	registry.secretsById.set(resolvedId, merged ?? null);
	return merged;
}
//#endregion
export { listBundledChannelPluginIds as i, getBootstrapChannelSecrets as n, iterateBootstrapChannelPlugins as r, getBootstrapChannelPlugin as t };
