// openclaw/dist/bundled-CdLjswoh.js
import { i as formatErrorMessage } from "./errors-D8p6rxH8.js";
import { i as normalizeLowercaseStringOrEmpty } from "./string-coerce-BUSzWgUA.js";
import { i as openBoundaryFileSync } from "./boundary-file-read-C2FDLxpR.js";
import { n as resolveOpenClawPackageRootSync } from "./openclaw-root-DtKxDfkA.js";
import { t as createSubsystemLogger } from "./subsystem-D2gFnOcq.js";
import { n as buildPluginLoaderJitiOptions, u as resolvePluginLoaderJitiConfig } from "./sdk-alias-BIcUDpnA.js";
import { n as listBundledPluginMetadata, r as resolveBundledPluginGeneratedPath } from "./bundled-plugin-metadata-uzOGdZK8.js";
import { createRequire } from "node:module";
import { fileURLToPath } from "node:url";
import fs from "node:fs";
import path from "node:path";
import { createJiti } from "jiti";
//#region src/plugins/bundled-channel-runtime.ts
function listBundledChannelPluginMetadata(params) {
	return listBundledPluginMetadata(params);
}
function resolveBundledChannelGeneratedPath(rootDir, entry, pluginDirName) {
	return resolveBundledPluginGeneratedPath(rootDir, entry, pluginDirName);
}
//#endregion
//#region src/plugins/module-export.ts
function unwrapDefaultModuleExport(moduleExport) {
	let resolved = moduleExport;
	const seen = /* @__PURE__ */ new Set();
	while (resolved && typeof resolved === "object" && "default" in resolved && !seen.has(resolved)) {
		seen.add(resolved);
		resolved = resolved.default;
	}
	return resolved;
}
//#endregion
//#region src/channels/plugins/module-loader.ts
const nodeRequire = createRequire(import.meta.url);
function createModuleLoader() {
	const jitiLoaders = /* @__PURE__ */ new Map();
	return (modulePath) => {
		const { tryNative, aliasMap, cacheKey } = resolvePluginLoaderJitiConfig({
			modulePath,
			argv1: process.argv[1],
			moduleUrl: import.meta.url,
			preferBuiltDist: true
		});
		const cached = jitiLoaders.get(cacheKey);
		if (cached) return cached;
		const loader = createJiti(import.meta.url, {
			...buildPluginLoaderJitiOptions(aliasMap),
			tryNative
		});
		jitiLoaders.set(cacheKey, loader);
		return loader;
	};
}
let loadModule = createModuleLoader();
function isJavaScriptModulePath(modulePath) {
	return [
		".js",
		".mjs",
		".cjs"
	].includes(normalizeLowercaseStringOrEmpty(path.extname(modulePath)));
}
function resolvePluginModuleCandidates(rootDir, specifier) {
	const normalizedSpecifier = specifier.replace(/\\/g, "/");
	const resolvedPath = path.resolve(rootDir, normalizedSpecifier);
	if (path.extname(resolvedPath)) return [resolvedPath];
	return [
		resolvedPath,
		`${resolvedPath}.ts`,
		`${resolvedPath}.js`,
		`${resolvedPath}.mjs`,
		`${resolvedPath}.cjs`
	];
}
function resolveExistingPluginModulePath(rootDir, specifier) {
	for (const candidate of resolvePluginModuleCandidates(rootDir, specifier)) if (fs.existsSync(candidate)) return candidate;
	return path.resolve(rootDir, specifier);
}
function loadChannelPluginModule(params) {
	const opened = openBoundaryFileSync({
		absolutePath: params.modulePath,
		rootPath: params.boundaryRootDir ?? params.rootDir,
		boundaryLabel: params.boundaryLabel ?? "plugin root",
		rejectHardlinks: false,
		skipLexicalRootCheck: true
	});
	if (!opened.ok) throw new Error(`${params.boundaryLabel ?? "plugin"} module path escapes plugin root or fails alias checks`);
	const safePath = opened.path;
	fs.closeSync(opened.fd);
	if (process.platform === "win32" && params.shouldTryNativeRequire?.(safePath)) try {
		return nodeRequire(safePath);
	} catch {}
	return loadModule(safePath)(safePath);
}
//#endregion
//#region src/channels/plugins/bundled.ts
const log = createSubsystemLogger("channels");
const OPENCLAW_PACKAGE_ROOT = resolveOpenClawPackageRootSync({
	argv1: process.argv[1],
	cwd: process.cwd(),
	moduleUrl: import.meta.url.startsWith("file:") ? import.meta.url : void 0
}) ?? (import.meta.url.startsWith("file:") ? path.resolve(fileURLToPath(new URL("../../..", import.meta.url))) : process.cwd());
function resolveChannelPluginModuleEntry(moduleExport) {
	const resolved = unwrapDefaultModuleExport(moduleExport);
	if (!resolved || typeof resolved !== "object") return null;
	const record = resolved;
	if (record.kind !== "bundled-channel-entry") return null;
	if (typeof record.id !== "string" || typeof record.name !== "string" || typeof record.description !== "string" || typeof record.register !== "function" || typeof record.loadChannelPlugin !== "function") return null;
	return record;
}
function resolveChannelSetupModuleEntry(moduleExport) {
	const resolved = unwrapDefaultModuleExport(moduleExport);
	if (!resolved || typeof resolved !== "object") return null;
	const record = resolved;
	if (record.kind !== "bundled-channel-setup-entry") return null;
	if (typeof record.loadSetupPlugin !== "function") return null;
	return record;
}
function resolveBundledChannelBoundaryRoot(params) {
	const distRoot = path.resolve(OPENCLAW_PACKAGE_ROOT, "dist", "extensions", params.metadata.dirName);
	if (params.modulePath === distRoot || params.modulePath.startsWith(`${distRoot}${path.sep}`)) return distRoot;
	return path.resolve(OPENCLAW_PACKAGE_ROOT, "extensions", params.metadata.dirName);
}
function resolveGeneratedBundledChannelModulePath(params) {
	if (!params.entry) return null;
	const resolved = resolveBundledChannelGeneratedPath(OPENCLAW_PACKAGE_ROOT, params.entry, params.metadata.dirName);
	if (resolved) return resolved;
	return null;
}
function loadGeneratedBundledChannelModule(params) {
	const modulePath = resolveGeneratedBundledChannelModulePath(params);
	if (!modulePath) throw new Error(`missing generated module for bundled channel ${params.metadata.manifest.id}`);
	return loadChannelPluginModule({
		modulePath,
		rootDir: resolveBundledChannelBoundaryRoot({
			metadata: params.metadata,
			modulePath
		}),
		boundaryRootDir: resolveBundledChannelBoundaryRoot({
			metadata: params.metadata,
			modulePath
		}),
		shouldTryNativeRequire: (safePath) => safePath.includes(`${path.sep}dist${path.sep}`) && isJavaScriptModulePath(safePath)
	});
}
function loadGeneratedBundledChannelEntry(params) {
	try {
		const entry = resolveChannelPluginModuleEntry(loadGeneratedBundledChannelModule({
			metadata: params.metadata,
			entry: params.metadata.source
		}));
		if (!entry) {
			log.warn(`[channels] bundled channel entry ${params.metadata.manifest.id} missing bundled-channel-entry contract; skipping`);
			return null;
		}
		const setupEntry = params.includeSetup && params.metadata.setupSource ? resolveChannelSetupModuleEntry(loadGeneratedBundledChannelModule({
			metadata: params.metadata,
			entry: params.metadata.setupSource
		})) : null;
		return {
			id: params.metadata.manifest.id,
			entry,
			...setupEntry ? { setupEntry } : {}
		};
	} catch (error) {
		const detail = formatErrorMessage(error);
		log.warn(`[channels] failed to load bundled channel ${params.metadata.manifest.id}: ${detail}`);
		return null;
	}
}
let cachedBundledChannelMetadata = null;
function listBundledChannelMetadata() {
	cachedBundledChannelMetadata ??= listBundledChannelPluginMetadata({
		includeChannelConfigs: false,
		includeSyntheticChannelConfigs: false
	}).filter((metadata) => (metadata.manifest.channels?.length ?? 0) > 0);
	return cachedBundledChannelMetadata;
}
function listBundledChannelPluginIds() {
	return listBundledChannelMetadata().map((metadata) => metadata.manifest.id).toSorted((left, right) => left.localeCompare(right));
}
const pluginLoadInProgressIds = /* @__PURE__ */ new Set();
const setupPluginLoadInProgressIds = /* @__PURE__ */ new Set();
const entryLoadInProgressIds = /* @__PURE__ */ new Set();
const lazyEntriesById = /* @__PURE__ */ new Map();
const lazyPluginsById = /* @__PURE__ */ new Map();
const lazySetupPluginsById = /* @__PURE__ */ new Map();
const lazySecretsById = /* @__PURE__ */ new Map();
const lazySetupSecretsById = /* @__PURE__ */ new Map();
function resolveBundledChannelMetadata(id) {
	return listBundledChannelMetadata().find((metadata) => metadata.manifest.id === id || metadata.manifest.channels?.includes(id));
}
function getLazyGeneratedBundledChannelEntry(id, params) {
	const cached = lazyEntriesById.get(id);
	if (cached && (!params?.includeSetup || cached.setupEntry)) return cached;
	if (cached === null && !params?.includeSetup) return null;
	const metadata = resolveBundledChannelMetadata(id);
	if (!metadata) {
		lazyEntriesById.set(id, null);
		return null;
	}
	if (entryLoadInProgressIds.has(id)) return null;
	entryLoadInProgressIds.add(id);
	try {
		const entry = loadGeneratedBundledChannelEntry({
			metadata,
			includeSetup: params?.includeSetup === true
		});
		lazyEntriesById.set(id, entry);
		if (entry?.entry.id && entry.entry.id !== id) lazyEntriesById.set(entry.entry.id, entry);
		return entry;
	} finally {
		entryLoadInProgressIds.delete(id);
	}
}
function listBundledChannelPlugins() {
	return listBundledChannelPluginIds().flatMap((id) => {
		const plugin = getBundledChannelPlugin(id);
		return plugin ? [plugin] : [];
	});
}
function listBundledChannelSetupPlugins() {
	return listBundledChannelPluginIds().flatMap((id) => {
		const plugin = getBundledChannelSetupPlugin(id);
		return plugin ? [plugin] : [];
	});
}
function getBundledChannelPlugin(id) {
	const cached = lazyPluginsById.get(id);
	if (cached) return cached;
	if (pluginLoadInProgressIds.has(id)) return;
	const entry = getLazyGeneratedBundledChannelEntry(id)?.entry;
	if (!entry) return;
	pluginLoadInProgressIds.add(id);
	try {
		const plugin = entry.loadChannelPlugin();
		lazyPluginsById.set(id, plugin);
		return plugin;
	} finally {
		pluginLoadInProgressIds.delete(id);
	}
}
function getBundledChannelSecrets(id) {
	if (lazySecretsById.has(id)) return lazySecretsById.get(id) ?? void 0;
	const entry = getLazyGeneratedBundledChannelEntry(id)?.entry;
	if (!entry) return;
	const secrets = entry.loadChannelSecrets?.() ?? getBundledChannelPlugin(id)?.secrets;
	lazySecretsById.set(id, secrets ?? null);
	return secrets;
}
function getBundledChannelSetupPlugin(id) {
	const cached = lazySetupPluginsById.get(id);
	if (cached) return cached;
	if (setupPluginLoadInProgressIds.has(id)) return;
	const entry = getLazyGeneratedBundledChannelEntry(id, { includeSetup: true })?.setupEntry;
	if (!entry) return;
	setupPluginLoadInProgressIds.add(id);
	try {
		const plugin = entry.loadSetupPlugin();
		lazySetupPluginsById.set(id, plugin);
		return plugin;
	} finally {
		setupPluginLoadInProgressIds.delete(id);
	}
}
function getBundledChannelSetupSecrets(id) {
	if (lazySetupSecretsById.has(id)) return lazySetupSecretsById.get(id) ?? void 0;
	const entry = getLazyGeneratedBundledChannelEntry(id, { includeSetup: true })?.setupEntry;
	if (!entry) return;
	const secrets = entry.loadSetupSecrets?.() ?? getBundledChannelSetupPlugin(id)?.secrets;
	lazySetupSecretsById.set(id, secrets ?? null);
	return secrets;
}
function requireBundledChannelPlugin(id) {
	const plugin = getBundledChannelPlugin(id);
	if (!plugin) throw new Error(`missing bundled channel plugin: ${id}`);
	return plugin;
}
function setBundledChannelRuntime(id, runtime) {
	const setter = getLazyGeneratedBundledChannelEntry(id)?.entry.setChannelRuntime;
	if (!setter) throw new Error(`missing bundled channel runtime setter: ${id}`);
	setter(runtime);
}
//#endregion
export { listBundledChannelPluginIds as a, requireBundledChannelPlugin as c, loadChannelPluginModule as d, resolveExistingPluginModulePath as f, getBundledChannelSetupSecrets as i, setBundledChannelRuntime as l, getBundledChannelSecrets as n, listBundledChannelPlugins as o, unwrapDefaultModuleExport as p, getBundledChannelSetupPlugin as r, listBundledChannelSetupPlugins as s, getBundledChannelPlugin as t, isJavaScriptModulePath as u };
