// openclaw/dist/bus-server-2ab_NN8Y.js
import { i as formatErrorMessage } from "./errors-D8p6rxH8.js";
import "./error-runtime-Da0X1TOQ.js";
import { createServer } from "node:http";
//#region extensions/qa-lab/src/bus-server.ts
async function readJson(req) {
	const chunks = [];
	for await (const chunk of req) chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
	const text = Buffer.concat(chunks).toString("utf8").trim();
	return text ? JSON.parse(text) : {};
}
function writeJson(res, statusCode, body) {
	const payload = JSON.stringify(body);
	res.writeHead(statusCode, {
		"content-type": "application/json; charset=utf-8",
		"content-length": Buffer.byteLength(payload)
	});
	res.end(payload);
}
function writeError(res, statusCode, error) {
	writeJson(res, statusCode, { error: formatErrorMessage(error) });
}
async function closeQaHttpServer(server) {
	let forceCloseTimer;
	try {
		await new Promise((resolve, reject) => {
			server.close((error) => error ? reject(error) : resolve());
			server.closeIdleConnections?.();
			forceCloseTimer = setTimeout(() => {
				server.closeAllConnections?.();
			}, 250);
			forceCloseTimer.unref();
		});
	} finally {
		if (forceCloseTimer) clearTimeout(forceCloseTimer);
	}
}
async function handleQaBusRequest(params) {
	const method = params.req.method ?? "GET";
	const url = new URL(params.req.url ?? "/", "http://127.0.0.1");
	if (method === "GET" && url.pathname === "/health") {
		writeJson(params.res, 200, { ok: true });
		return true;
	}
	if (method === "GET" && url.pathname === "/v1/state") {
		writeJson(params.res, 200, params.state.getSnapshot());
		return true;
	}
	if (!url.pathname.startsWith("/v1/")) return false;
	if (method !== "POST") {
		writeError(params.res, 405, "method not allowed");
		return true;
	}
	const body = await readJson(params.req);
	try {
		switch (url.pathname) {
			case "/v1/reset":
				params.state.reset();
				writeJson(params.res, 200, { ok: true });
				return true;
			case "/v1/inbound/message":
				writeJson(params.res, 200, { message: params.state.addInboundMessage(body) });
				return true;
			case "/v1/outbound/message":
				writeJson(params.res, 200, { message: params.state.addOutboundMessage(body) });
				return true;
			case "/v1/actions/thread-create":
				writeJson(params.res, 200, { thread: params.state.createThread(body) });
				return true;
			case "/v1/actions/react":
				writeJson(params.res, 200, { message: params.state.reactToMessage(body) });
				return true;
			case "/v1/actions/edit":
				writeJson(params.res, 200, { message: params.state.editMessage(body) });
				return true;
			case "/v1/actions/delete":
				writeJson(params.res, 200, { message: params.state.deleteMessage(body) });
				return true;
			case "/v1/actions/read":
				writeJson(params.res, 200, { message: params.state.readMessage(body) });
				return true;
			case "/v1/actions/search":
				writeJson(params.res, 200, { messages: params.state.searchMessages(body) });
				return true;
			case "/v1/poll": {
				const input = body;
				const timeoutMs = Math.max(0, Math.min(input.timeoutMs ?? 0, 3e4));
				const initial = params.state.poll(input);
				if (initial.events.length > 0 || timeoutMs === 0) {
					writeJson(params.res, 200, initial);
					return true;
				}
				try {
					await params.state.waitFor({
						kind: "event-kind",
						eventKind: "inbound-message",
						timeoutMs
					});
				} catch {}
				writeJson(params.res, 200, params.state.poll(input));
				return true;
			}
			case "/v1/wait":
				writeJson(params.res, 200, { match: await params.state.waitFor(body) });
				return true;
			default:
				writeError(params.res, 404, "not found");
				return true;
		}
	} catch (error) {
		writeError(params.res, 400, error);
		return true;
	}
}
function createQaBusServer(state) {
	return createServer(async (req, res) => {
		if (!await handleQaBusRequest({
			req,
			res,
			state
		})) writeError(res, 404, "not found");
	});
}
async function startQaBusServer(params) {
	const server = createQaBusServer(params.state);
	await new Promise((resolve, reject) => {
		server.once("error", reject);
		server.listen(params.port ?? 0, "127.0.0.1", () => resolve());
	});
	const address = server.address();
	if (!address || typeof address === "string") throw new Error("qa-bus failed to bind");
	return {
		server,
		port: address.port,
		baseUrl: `http://127.0.0.1:${address.port}`,
		async stop() {
			await closeQaHttpServer(server);
		}
	};
}
//#endregion
export { writeError as a, startQaBusServer as i, createQaBusServer as n, writeJson as o, handleQaBusRequest as r, closeQaHttpServer as t };
