// openclaw/dist/channel-reply-pipeline-CJZpAxBU.js
import { i as normalizeChannelId, t as getChannelPlugin } from "./registry-CAnQx2oU.js";
import "./plugins-DB_0wnJl.js";
import { n as createReplyPrefixOptions } from "./reply-prefix-DWLvut12.js";
import { t as createTypingCallbacks } from "./typing-3TQsvlmc.js";
//#region src/plugin-sdk/channel-reply-pipeline.ts
function createChannelReplyPipeline(params) {
	const channelId = params.channel ? normalizeChannelId(params.channel) ?? params.channel : void 0;
	const plugin = params.transformReplyPayload ? void 0 : channelId ? getChannelPlugin(channelId) : void 0;
	const transformReplyPayload = params.transformReplyPayload ?? (plugin?.messaging?.transformReplyPayload ? (payload) => plugin.messaging?.transformReplyPayload?.({
		payload,
		cfg: params.cfg,
		accountId: params.accountId
	}) ?? payload : void 0);
	return {
		...createReplyPrefixOptions({
			cfg: params.cfg,
			agentId: params.agentId,
			channel: params.channel,
			accountId: params.accountId
		}),
		...transformReplyPayload ? { transformReplyPayload } : {},
		...params.typingCallbacks ? { typingCallbacks: params.typingCallbacks } : params.typing ? { typingCallbacks: createTypingCallbacks(params.typing) } : {}
	};
}
//#endregion
export { createChannelReplyPipeline as t };
