// openclaw/dist/channel-plugin-ids-D4rzcwpO.js
import { n as loadPluginManifestRegistry } from "./manifest-registry-DRnxzk1J.js";
import { r as hasKind } from "./slots-V5i3G255.js";
import { a as normalizePluginsConfig, i as normalizePluginId, n as createPluginActivationSource, s as resolveEffectivePluginActivationState } from "./config-state-DmPson1J.js";
import { n as listPotentialConfiguredChannelIds } from "./config-presence-Bi1jPo3m.js";
import { I as resolveMemoryDreamingConfig, L as resolveMemoryDreamingPluginConfig, R as resolveMemoryDreamingPluginId } from "./dreaming-BqljXKK2.js";
//#region src/plugins/channel-plugin-ids.ts
function hasRuntimeContractSurface(plugin) {
	return Boolean(plugin.providers.length > 0 || plugin.cliBackends.length > 0 || plugin.contracts?.speechProviders?.length || plugin.contracts?.mediaUnderstandingProviders?.length || plugin.contracts?.imageGenerationProviders?.length || plugin.contracts?.videoGenerationProviders?.length || plugin.contracts?.musicGenerationProviders?.length || plugin.contracts?.webFetchProviders?.length || plugin.contracts?.webSearchProviders?.length || plugin.contracts?.memoryEmbeddingProviders?.length || hasKind(plugin.kind, "memory"));
}
function isGatewayStartupMemoryPlugin(plugin) {
	return hasKind(plugin.kind, "memory");
}
function isGatewayStartupSidecar(plugin) {
	return plugin.channels.length === 0 && !hasRuntimeContractSurface(plugin);
}
function resolveGatewayStartupDreamingPluginIds(config) {
	if (!resolveMemoryDreamingConfig({
		pluginConfig: resolveMemoryDreamingPluginConfig(config),
		cfg: config
	}).enabled) return /* @__PURE__ */ new Set();
	return new Set(["memory-core", resolveMemoryDreamingPluginId(config)]);
}
function resolveExplicitMemorySlotStartupPluginId(config) {
	const configuredSlot = config.plugins?.slots?.memory?.trim();
	if (!configuredSlot || configuredSlot.toLowerCase() === "none") return;
	return normalizePluginId(configuredSlot);
}
function shouldConsiderForGatewayStartup(params) {
	if (isGatewayStartupSidecar(params.plugin)) return true;
	if (!isGatewayStartupMemoryPlugin(params.plugin)) return false;
	if (params.startupDreamingPluginIds.has(params.plugin.id)) return true;
	return params.explicitMemorySlotStartupPluginId === params.plugin.id;
}
function resolveChannelPluginIds(params) {
	return loadPluginManifestRegistry({
		config: params.config,
		workspaceDir: params.workspaceDir,
		env: params.env
	}).plugins.filter((plugin) => plugin.channels.length > 0).map((plugin) => plugin.id);
}
function resolveConfiguredChannelPluginIds(params) {
	const configuredChannelIds = new Set(listPotentialConfiguredChannelIds(params.config, params.env).map((id) => id.trim()));
	if (configuredChannelIds.size === 0) return [];
	return resolveChannelPluginIds(params).filter((pluginId) => configuredChannelIds.has(pluginId));
}
function resolveConfiguredDeferredChannelPluginIds(params) {
	const configuredChannelIds = new Set(listPotentialConfiguredChannelIds(params.config, params.env).map((id) => id.trim()));
	if (configuredChannelIds.size === 0) return [];
	return loadPluginManifestRegistry({
		config: params.config,
		workspaceDir: params.workspaceDir,
		env: params.env
	}).plugins.filter((plugin) => plugin.channels.some((channelId) => configuredChannelIds.has(channelId)) && plugin.startupDeferConfiguredChannelFullLoadUntilAfterListen === true).map((plugin) => plugin.id);
}
function resolveGatewayStartupPluginIds(params) {
	const configuredChannelIds = new Set(listPotentialConfiguredChannelIds(params.config, params.env).map((id) => id.trim()));
	const pluginsConfig = normalizePluginsConfig(params.config.plugins);
	const activationSource = createPluginActivationSource({ config: params.activationSourceConfig ?? params.config });
	const startupDreamingPluginIds = resolveGatewayStartupDreamingPluginIds(params.config);
	const explicitMemorySlotStartupPluginId = resolveExplicitMemorySlotStartupPluginId(params.activationSourceConfig ?? params.config);
	return loadPluginManifestRegistry({
		config: params.config,
		workspaceDir: params.workspaceDir,
		env: params.env
	}).plugins.filter((plugin) => {
		if (plugin.channels.some((channelId) => configuredChannelIds.has(channelId))) return true;
		if (!shouldConsiderForGatewayStartup({
			plugin,
			startupDreamingPluginIds,
			explicitMemorySlotStartupPluginId
		})) return false;
		const activationState = resolveEffectivePluginActivationState({
			id: plugin.id,
			origin: plugin.origin,
			config: pluginsConfig,
			rootConfig: params.config,
			enabledByDefault: plugin.enabledByDefault,
			activationSource
		});
		if (!activationState.enabled) return false;
		if (plugin.origin !== "bundled") return activationState.explicitlyEnabled;
		return activationState.source === "explicit" || activationState.source === "default";
	}).map((plugin) => plugin.id);
}
//#endregion
export { resolveGatewayStartupPluginIds as i, resolveConfiguredChannelPluginIds as n, resolveConfiguredDeferredChannelPluginIds as r, resolveChannelPluginIds as t };
