// openclaw/dist/cli-Cb-0l48h.js
import { n as hasQaScenarioPack } from "./scenario-catalog-DQ-BbL_S.js";
//#region extensions/qa-lab/src/cli-options.ts
function collectString(value, previous) {
	const trimmed = value.trim();
	return trimmed ? [...previous, trimmed] : previous;
}
//#endregion
//#region extensions/qa-lab/src/live-transports/shared/live-transport-cli.ts
function createLazyCliRuntimeLoader(load) {
	let promise = null;
	return async () => {
		promise ??= load();
		return await promise;
	};
}
function mapLiveTransportQaCommanderOptions(opts) {
	return {
		repoRoot: opts.repoRoot,
		outputDir: opts.outputDir,
		providerMode: opts.providerMode,
		primaryModel: opts.model,
		alternateModel: opts.altModel,
		fastMode: opts.fast,
		scenarioIds: opts.scenario,
		sutAccountId: opts.sutAccount
	};
}
function registerLiveTransportQaCli(params) {
	params.qa.command(params.commandName).description(params.description).option("--repo-root <path>", "Repository root to target when running from a neutral cwd").option("--output-dir <path>", params.outputDirHelp).option("--provider-mode <mode>", "Provider mode: mock-openai or live-frontier (legacy live-openai still works)", "live-frontier").option("--model <ref>", "Primary provider/model ref").option("--alt-model <ref>", "Alternate provider/model ref").option("--scenario <id>", params.scenarioHelp, collectString, []).option("--fast", "Enable provider fast mode where supported", false).option("--sut-account <id>", params.sutAccountHelp, "sut").action(async (opts) => {
		await params.run(mapLiveTransportQaCommanderOptions(opts));
	});
}
function createLiveTransportQaCliRegistration(params) {
	return {
		commandName: params.commandName,
		register(qa) {
			registerLiveTransportQaCli({
				qa,
				commandName: params.commandName,
				description: params.description,
				outputDirHelp: params.outputDirHelp,
				scenarioHelp: params.scenarioHelp,
				sutAccountHelp: params.sutAccountHelp,
				run: params.run
			});
		}
	};
}
//#endregion
//#region extensions/qa-lab/src/live-transports/matrix/cli.ts
const loadMatrixQaCliRuntime = createLazyCliRuntimeLoader(() => import("./cli.runtime-BBLS9wrA.js"));
async function runQaMatrix(opts) {
	await (await loadMatrixQaCliRuntime()).runQaMatrixCommand(opts);
}
const matrixQaCliRegistration = createLiveTransportQaCliRegistration({
	commandName: "matrix",
	description: "Run the Docker-backed Matrix live QA lane against a disposable homeserver",
	outputDirHelp: "Matrix QA artifact directory",
	scenarioHelp: "Run only the named Matrix QA scenario (repeatable)",
	sutAccountHelp: "Temporary Matrix account id inside the QA gateway config",
	run: runQaMatrix
});
//#endregion
//#region extensions/qa-lab/src/live-transports/telegram/cli.ts
const loadTelegramQaCliRuntime = createLazyCliRuntimeLoader(() => import("./cli.runtime-CMiWJfIX.js"));
async function runQaTelegram(opts) {
	await (await loadTelegramQaCliRuntime()).runQaTelegramCommand(opts);
}
//#endregion
//#region extensions/qa-lab/src/live-transports/cli.ts
const LIVE_TRANSPORT_QA_CLI_REGISTRATIONS = [createLiveTransportQaCliRegistration({
	commandName: "telegram",
	description: "Run the manual Telegram live QA lane against a private bot-to-bot group harness",
	outputDirHelp: "Telegram QA artifact directory",
	scenarioHelp: "Run only the named Telegram QA scenario (repeatable)",
	sutAccountHelp: "Temporary Telegram account id inside the QA gateway config",
	run: runQaTelegram
}), matrixQaCliRegistration];
//#endregion
//#region extensions/qa-lab/src/cli.ts
let qaLabCliRuntimePromise = null;
async function loadQaLabCliRuntime() {
	qaLabCliRuntimePromise ??= import("./cli.runtime-CkfLDPdp.js");
	return await qaLabCliRuntimePromise;
}
async function runQaSelfCheck(opts) {
	await (await loadQaLabCliRuntime()).runQaLabSelfCheckCommand(opts);
}
async function runQaSuite(opts) {
	await (await loadQaLabCliRuntime()).runQaSuiteCommand(opts);
}
async function runQaParityReport(opts) {
	await (await loadQaLabCliRuntime()).runQaParityReportCommand(opts);
}
async function runQaCharacterEval(opts) {
	await (await loadQaLabCliRuntime()).runQaCharacterEvalCommand(opts);
}
async function runQaManualLane(opts) {
	await (await loadQaLabCliRuntime()).runQaManualLaneCommand(opts);
}
async function runQaUi(opts) {
	await (await loadQaLabCliRuntime()).runQaLabUiCommand(opts);
}
async function runQaDockerScaffold(opts) {
	await (await loadQaLabCliRuntime()).runQaDockerScaffoldCommand(opts);
}
async function runQaDockerBuildImage(opts) {
	await (await loadQaLabCliRuntime()).runQaDockerBuildImageCommand(opts);
}
async function runQaDockerUp(opts) {
	await (await loadQaLabCliRuntime()).runQaDockerUpCommand(opts);
}
async function runQaMockOpenAi(opts) {
	await (await loadQaLabCliRuntime()).runQaMockOpenAiCommand(opts);
}
function isQaLabCliAvailable() {
	return hasQaScenarioPack();
}
function registerQaLabCli(program) {
	const qa = program.command("qa").description("Run private QA automation flows and launch the QA debugger");
	qa.command("run").description("Run the bundled QA self-check and write a Markdown report").option("--repo-root <path>", "Repository root to target when running from a neutral cwd").option("--output <path>", "Report output path").action(async (opts) => {
		await runQaSelfCheck(opts);
	});
	qa.command("suite").description("Run repo-backed QA scenarios against the QA gateway lane").option("--repo-root <path>", "Repository root to target when running from a neutral cwd").option("--output-dir <path>", "Suite artifact directory").option("--runner <kind>", "Execution runner: host or multipass", "host").option("--provider-mode <mode>", "Provider mode: mock-openai or live-frontier (legacy live-openai still works)", "mock-openai").option("--model <ref>", "Primary provider/model ref").option("--alt-model <ref>", "Alternate provider/model ref").option("--cli-auth-mode <mode>", "CLI backend auth mode for live Claude CLI runs: auto, api-key, or subscription").option("--parity-pack <name>", "Preset scenario pack; currently only \"agentic\" is supported").option("--scenario <id>", "Run only the named QA scenario (repeatable)", collectString, []).option("--concurrency <count>", "Scenario worker concurrency", (value) => Number(value)).option("--fast", "Enable provider fast mode where supported", false).option("--image <alias>", "Multipass image alias").option("--cpus <count>", "Multipass vCPU count", (value) => Number(value)).option("--memory <size>", "Multipass memory size").option("--disk <size>", "Multipass disk size").action(async (opts) => {
		await runQaSuite({
			repoRoot: opts.repoRoot,
			outputDir: opts.outputDir,
			runner: opts.runner,
			providerMode: opts.providerMode,
			primaryModel: opts.model,
			alternateModel: opts.altModel,
			fastMode: opts.fast,
			cliAuthMode: opts.cliAuthMode,
			parityPack: opts.parityPack,
			scenarioIds: opts.scenario,
			concurrency: opts.concurrency,
			image: opts.image,
			cpus: opts.cpus,
			memory: opts.memory,
			disk: opts.disk
		});
	});
	qa.command("parity-report").description("Compare two QA suite summaries and write an agentic parity gate report").requiredOption("--candidate-summary <path>", "Candidate qa-suite-summary.json path").requiredOption("--baseline-summary <path>", "Baseline qa-suite-summary.json path").option("--repo-root <path>", "Repository root to target when running from a neutral cwd").option("--candidate-label <label>", "Candidate display label", "openai/gpt-5.4").option("--baseline-label <label>", "Baseline display label", "anthropic/claude-opus-4-6").option("--output-dir <path>", "Artifact directory for the parity report").action(async (opts) => {
		await runQaParityReport(opts);
	});
	for (const lane of LIVE_TRANSPORT_QA_CLI_REGISTRATIONS) lane.register(qa);
	qa.command("character-eval").description("Run the character QA scenario across live models and write a judged report").option("--repo-root <path>", "Repository root to target when running from a neutral cwd").option("--output-dir <path>", "Character eval artifact directory").option("--model <ref[,option]>", "Provider/model ref to evaluate; options: thinking=<level>, fast, no-fast, fast=<bool>", collectString, []).option("--scenario <id>", "Character scenario id", "character-vibes-gollum").option("--fast", "Enable provider fast mode for all candidate runs").option("--thinking <level>", "Candidate thinking default: off|minimal|low|medium|high|xhigh|adaptive").option("--model-thinking <ref=level>", "Deprecated: candidate thinking override for one model ref (repeatable)", collectString, []).option("--judge-model <ref[,option]>", "Judge provider/model ref; options: thinking=<level>, fast, no-fast, fast=<bool> (repeatable)", collectString, []).option("--judge-timeout-ms <ms>", "Override judge wait timeout", (value) => Number(value)).option("--blind-judge-models", "Hide candidate model refs from judge prompts; reports still map rankings back to real refs").option("--concurrency <count>", "Candidate model run concurrency", (value) => Number(value)).option("--judge-concurrency <count>", "Judge model run concurrency", (value) => Number(value)).action(async (opts) => {
		await runQaCharacterEval(opts);
	});
	qa.command("manual").description("Run a one-off QA agent prompt against the selected provider/model lane").requiredOption("--message <text>", "Prompt to send to the QA agent").option("--repo-root <path>", "Repository root to target when running from a neutral cwd").option("--provider-mode <mode>", "Provider mode: mock-openai or live-frontier (legacy live-openai still works)", "live-frontier").option("--model <ref>", "Primary provider/model ref (defaults by provider mode)").option("--alt-model <ref>", "Alternate provider/model ref").option("--fast", "Enable provider fast mode where supported", false).option("--timeout-ms <ms>", "Override agent.wait timeout", (value) => Number(value)).action(async (opts) => {
		await runQaManualLane({
			repoRoot: opts.repoRoot,
			providerMode: opts.providerMode,
			primaryModel: opts.model,
			alternateModel: opts.altModel,
			fastMode: opts.fast,
			message: opts.message,
			timeoutMs: opts.timeoutMs
		});
	});
	qa.command("ui").description("Start the private QA debugger UI and local QA bus").option("--repo-root <path>", "Repository root to target when running from a neutral cwd").option("--host <host>", "Bind host", "127.0.0.1").option("--port <port>", "Bind port", (value) => Number(value)).option("--advertise-host <host>", "Optional public host to advertise in bootstrap payloads").option("--advertise-port <port>", "Optional public port to advertise", (value) => Number(value)).option("--control-ui-url <url>", "Optional Control UI URL to embed beside the QA panel").option("--control-ui-token <token>", "Optional Control UI token for embedded links").option("--control-ui-proxy-target <url>", "Optional upstream Control UI target for /control-ui proxying").option("--ui-dist-dir <path>", "Optional QA Lab UI asset directory override").option("--auto-kickoff-target <kind>", "Kickoff default target (direct or channel)").option("--embedded-gateway <mode>", "Embedded gateway mode hint", "enabled").option("--send-kickoff-on-start", "Inject the repo-backed kickoff task when the UI starts", false).action(async (opts) => {
		await runQaUi(opts);
	});
	qa.command("docker-scaffold").description("Write a prebaked Docker scaffold for the QA dashboard + gateway lane").option("--repo-root <path>", "Repository root to target when running from a neutral cwd").requiredOption("--output-dir <path>", "Output directory for docker-compose + state files").option("--gateway-port <port>", "Gateway host port", (value) => Number(value)).option("--qa-lab-port <port>", "QA lab host port", (value) => Number(value)).option("--provider-base-url <url>", "Provider base URL for the QA gateway").option("--image <name>", "Prebaked image name", "openclaw:qa-local-prebaked").option("--use-prebuilt-image", "Use image: instead of build: in docker-compose", false).option("--bind-ui-dist", "Bind-mount extensions/qa-lab/web/dist into the qa-lab container for faster UI refresh", false).action(async (opts) => {
		await runQaDockerScaffold(opts);
	});
	qa.command("docker-build-image").description("Build the prebaked QA Docker image with qa-channel + qa-lab bundled").option("--repo-root <path>", "Repository root to target when running from a neutral cwd").option("--image <name>", "Image tag", "openclaw:qa-local-prebaked").action(async (opts) => {
		await runQaDockerBuildImage(opts);
	});
	qa.command("up").description("Build the QA site, start the Docker-backed QA stack, and print the QA Lab URL").option("--repo-root <path>", "Repository root to target when running from a neutral cwd").option("--output-dir <path>", "Output directory for docker-compose + state files").option("--gateway-port <port>", "Gateway host port", (value) => Number(value)).option("--qa-lab-port <port>", "QA lab host port", (value) => Number(value)).option("--provider-base-url <url>", "Provider base URL for the QA gateway").option("--image <name>", "Image tag", "openclaw:qa-local-prebaked").option("--use-prebuilt-image", "Use image: instead of build: in docker-compose", false).option("--bind-ui-dist", "Bind-mount extensions/qa-lab/web/dist into the qa-lab container for faster UI refresh", false).option("--skip-ui-build", "Skip pnpm qa:lab:build before starting Docker", false).action(async (opts) => {
		await runQaDockerUp(opts);
	});
	qa.command("mock-openai").description("Run the local mock OpenAI Responses API server for QA").option("--host <host>", "Bind host", "127.0.0.1").option("--port <port>", "Bind port", (value) => Number(value)).action(async (opts) => {
		await runQaMockOpenAi(opts);
	});
}
//#endregion
export { registerQaLabCli as n, isQaLabCliAvailable as t };
