// openclaw/dist/cli.runtime-CkfLDPdp.js
import { i as formatErrorMessage } from "./errors-D8p6rxH8.js";
import { n as resolvePreferredOpenClawTmpDir } from "./tmp-openclaw-dir-DgmFzeYV.js";
import "./temp-path-B39Iubqg.js";
import "./error-runtime-Da0X1TOQ.js";
import { t as startQaLabServer } from "./lab-server-Dq4QthbC.js";
import { r as isQaFastModeModelRef } from "./model-selection-ClMg2tje.js";
import { i as normalizeQaProviderMode, r as defaultQaModelForMode } from "./run-config-B359hkWD.js";
import { o as startQaGatewayChild, s as startQaMockOpenAiServer } from "./gateway-child-Du_AnSFD.js";
import { i as normalizeQaThinkingLevel } from "./qa-gateway-config-DunCiJBX.js";
import { n as writeQaDockerHarnessFiles, t as buildQaDockerHarnessImage } from "./docker-harness-DsccVB1q.js";
import { n as extractQaVisibleReplyLeakText, r as resolveQaLiveTurnTimeoutMs } from "./reply-failure-CNRsmk_Z.js";
import { t as resolveRepoRelativeOutputDir } from "./cli-paths-C0Fbq3dW.js";
import { a as waitForDockerServiceHealth, i as resolveHostPort, n as fetchHealthUrl, o as waitForHealth, r as resolveComposeServiceUrl, t as execCommand } from "./docker-runtime-C6coNoAQ.js";
import fs from "node:fs";
import path from "node:path";
import { execFile } from "node:child_process";
import os from "node:os";
import fs$1, { access, appendFile, mkdir, writeFile } from "node:fs/promises";
import { randomUUID } from "node:crypto";
import { setTimeout as setTimeout$1 } from "node:timers/promises";
//#region extensions/qa-lab/src/agentic-parity.ts
const QA_AGENTIC_PARITY_PACK = "agentic";
const QA_AGENTIC_PARITY_SCENARIOS = [
	{
		id: "approval-turn-tool-followthrough",
		title: "Approval turn tool followthrough"
	},
	{
		id: "model-switch-tool-continuity",
		title: "Model switch with tool continuity"
	},
	{
		id: "source-docs-discovery-report",
		title: "Source and docs discovery report"
	},
	{
		id: "image-understanding-attachment",
		title: "Image understanding from attachment"
	},
	{
		id: "compaction-retry-mutating-tool",
		title: "Compaction retry after mutating tool"
	}
];
const QA_AGENTIC_PARITY_SCENARIO_IDS = QA_AGENTIC_PARITY_SCENARIOS.map(({ id }) => id);
const QA_AGENTIC_PARITY_SCENARIO_TITLES = QA_AGENTIC_PARITY_SCENARIOS.map(({ title }) => title);
function resolveQaParityPackScenarioIds(params) {
	const normalizedPack = params.parityPack?.trim().toLowerCase();
	const explicitScenarioIds = [...new Set(params.scenarioIds ?? [])];
	if (!normalizedPack) return explicitScenarioIds;
	if (normalizedPack !== "agentic") throw new Error(`--parity-pack must be "${QA_AGENTIC_PARITY_PACK}", got "${params.parityPack}"`);
	return [...new Set([...explicitScenarioIds, ...QA_AGENTIC_PARITY_SCENARIO_IDS])];
}
//#endregion
//#region extensions/qa-lab/src/agentic-parity-report.ts
const UNINTENDED_STOP_PATTERNS = [
	/incomplete turn/i,
	/\btimed out\b/i,
	/\btimeout\b/i,
	/\bstopped\b/i,
	/\bblocked\b/i,
	/\babandoned\b/i,
	/did not continue/i
];
const SUSPICIOUS_PASS_PATTERNS = [
	/incomplete turn/i,
	/\btimed out\b/i,
	/\btimeout\b/i,
	/\bfailed to\b/i,
	/\bcould not\b/i,
	/\bunable to\b/i,
	/did not continue/i,
	/error occurred/i,
	/an error was/i
];
function normalizeScenarioStatus(status) {
	return status === "pass" || status === "fail" || status === "skip" ? status : "fail";
}
function scenarioText(scenario) {
	const parts = [scenario.details ?? ""];
	for (const step of scenario.steps ?? []) parts.push(step.details ?? "");
	return parts.filter(Boolean).join("\n");
}
function scenarioHasPattern(scenario, patterns) {
	const text = scenarioText(scenario);
	return text.length > 0 && patterns.some((pattern) => pattern.test(text));
}
function computeQaAgenticParityMetrics(summary) {
	const scenarios = summary.scenarios.map((scenario) => ({
		...scenario,
		status: normalizeScenarioStatus(scenario.status)
	}));
	const totalScenarios = summary.counts?.total ?? scenarios.length;
	const passedScenarios = summary.counts?.passed ?? scenarios.filter((scenario) => scenario.status === "pass").length;
	const failedScenarios = summary.counts?.failed ?? scenarios.filter((scenario) => scenario.status === "fail").length;
	const unintendedStopCount = scenarios.filter((scenario) => scenario.status !== "pass" && scenarioHasPattern(scenario, UNINTENDED_STOP_PATTERNS)).length;
	const fakeSuccessCount = scenarios.filter((scenario) => scenario.status === "pass" && scenarioHasPattern(scenario, SUSPICIOUS_PASS_PATTERNS)).length;
	const validToolCallCount = passedScenarios;
	const rate = (value) => totalScenarios > 0 ? value / totalScenarios : 0;
	return {
		totalScenarios,
		passedScenarios,
		failedScenarios,
		completionRate: rate(passedScenarios),
		unintendedStopCount,
		unintendedStopRate: rate(unintendedStopCount),
		validToolCallCount,
		validToolCallRate: rate(validToolCallCount),
		fakeSuccessCount
	};
}
function formatPercent(value) {
	return `${(value * 100).toFixed(1)}%`;
}
function requiredCoverageStatus(scenario) {
	return scenario ? normalizeScenarioStatus(scenario.status) : "missing";
}
function scopeSummaryToParityPack(summary, parityTitleSet) {
	return { scenarios: summary.scenarios.filter((scenario) => parityTitleSet.has(scenario.name)) };
}
function buildQaAgenticParityComparison(params) {
	const parityTitleSet = new Set(QA_AGENTIC_PARITY_SCENARIO_TITLES);
	const candidateMetrics = computeQaAgenticParityMetrics(scopeSummaryToParityPack(params.candidateSummary, parityTitleSet));
	const baselineMetrics = computeQaAgenticParityMetrics(scopeSummaryToParityPack(params.baselineSummary, parityTitleSet));
	const scenarioNames = new Set([
		...QA_AGENTIC_PARITY_SCENARIO_TITLES,
		...params.candidateSummary.scenarios.map((scenario) => scenario.name),
		...params.baselineSummary.scenarios.map((scenario) => scenario.name)
	]);
	const candidateByName = new Map(params.candidateSummary.scenarios.map((scenario) => [scenario.name, scenario]));
	const baselineByName = new Map(params.baselineSummary.scenarios.map((scenario) => [scenario.name, scenario]));
	const scenarioComparisons = [...scenarioNames].toSorted((left, right) => left.localeCompare(right)).map((name) => {
		const candidate = candidateByName.get(name);
		const baseline = baselineByName.get(name);
		return {
			name,
			candidateStatus: candidate ? normalizeScenarioStatus(candidate.status) : "missing",
			baselineStatus: baseline ? normalizeScenarioStatus(baseline.status) : "missing",
			...candidate?.details ? { candidateDetails: candidate.details } : {},
			...baseline?.details ? { baselineDetails: baseline.details } : {}
		};
	});
	const failures = [];
	const requiredScenarioCoverage = QA_AGENTIC_PARITY_SCENARIO_TITLES.map((name) => {
		const candidate = candidateByName.get(name);
		const baseline = baselineByName.get(name);
		return {
			name,
			candidateStatus: requiredCoverageStatus(candidate),
			baselineStatus: requiredCoverageStatus(baseline)
		};
	}).filter((scenario) => scenario.candidateStatus === "missing" || scenario.baselineStatus === "missing" || scenario.candidateStatus === "skip" || scenario.baselineStatus === "skip");
	for (const scenario of requiredScenarioCoverage) failures.push(`Missing required parity scenario coverage for ${scenario.name}: ${params.candidateLabel}=${scenario.candidateStatus}, ${params.baselineLabel}=${scenario.baselineStatus}.`);
	const coverageMismatch = scenarioComparisons.filter((scenario) => !parityTitleSet.has(scenario.name) && (scenario.candidateStatus === "missing" || scenario.baselineStatus === "missing"));
	for (const scenario of coverageMismatch) failures.push(`Scenario coverage mismatch for ${scenario.name}: ${params.candidateLabel}=${scenario.candidateStatus}, ${params.baselineLabel}=${scenario.baselineStatus}.`);
	if (candidateMetrics.completionRate < baselineMetrics.completionRate) failures.push(`${params.candidateLabel} completion rate ${formatPercent(candidateMetrics.completionRate)} is below ${params.baselineLabel} ${formatPercent(baselineMetrics.completionRate)}.`);
	if (candidateMetrics.unintendedStopRate > baselineMetrics.unintendedStopRate) failures.push(`${params.candidateLabel} unintended-stop rate ${formatPercent(candidateMetrics.unintendedStopRate)} exceeds ${params.baselineLabel} ${formatPercent(baselineMetrics.unintendedStopRate)}.`);
	if (candidateMetrics.validToolCallRate < baselineMetrics.validToolCallRate) failures.push(`${params.candidateLabel} valid-tool-call rate ${formatPercent(candidateMetrics.validToolCallRate)} is below ${params.baselineLabel} ${formatPercent(baselineMetrics.validToolCallRate)}.`);
	if (candidateMetrics.fakeSuccessCount > 0) failures.push(`${params.candidateLabel} produced ${candidateMetrics.fakeSuccessCount} suspicious pass result(s); fake-success count must be 0.`);
	if (baselineMetrics.fakeSuccessCount > 0) failures.push(`${params.baselineLabel} produced ${baselineMetrics.fakeSuccessCount} suspicious pass result(s); baseline fake-success count must also be 0.`);
	return {
		candidateLabel: params.candidateLabel,
		baselineLabel: params.baselineLabel,
		comparedAt: params.comparedAt ?? (/* @__PURE__ */ new Date()).toISOString(),
		candidateMetrics,
		baselineMetrics,
		scenarioComparisons,
		pass: failures.length === 0,
		failures,
		notes: ["First-wave valid-tool-call rate is scenario-level and uses passing tool-mediated scenarios as the verified numerator.", "Auth/proxy/DNS correctness is intentionally out of scope for this parity report and should be gated by the deterministic runtime-truthfulness suites."]
	};
}
function renderQaAgenticParityMarkdownReport(comparison) {
	const lines = [
		"# OpenClaw GPT-5.4 / Opus 4.6 Agentic Parity Report",
		"",
		`- Compared at: ${comparison.comparedAt}`,
		`- Candidate: ${comparison.candidateLabel}`,
		`- Baseline: ${comparison.baselineLabel}`,
		`- Verdict: ${comparison.pass ? "pass" : "fail"}`,
		"",
		"## Aggregate Metrics",
		"",
		"| Metric | Candidate | Baseline |",
		"| --- | ---: | ---: |",
		`| Completion rate | ${formatPercent(comparison.candidateMetrics.completionRate)} | ${formatPercent(comparison.baselineMetrics.completionRate)} |`,
		`| Unintended-stop rate | ${formatPercent(comparison.candidateMetrics.unintendedStopRate)} | ${formatPercent(comparison.baselineMetrics.unintendedStopRate)} |`,
		`| Valid-tool-call rate | ${formatPercent(comparison.candidateMetrics.validToolCallRate)} | ${formatPercent(comparison.baselineMetrics.validToolCallRate)} |`,
		`| Fake-success count | ${comparison.candidateMetrics.fakeSuccessCount} | ${comparison.baselineMetrics.fakeSuccessCount} |`,
		""
	];
	if (comparison.failures.length > 0) {
		lines.push("## Gate Failures", "");
		for (const failure of comparison.failures) lines.push(`- ${failure}`);
		lines.push("");
	}
	lines.push("## Scenario Comparison", "");
	for (const scenario of comparison.scenarioComparisons) {
		lines.push(`### ${scenario.name}`, "");
		lines.push(`- ${comparison.candidateLabel}: ${scenario.candidateStatus}`);
		lines.push(`- ${comparison.baselineLabel}: ${scenario.baselineStatus}`);
		if (scenario.candidateDetails) lines.push(`- ${comparison.candidateLabel} details: ${scenario.candidateDetails}`);
		if (scenario.baselineDetails) lines.push(`- ${comparison.baselineLabel} details: ${scenario.baselineDetails}`);
		lines.push("");
	}
	lines.push("## Notes", "");
	for (const note of comparison.notes) lines.push(`- ${note}`);
	lines.push("");
	return lines.join("\n");
}
//#endregion
//#region extensions/qa-lab/src/manual-lane.runtime.ts
function resolveManualLaneTimeoutMs(params) {
	if (typeof params.timeoutMs === "number" && Number.isFinite(params.timeoutMs) && params.timeoutMs > 0) return params.timeoutMs;
	return resolveQaLiveTurnTimeoutMs({
		providerMode: params.providerMode,
		primaryModel: params.primaryModel,
		alternateModel: params.alternateModel
	}, 12e4, params.primaryModel);
}
async function runQaManualLane(params) {
	const sessionSuffix = params.primaryModel.replace(/[^a-z0-9._-]+/gi, "-");
	const lab = await startQaLabServer({
		repoRoot: params.repoRoot,
		embeddedGateway: "disabled"
	});
	const mock = params.providerMode === "mock-openai" ? await startQaMockOpenAiServer({
		host: "127.0.0.1",
		port: 0
	}) : null;
	const gateway = await startQaGatewayChild({
		repoRoot: params.repoRoot,
		providerBaseUrl: mock ? `${mock.baseUrl}/v1` : void 0,
		qaBusBaseUrl: lab.listenUrl,
		providerMode: params.providerMode,
		primaryModel: params.primaryModel,
		alternateModel: params.alternateModel,
		fastMode: params.fastMode,
		thinkingDefault: params.thinkingDefault,
		controlUiEnabled: false
	});
	const timeoutMs = resolveManualLaneTimeoutMs({
		providerMode: params.providerMode,
		primaryModel: params.primaryModel,
		alternateModel: params.alternateModel,
		timeoutMs: params.timeoutMs
	});
	try {
		const started = await gateway.call("agent", {
			idempotencyKey: randomUUID(),
			agentId: "qa",
			sessionKey: `agent:qa:manual:${sessionSuffix}`,
			message: params.message,
			deliver: true,
			channel: "qa-channel",
			to: "dm:qa-operator",
			replyChannel: "qa-channel",
			replyTo: "dm:qa-operator"
		}, { timeoutMs: 3e4 });
		if (!started.runId) throw new Error(`agent call did not return a runId: ${JSON.stringify(started)}`);
		const waited = await gateway.call("agent.wait", {
			runId: started.runId,
			timeoutMs
		}, { timeoutMs: timeoutMs + 5e3 });
		await setTimeout$1(500);
		const reply = lab.state.getSnapshot().messages.findLast((candidate) => candidate.direction === "outbound" && candidate.conversation.id === "qa-operator")?.text ?? null;
		return {
			model: params.primaryModel,
			waited,
			reply,
			watchUrl: lab.baseUrl
		};
	} catch (error) {
		throw new Error(formatErrorMessage(error), { cause: error });
	} finally {
		await gateway.stop();
		await mock?.stop();
		await lab.stop();
	}
}
//#endregion
//#region extensions/qa-lab/src/suite-launch.runtime.ts
async function loadQaLabServerRuntime() {
	const { startQaLabServer } = await import("./lab-server-bcPgwyp-.js");
	return startQaLabServer;
}
async function runQaSuiteFromRuntime(...args) {
	const { runQaSuite } = await import("./suite-DnvG1Tyv.js");
	const params = args[0];
	return await runQaSuite({
		...params,
		startLab: params?.startLab ?? await loadQaLabServerRuntime()
	});
}
//#endregion
//#region extensions/qa-lab/src/character-eval.ts
const DEFAULT_CHARACTER_SCENARIO_ID = "character-vibes-gollum";
const DEFAULT_CHARACTER_EVAL_MODELS = Object.freeze([
	"openai/gpt-5.4",
	"openai/gpt-5.2",
	"openai/gpt-5",
	"anthropic/claude-opus-4-6",
	"anthropic/claude-sonnet-4-6",
	"zai/glm-5.1",
	"moonshot/kimi-k2.5",
	"google/gemini-3.1-pro-preview"
]);
const DEFAULT_CHARACTER_THINKING = "high";
const DEFAULT_CHARACTER_EVAL_CONCURRENCY = 16;
const DEFAULT_CHARACTER_THINKING_BY_MODEL = Object.freeze({
	"openai/gpt-5.4": "xhigh",
	"openai/gpt-5.2": "xhigh",
	"openai/gpt-5": "xhigh"
});
const DEFAULT_JUDGE_MODELS = Object.freeze(["openai/gpt-5.4", "anthropic/claude-opus-4-6"]);
const DEFAULT_JUDGE_THINKING = "xhigh";
const DEFAULT_JUDGE_TIMEOUT_MS = 3e5;
const DEFAULT_JUDGE_MODEL_OPTIONS = Object.freeze({
	"openai/gpt-5.4": { thinkingDefault: "xhigh" },
	"anthropic/claude-opus-4-6": { thinkingDefault: "high" }
});
function normalizeModelRefs(models) {
	return [...new Set(models.map((model) => model.trim()).filter((model) => model.length > 0))];
}
function resolveCandidateThinkingDefault(params) {
	return params.candidateModelOptions?.[params.model]?.thinkingDefault ?? params.candidateThinkingByModel?.[params.model] ?? params.candidateThinkingDefault ?? DEFAULT_CHARACTER_THINKING_BY_MODEL[params.model] ?? DEFAULT_CHARACTER_THINKING;
}
function resolveCandidateFastMode(params) {
	return params.candidateModelOptions?.[params.model]?.fastMode ?? params.candidateFastMode ?? isQaFastModeModelRef(params.model);
}
function resolveJudgeOptions(params) {
	const modelDefaults = DEFAULT_JUDGE_MODEL_OPTIONS[params.model];
	const modelOptions = params.judgeModelOptions?.[params.model];
	return {
		thinkingDefault: modelOptions?.thinkingDefault ?? params.judgeThinkingDefault ?? modelDefaults?.thinkingDefault ?? DEFAULT_JUDGE_THINKING,
		fastMode: modelOptions?.fastMode ?? modelDefaults?.fastMode ?? false
	};
}
function sanitizePathPart(value) {
	return value.replace(/[^a-z0-9._-]+/gi, "-").replace(/^-+|-+$/g, "") || "model";
}
function normalizeConcurrency(value, fallback = 1) {
	if (value === void 0) return fallback;
	if (!Number.isFinite(value)) return fallback;
	return Math.max(1, Math.floor(value));
}
async function mapWithConcurrency(items, concurrency, mapper) {
	const results = Array.from({ length: items.length });
	let nextIndex = 0;
	const workerCount = Math.min(normalizeConcurrency(concurrency), items.length);
	const workers = Array.from({ length: workerCount }, async () => {
		while (nextIndex < items.length) {
			const index = nextIndex;
			nextIndex += 1;
			results[index] = await mapper(items[index], index);
		}
	});
	await Promise.all(workers);
	return results;
}
function extractTranscript(result) {
	return result.scenarios.flatMap((scenario) => scenario.steps.map((step) => step.details).filter((detail) => Boolean(detail))).toSorted((left, right) => right.length - left.length)[0] ?? result.report;
}
function collectTranscriptStats(transcript) {
	return {
		transcriptChars: transcript.length,
		transcriptLines: transcript.length === 0 ? 0 : transcript.split(/\r?\n/).length,
		userTurns: transcript.match(/^USER\b/gm)?.length ?? 0,
		assistantTurns: transcript.match(/^ASSISTANT\b/gm)?.length ?? 0
	};
}
function detectTranscriptFailure(transcript) {
	if (extractQaVisibleReplyLeakText(transcript)) return "internal harness/meta text leaked into transcript";
	return [
		[/\bmodel `[^`]+` is not supported\b/i, "model unsupported error leaked into transcript"],
		[/\binsufficient account balance\b/i, "account balance error leaked into transcript"],
		[/\b(?:backend|transport|internal) error\b/i, "backend error leaked into transcript"],
		[/\bsomething went wrong while processing your request\b/i, "generic request failure leaked into transcript"],
		[/\buse \/new to start a fresh session\b/i, "generic request failure leaked into transcript"],
		[/\bmodel did not produce a response before the LLM idle timeout\b/i, "LLM timeout leaked into transcript"],
		[/\btool failed\b/i, "tool failure leaked into transcript"],
		[/\b(?:read|write|edit|patch):[^\n]*\bfailed\b/i, "tool failure leaked into transcript"],
		[/\bnot configured\b/i, "configuration error leaked into transcript"]
	].find(([pattern]) => pattern.test(transcript))?.[1];
}
function formatDuration(ms) {
	if (!Number.isFinite(ms) || ms < 0) return "unknown";
	if (ms < 1e3) return `${Math.round(ms)}ms`;
	if (ms < 6e4) {
		const seconds = ms / 1e3;
		return `${seconds >= 10 ? Math.round(seconds) : Number(seconds.toFixed(1))}s`;
	}
	const totalSeconds = Math.round(ms / 1e3);
	const minutes = Math.floor(totalSeconds / 60);
	const seconds = totalSeconds % 60;
	return seconds === 0 ? `${minutes}m` : `${minutes}m ${seconds}s`;
}
function logCharacterEvalProgress(progress, message) {
	progress?.(`[qa-character] ${message}`);
}
function formatEvalIndex(index, total) {
	return `${index + 1}/${total}`;
}
function summarizeRunStats(run) {
	return [
		`status=${run.status}`,
		`duration=${formatDuration(run.durationMs)}`,
		`turns=${run.stats.userTurns}/${run.stats.assistantTurns}`,
		`chars=${run.stats.transcriptChars}`,
		...run.error ? [`error="${run.error}"`] : []
	].join(" ");
}
function formatBlindCandidateLabel(index) {
	return `candidate-${String(index + 1).padStart(2, "0")}`;
}
function buildJudgePrompt(params) {
	const labelToModel = /* @__PURE__ */ new Map();
	const runBlocks = params.runs.map((run, index) => {
		const label = params.blindModels ? formatBlindCandidateLabel(index) : run.model;
		labelToModel.set(label, run.model);
		return `## CANDIDATE ${label}

Status: ${run.status}
Duration ms (not used for ranking): ${run.durationMs}
Fast mode: ${run.fastMode ? "on" : "off"}
Thinking: ${run.thinkingDefault}
Transcript chars: ${run.stats.transcriptChars}
Assistant turns: ${run.stats.assistantTurns}
Error: ${run.error ?? "none"}

\`\`\`text
${run.transcript}
\`\`\``;
	}).join("\n\n");
	return {
		prompt: `You are grading OpenClaw natural character conversation transcripts for naturalness, vibes, and funniness.

Scenario id: ${params.scenarioId}

Rank the models by:
- natural conversational reaction
- playful character commitment
- funny, surprising details
- coherence across turns
- completing real user tasks without becoming generic
- not sounding aware of an eval or test
- avoiding tool/backend/error leakage

Treat candidate labels as opaque identifiers. Do not assume quality from the label.
Duration is recorded for separate benchmark analysis only. Do not rank models by speed.

Return strict JSON only with this shape:
{
  "rankings": [
    {
      "model": "same candidate label",
      "rank": 1,
      "score": 9.2,
      "summary": "one sentence",
      "strengths": ["short"],
      "weaknesses": ["short"]
    }
  ]
}

${runBlocks}`,
		labelToModel
	};
}
function normalizeJudgment(value, allowedModels) {
	const payload = value && typeof value === "object" ? value : {};
	return (Array.isArray(payload.rankings) ? payload.rankings : []).map((entry) => {
		if (!entry || typeof entry !== "object") return null;
		const record = entry;
		const model = typeof record.model === "string" ? record.model : "";
		if (!allowedModels.has(model)) return null;
		const rank = typeof record.rank === "number" ? record.rank : Number(record.rank);
		const score = typeof record.score === "number" ? record.score : Number(record.score);
		const summary = typeof record.summary === "string" ? record.summary : "";
		const strengths = Array.isArray(record.strengths) ? record.strengths.filter((item) => typeof item === "string") : [];
		const weaknesses = Array.isArray(record.weaknesses) ? record.weaknesses.filter((item) => typeof item === "string") : [];
		if (!Number.isFinite(rank) || !Number.isFinite(score)) return null;
		return {
			model,
			rank,
			score,
			summary,
			strengths,
			weaknesses
		};
	}).filter((entry) => Boolean(entry)).toSorted((left, right) => left.rank - right.rank || right.score - left.score);
}
function parseJudgeReply(reply, allowedModels) {
	if (!reply) throw new Error("judge did not return a reply");
	const trimmed = reply.trim();
	const jsonText = trimmed.match(/```(?:json)?\s*([\s\S]*?)```/)?.[1]?.trim() ?? trimmed.match(/\{[\s\S]*\}/)?.[0]?.trim() ?? trimmed;
	const rankings = normalizeJudgment(JSON.parse(jsonText), allowedModels);
	if (rankings.length === 0) throw new Error("judge reply did not contain valid rankings");
	return rankings;
}
async function defaultRunJudge(params) {
	return (await runQaManualLane({
		repoRoot: params.repoRoot,
		providerMode: "live-frontier",
		primaryModel: params.judgeModel,
		alternateModel: params.judgeModel,
		fastMode: params.judgeFastMode,
		thinkingDefault: params.judgeThinkingDefault,
		message: params.prompt,
		timeoutMs: params.timeoutMs
	})).reply;
}
function renderCharacterEvalReport(params) {
	const lines = [
		"# OpenClaw Character Eval Report",
		"",
		`- Started: ${params.startedAt.toISOString()}`,
		`- Finished: ${params.finishedAt.toISOString()}`,
		`- Duration: ${formatDuration(params.finishedAt.getTime() - params.startedAt.getTime())}`,
		`- Scenario: ${params.scenarioId}`,
		"- Execution: local QA gateway child processes, not Docker",
		`- Judges: ${params.judgments.map((judgment) => judgment.model).join(", ")}`,
		`- Judge thinking: ${params.judgments[0]?.thinkingDefault ?? DEFAULT_JUDGE_THINKING}`,
		`- Judge fast mode: ${params.judgments.every((judgment) => judgment.fastMode) ? "on" : "mixed"}`,
		`- Judge model labels: ${params.judgments.every((judgment) => judgment.blindModels) ? "blind" : "visible"}`,
		"",
		"## Judge Rankings",
		""
	];
	for (const judgment of params.judgments) {
		lines.push(`### ${judgment.model}`, "");
		lines.push(`- Duration: ${formatDuration(judgment.durationMs)}`, "");
		lines.push(`- Timeout: ${formatDuration(judgment.timeoutMs)}`, "");
		if (judgment.rankings.length > 0) for (const ranking of judgment.rankings) {
			lines.push(`${ranking.rank}. ${ranking.model} - ${ranking.score.toFixed(1)} - ${ranking.summary}`);
			if (ranking.strengths.length > 0) lines.push(`   Strengths: ${ranking.strengths.join("; ")}`);
			if (ranking.weaknesses.length > 0) lines.push(`   Weaknesses: ${ranking.weaknesses.join("; ")}`);
		}
		else {
			lines.push("- Judge ranking unavailable.");
			if (judgment.error) lines.push(`- Judge error: ${judgment.error}`);
		}
		lines.push("");
	}
	lines.push("## Run Stats", "");
	lines.push("| Model | Thinking | Fast mode | Status | Duration | User turns | Assistant turns | Transcript chars |");
	lines.push("| --- | --- | --- | --- | ---: | ---: | ---: | ---: |");
	for (const run of params.runs) lines.push(`| ${run.model} | ${run.thinkingDefault} | ${run.fastMode ? "on" : "off"} | ${run.status} | ${formatDuration(run.durationMs)} | ${run.stats.userTurns} | ${run.stats.assistantTurns} | ${run.stats.transcriptChars} |`);
	lines.push("", "## Transcripts", "");
	for (const run of params.runs) {
		lines.push(`### ${run.model}`, "");
		lines.push(`- Status: ${run.status}`);
		lines.push(`- Thinking: ${run.thinkingDefault}`);
		lines.push(`- Fast mode: ${run.fastMode ? "on" : "off"}`);
		lines.push(`- Duration: ${formatDuration(run.durationMs)}`);
		lines.push(`- Report: ${run.reportPath ?? "unavailable"}`);
		if (run.error) lines.push(`- Error: ${run.error}`);
		lines.push("", "```text", run.transcript.trim() || "(empty transcript)", "```", "");
	}
	return `${lines.join("\n")}\n`;
}
async function runQaCharacterEval(params) {
	const startedAt = /* @__PURE__ */ new Date();
	const repoRoot = path.resolve(params.repoRoot ?? process.cwd());
	const scenarioId = params.scenarioId?.trim() || DEFAULT_CHARACTER_SCENARIO_ID;
	const models = normalizeModelRefs(params.models.length > 0 ? params.models : DEFAULT_CHARACTER_EVAL_MODELS);
	if (models.length === 0) throw new Error("qa character-eval needs at least one --model <provider/model> ref");
	const outputDir = params.outputDir ?? path.join(repoRoot, ".artifacts", "qa-e2e", `character-eval-${Date.now().toString(36)}`);
	const runsDir = path.join(outputDir, "runs");
	await fs$1.mkdir(runsDir, { recursive: true });
	const runSuite = params.runSuite ?? runQaSuiteFromRuntime;
	const candidateConcurrency = normalizeConcurrency(params.candidateConcurrency, DEFAULT_CHARACTER_EVAL_CONCURRENCY);
	logCharacterEvalProgress(params.progress, `start scenario=${scenarioId} candidates=${models.length} candidateConcurrency=${candidateConcurrency} output=${outputDir}`);
	const candidatesStartedAt = Date.now();
	const runs = await mapWithConcurrency(models, candidateConcurrency, async (model, index) => {
		const thinkingDefault = resolveCandidateThinkingDefault({
			model,
			candidateThinkingDefault: params.candidateThinkingDefault,
			candidateThinkingByModel: params.candidateThinkingByModel,
			candidateModelOptions: params.candidateModelOptions
		});
		const fastMode = resolveCandidateFastMode({
			model,
			candidateFastMode: params.candidateFastMode,
			candidateModelOptions: params.candidateModelOptions
		});
		const modelOutputDir = path.join(runsDir, sanitizePathPart(model));
		const runStartedAt = Date.now();
		logCharacterEvalProgress(params.progress, `candidate start ${formatEvalIndex(index, models.length)} model=${model} thinking=${thinkingDefault} fast=${fastMode ? "on" : "off"}`);
		try {
			const result = await runSuite({
				repoRoot,
				outputDir: modelOutputDir,
				providerMode: "live-frontier",
				primaryModel: model,
				alternateModel: model,
				fastMode,
				thinkingDefault,
				scenarioIds: [scenarioId]
			});
			const transcript = extractTranscript(result);
			const transcriptFailure = detectTranscriptFailure(transcript);
			const run = {
				model,
				status: result.scenarios.some((scenario) => scenario.status === "fail") || transcriptFailure ? "fail" : "pass",
				durationMs: Date.now() - runStartedAt,
				outputDir: modelOutputDir,
				thinkingDefault,
				fastMode,
				reportPath: result.reportPath,
				summaryPath: result.summaryPath,
				transcript,
				stats: collectTranscriptStats(transcript),
				...transcriptFailure ? { error: transcriptFailure } : {}
			};
			logCharacterEvalProgress(params.progress, `candidate done ${formatEvalIndex(index, models.length)} model=${model} ${summarizeRunStats(run)}`);
			return run;
		} catch (error) {
			const transcript = "";
			const run = {
				model,
				status: "fail",
				durationMs: Date.now() - runStartedAt,
				outputDir: modelOutputDir,
				thinkingDefault,
				fastMode,
				transcript,
				stats: collectTranscriptStats(transcript),
				error: formatErrorMessage(error)
			};
			logCharacterEvalProgress(params.progress, `candidate done ${formatEvalIndex(index, models.length)} model=${model} ${summarizeRunStats(run)}`);
			return run;
		}
	});
	const failedCandidateCount = runs.filter((run) => run.status === "fail").length;
	logCharacterEvalProgress(params.progress, `candidates done pass=${runs.length - failedCandidateCount} fail=${failedCandidateCount} duration=${formatDuration(Date.now() - candidatesStartedAt)}`);
	const judgeModels = normalizeModelRefs(params.judgeModels && params.judgeModels.length > 0 ? params.judgeModels : params.judgeModel ? [params.judgeModel] : DEFAULT_JUDGE_MODELS);
	const runJudge = params.runJudge ?? defaultRunJudge;
	const judgeConcurrency = normalizeConcurrency(params.judgeConcurrency, DEFAULT_CHARACTER_EVAL_CONCURRENCY);
	const judgeTimeoutMs = params.judgeTimeoutMs ?? DEFAULT_JUDGE_TIMEOUT_MS;
	logCharacterEvalProgress(params.progress, `judges start judges=${judgeModels.length} judgeConcurrency=${judgeConcurrency} timeout=${formatDuration(judgeTimeoutMs)} labels=${params.judgeBlindModels === true ? "blind" : "visible"}`);
	const judgesStartedAt = Date.now();
	const judgments = await mapWithConcurrency(judgeModels, judgeConcurrency, async (judgeModel, index) => {
		const judgeOptions = resolveJudgeOptions({
			model: judgeModel,
			judgeThinkingDefault: params.judgeThinkingDefault,
			judgeModelOptions: params.judgeModelOptions
		});
		let rankings = [];
		let judgeError;
		const judgeStartedAt = Date.now();
		logCharacterEvalProgress(params.progress, `judge start ${formatEvalIndex(index, judgeModels.length)} model=${judgeModel} thinking=${judgeOptions.thinkingDefault} fast=${judgeOptions.fastMode ? "on" : "off"} timeout=${formatDuration(judgeTimeoutMs)}`);
		try {
			const judgePrompt = buildJudgePrompt({
				scenarioId,
				runs,
				blindModels: params.judgeBlindModels
			});
			rankings = parseJudgeReply(await runJudge({
				repoRoot,
				judgeModel,
				judgeThinkingDefault: judgeOptions.thinkingDefault,
				judgeFastMode: judgeOptions.fastMode,
				prompt: judgePrompt.prompt,
				timeoutMs: judgeTimeoutMs
			}), new Set(judgePrompt.labelToModel.keys())).map((ranking) => ({
				...ranking,
				model: judgePrompt.labelToModel.get(ranking.model) ?? ranking.model
			}));
		} catch (error) {
			judgeError = formatErrorMessage(error);
		}
		const judgment = {
			model: judgeModel,
			thinkingDefault: judgeOptions.thinkingDefault,
			fastMode: judgeOptions.fastMode,
			blindModels: params.judgeBlindModels === true,
			timeoutMs: judgeTimeoutMs,
			durationMs: Date.now() - judgeStartedAt,
			rankings,
			...judgeError ? { error: judgeError } : {}
		};
		logCharacterEvalProgress(params.progress, `judge done ${formatEvalIndex(index, judgeModels.length)} model=${judgeModel} rankings=${rankings.length} duration=${formatDuration(judgment.durationMs)}${judgeError ? ` error="${judgeError}"` : ""}`);
		return judgment;
	});
	const failedJudgeCount = judgments.filter((judgment) => judgment.rankings.length === 0).length;
	logCharacterEvalProgress(params.progress, `judges done ranked=${judgments.length - failedJudgeCount} failed=${failedJudgeCount} duration=${formatDuration(Date.now() - judgesStartedAt)}`);
	const finishedAt = /* @__PURE__ */ new Date();
	const report = renderCharacterEvalReport({
		scenarioId,
		startedAt,
		finishedAt,
		runs,
		judgments
	});
	const reportPath = path.join(outputDir, "character-eval-report.md");
	const summaryPath = path.join(outputDir, "character-eval-summary.json");
	await fs$1.writeFile(reportPath, report, "utf8");
	await fs$1.writeFile(summaryPath, `${JSON.stringify({
		scenarioId,
		runs,
		judgments
	}, null, 2)}\n`, "utf8");
	logCharacterEvalProgress(params.progress, `report written duration=${formatDuration(finishedAt.getTime() - startedAt.getTime())} report=${reportPath} summary=${summaryPath}`);
	return {
		outputDir,
		reportPath,
		summaryPath,
		runs,
		judgments
	};
}
//#endregion
//#region extensions/qa-lab/src/docker-up.runtime.ts
function resolveDefaultQaDockerDir(repoRoot) {
	return path.resolve(repoRoot, ".artifacts/qa-docker");
}
async function runQaDockerUp(params, deps) {
	const repoRoot = path.resolve(params.repoRoot ?? process.cwd());
	const resolveHostPortImpl = deps?.resolveHostPortImpl ?? resolveHostPort;
	const outputDir = path.resolve(params.outputDir ?? resolveDefaultQaDockerDir(repoRoot));
	const gatewayPort = await resolveHostPortImpl(params.gatewayPort ?? 18789, params.gatewayPort != null);
	const qaLabPort = await resolveHostPortImpl(params.qaLabPort ?? 43124, params.qaLabPort != null);
	const runCommand = deps?.runCommand ?? execCommand;
	const fetchImpl = deps?.fetchImpl ?? fetchHealthUrl;
	const sleepImpl = deps?.sleepImpl ?? setTimeout$1;
	if (!params.skipUiBuild) await runCommand("pnpm", ["qa:lab:build"], repoRoot);
	await writeQaDockerHarnessFiles({
		outputDir,
		repoRoot,
		gatewayPort,
		qaLabPort,
		providerBaseUrl: params.providerBaseUrl,
		imageName: params.image,
		usePrebuiltImage: params.usePrebuiltImage,
		bindUiDist: params.bindUiDist,
		includeQaLabUi: true
	});
	const composeFile = path.join(outputDir, "docker-compose.qa.yml");
	try {
		await runCommand("docker", [
			"compose",
			"-f",
			composeFile,
			"down",
			"--remove-orphans"
		], repoRoot);
	} catch {}
	const composeArgs = [
		"compose",
		"-f",
		composeFile,
		"up"
	];
	if (!params.usePrebuiltImage) composeArgs.push("--build");
	composeArgs.push("-d");
	await runCommand("docker", composeArgs, repoRoot);
	await sleepImpl(3e3);
	const qaLabUrl = `http://127.0.0.1:${qaLabPort}`;
	const hostGatewayUrl = `http://127.0.0.1:${gatewayPort}/`;
	await waitForHealth(`${qaLabUrl}/healthz`, {
		label: "QA Lab",
		fetchImpl,
		sleepImpl,
		composeFile
	});
	await waitForDockerServiceHealth("openclaw-qa-gateway", composeFile, repoRoot, runCommand, sleepImpl);
	let gatewayUrl = hostGatewayUrl;
	if (!await fetchImpl(`${hostGatewayUrl}healthz`).then((response) => response.ok).catch(() => false)) {
		const containerGatewayUrl = await resolveComposeServiceUrl("openclaw-qa-gateway", 18789, composeFile, repoRoot, runCommand, fetchImpl);
		if (containerGatewayUrl) gatewayUrl = containerGatewayUrl;
	}
	return {
		outputDir,
		composeFile,
		qaLabUrl,
		gatewayUrl,
		stopCommand: `docker compose -f ${composeFile} down`
	};
}
//#endregion
//#region extensions/qa-lab/src/multipass.runtime.ts
const MULTIPASS_MOUNTED_REPO_PATH = "/workspace/openclaw-host";
const MULTIPASS_GUEST_REPO_PATH = "/workspace/openclaw";
const MULTIPASS_GUEST_CODEX_HOME_PATH = "/workspace/openclaw-codex-home";
const MULTIPASS_GUEST_PACKAGES = [
	"build-essential",
	"ca-certificates",
	"curl",
	"pkg-config",
	"python3",
	"rsync",
	"xz-utils"
];
const MULTIPASS_REPO_SYNC_EXCLUDES = [
	".git",
	"node_modules",
	".artifacts",
	".tmp",
	".turbo",
	"coverage",
	"*.heapsnapshot"
];
const MULTIPASS_EXEC_MAX_BUFFER = 64 * 1024 * 1024;
const MULTIPASS_GUEST_RUN_TIMEOUT_MS = 3600 * 1e3;
const QA_LIVE_ENV_ALIASES = Object.freeze([
	{
		liveVar: "OPENCLAW_LIVE_OPENAI_KEY",
		providerVar: "OPENAI_API_KEY"
	},
	{
		liveVar: "OPENCLAW_LIVE_ANTHROPIC_KEY",
		providerVar: "ANTHROPIC_API_KEY"
	},
	{
		liveVar: "OPENCLAW_LIVE_GEMINI_KEY",
		providerVar: "GEMINI_API_KEY"
	}
]);
const QA_LIVE_ALLOWED_ENV_VARS = Object.freeze([
	"ANTHROPIC_API_KEY",
	"ANTHROPIC_OAUTH_TOKEN",
	"AWS_ACCESS_KEY_ID",
	"AWS_BEARER_TOKEN_BEDROCK",
	"AWS_REGION",
	"AWS_SECRET_ACCESS_KEY",
	"AWS_SESSION_TOKEN",
	"GEMINI_API_KEY",
	"GEMINI_API_KEYS",
	"GOOGLE_API_KEY",
	"MISTRAL_API_KEY",
	"OPENAI_API_KEY",
	"OPENAI_API_KEYS",
	"OPENAI_BASE_URL",
	"OPENCLAW_LIVE_ANTHROPIC_KEY",
	"OPENCLAW_LIVE_ANTHROPIC_KEYS",
	"OPENCLAW_LIVE_GEMINI_KEY",
	"OPENCLAW_LIVE_OPENAI_KEY",
	"OPENCLAW_QA_LIVE_PROVIDER_CONFIG_PATH",
	"OPENCLAW_CONFIG_PATH",
	"VOYAGE_API_KEY"
]);
const QA_LIVE_ALLOWED_ENV_PATTERNS = Object.freeze([
	/^[A-Z0-9_]+_API_KEYS$/u,
	/^[A-Z0-9_]+_API_KEY_[0-9]+$/u,
	/^OPENCLAW_LIVE_[A-Z0-9_]+_KEYS$/u
]);
const qaMultipassDefaultResources = {
	image: "lts",
	cpus: 2,
	memory: "4G",
	disk: "24G"
};
function shellQuote(value) {
	return `'${value.replaceAll("'", `'"'"'`)}'`;
}
function createOutputStamp() {
	return (/* @__PURE__ */ new Date()).toISOString().replaceAll(":", "").replaceAll(".", "").replace("T", "-");
}
function createVmSuffix() {
	return `${Date.now().toString(36)}-${randomUUID().slice(0, 8)}`;
}
function sleep(ms) {
	return new Promise((resolve) => setTimeout(resolve, ms));
}
function execFileAsync(file, args, options = {}) {
	return new Promise((resolve, reject) => {
		execFile(file, args, {
			encoding: "utf8",
			maxBuffer: MULTIPASS_EXEC_MAX_BUFFER,
			timeout: options.timeoutMs
		}, (error, stdout, stderr) => {
			if (error) {
				const message = stderr.trim() || stdout.trim() || error.message;
				const wrappedError = new Error(message, { cause: error });
				wrappedError.code = error.code;
				reject(wrappedError);
				return;
			}
			resolve({
				stdout,
				stderr
			});
		});
	});
}
function resolveRealPath(value) {
	return fs.realpathSync.native?.(value) ?? fs.realpathSync(value);
}
function resolveExistingPath(value) {
	let currentPath = value;
	while (!fs.existsSync(currentPath)) {
		const parentPath = path.dirname(currentPath);
		if (parentPath === currentPath) throw new Error(`unable to resolve existing path for ${value}`);
		currentPath = parentPath;
	}
	return currentPath;
}
function isPathInside(parentPath, childPath) {
	const relativePath = path.relative(parentPath, childPath);
	return !relativePath.startsWith("..") && !path.isAbsolute(relativePath);
}
function validatePnpmVersion(version) {
	if (!/^[0-9A-Za-z.+_-]+$/u.test(version)) throw new Error(`unsupported pnpm version in packageManager: ${version}`);
	return version;
}
function resolveMountedOutputPath(repoRoot, hostPath) {
	const relativePath = path.relative(repoRoot, hostPath);
	if (relativePath.startsWith("..") || path.isAbsolute(relativePath) || relativePath.length === 0) throw new Error(`qa suite --runner multipass requires --output-dir to stay under the repo root (${repoRoot}), got ${hostPath}.`);
	const realRepoRoot = resolveRealPath(repoRoot);
	const realExistingHostPath = resolveRealPath(resolveExistingPath(hostPath));
	if (!isPathInside(realRepoRoot, realExistingHostPath) && realExistingHostPath !== realRepoRoot) throw new Error(`qa suite --runner multipass requires --output-dir to stay under the repo root (${repoRoot}), got ${hostPath}.`);
	return path.posix.join(MULTIPASS_MOUNTED_REPO_PATH, ...relativePath.split(path.sep));
}
function resolvePnpmVersion(repoRoot) {
	const packageJsonPath = path.join(repoRoot, "package.json");
	const packageManager = JSON.parse(fs.readFileSync(packageJsonPath, "utf8")).packageManager ?? "";
	const match = /^pnpm@(.+)$/.exec(packageManager);
	if (!match?.[1]) throw new Error(`unable to resolve pnpm version from packageManager in ${packageJsonPath}`);
	return match[1];
}
function resolveMultipassInstallHint() {
	if (process.platform === "darwin") return "brew install --cask multipass";
	if (process.platform === "win32") return "winget install Canonical.Multipass";
	if (process.platform === "linux") return "sudo snap install multipass";
	return "https://multipass.run/install";
}
function resolveUserPath(value, env = process.env) {
	if (value === "~") return env.HOME ?? os.homedir();
	if (value.startsWith("~/")) return path.join(env.HOME ?? os.homedir(), value.slice(2));
	return path.resolve(value);
}
function resolveLiveProviderConfigPath(env = process.env) {
	const explicit = env.OPENCLAW_QA_LIVE_PROVIDER_CONFIG_PATH?.trim() || env.OPENCLAW_CONFIG_PATH?.trim();
	return explicit ? {
		path: resolveUserPath(explicit, env),
		explicit: true
	} : {
		path: path.join(os.homedir(), ".openclaw", "openclaw.json"),
		explicit: false
	};
}
function resolveQaLiveCliAuthEnv(baseEnv) {
	const configuredCodexHome = baseEnv.CODEX_HOME?.trim();
	if (configuredCodexHome) {
		const codexHome = resolveUserPath(configuredCodexHome, baseEnv);
		return fs.existsSync(codexHome) ? { CODEX_HOME: codexHome } : {};
	}
	const hostHome = baseEnv.HOME?.trim() || os.homedir();
	const codexHome = path.join(hostHome, ".codex");
	return fs.existsSync(codexHome) ? { CODEX_HOME: codexHome } : {};
}
function resolveForwardedLiveEnv(baseEnv = process.env) {
	const forwarded = {};
	for (const [key, rawValue] of Object.entries(baseEnv)) {
		if (!QA_LIVE_ALLOWED_ENV_VARS.includes(key) && !QA_LIVE_ALLOWED_ENV_PATTERNS.some((pattern) => pattern.test(key))) continue;
		const value = rawValue?.trim();
		if (value) forwarded[key] = value;
	}
	for (const { liveVar, providerVar } of QA_LIVE_ENV_ALIASES) {
		const liveValue = forwarded[liveVar]?.trim();
		if (liveValue && !forwarded[providerVar]?.trim()) forwarded[providerVar] = liveValue;
	}
	const liveCliAuth = resolveQaLiveCliAuthEnv(baseEnv);
	if (liveCliAuth.CODEX_HOME) forwarded.CODEX_HOME = liveCliAuth.CODEX_HOME;
	return forwarded;
}
function createQaMultipassOutputDir(repoRoot) {
	return path.join(repoRoot, ".artifacts", "qa-e2e", `multipass-${createOutputStamp()}`);
}
function resolveGuestMountedPath(repoRoot, hostPath) {
	return resolveMountedOutputPath(repoRoot, hostPath);
}
function appendScenarioArgs(command, scenarioIds) {
	for (const scenarioId of scenarioIds) command.push("--scenario", scenarioId);
	return command;
}
function createQaMultipassPlan(params) {
	const outputDir = params.outputDir ?? createQaMultipassOutputDir(params.repoRoot);
	const scenarioIds = [...new Set(params.scenarioIds ?? [])];
	const providerMode = params.providerMode ?? "mock-openai";
	const forwardedEnv = providerMode === "live-frontier" ? resolveForwardedLiveEnv() : {};
	const hostCodexHomePath = forwardedEnv.CODEX_HOME;
	const liveProviderConfig = providerMode === "live-frontier" ? resolveLiveProviderConfigPath() : void 0;
	const hostLiveProviderConfigPath = liveProviderConfig && fs.existsSync(liveProviderConfig.path) ? liveProviderConfig.path : void 0;
	const vmName = `openclaw-qa-${createVmSuffix()}`;
	const guestOutputDir = resolveGuestMountedPath(params.repoRoot, outputDir);
	const qaCommand = appendScenarioArgs([
		"pnpm",
		"openclaw",
		"qa",
		"suite",
		"--provider-mode",
		providerMode,
		"--output-dir",
		guestOutputDir,
		...params.primaryModel ? ["--model", params.primaryModel] : [],
		...params.alternateModel ? ["--alt-model", params.alternateModel] : [],
		...params.fastMode ? ["--fast"] : [],
		...params.concurrency ? ["--concurrency", String(params.concurrency)] : []
	], scenarioIds);
	return {
		repoRoot: params.repoRoot,
		outputDir,
		reportPath: path.join(outputDir, "qa-suite-report.md"),
		summaryPath: path.join(outputDir, "qa-suite-summary.json"),
		hostLogPath: path.join(outputDir, "multipass-host.log"),
		hostBootstrapLogPath: path.join(outputDir, "multipass-guest-bootstrap.log"),
		hostGuestScriptPath: path.join(outputDir, "multipass-guest-run.sh"),
		vmName,
		image: params.image ?? qaMultipassDefaultResources.image,
		cpus: params.cpus ?? qaMultipassDefaultResources.cpus,
		memory: params.memory ?? qaMultipassDefaultResources.memory,
		disk: params.disk ?? qaMultipassDefaultResources.disk,
		pnpmVersion: validatePnpmVersion(resolvePnpmVersion(params.repoRoot)),
		providerMode,
		primaryModel: params.primaryModel,
		alternateModel: params.alternateModel,
		fastMode: params.fastMode,
		scenarioIds,
		forwardedEnv,
		hostCodexHomePath,
		guestCodexHomePath: hostCodexHomePath ? MULTIPASS_GUEST_CODEX_HOME_PATH : void 0,
		hostLiveProviderConfigPath,
		guestLiveProviderConfigPath: hostLiveProviderConfigPath ? `/tmp/${vmName}-live-provider-config.json` : void 0,
		guestMountedRepoPath: MULTIPASS_MOUNTED_REPO_PATH,
		guestRepoPath: MULTIPASS_GUEST_REPO_PATH,
		guestOutputDir,
		guestScriptPath: `/tmp/${vmName}-qa-suite.sh`,
		guestBootstrapLogPath: `/tmp/${vmName}-bootstrap.log`,
		qaCommand
	};
}
function renderQaMultipassGuestScript(plan, options = {}) {
	const redactSecrets = options.redactSecrets ?? false;
	const rsyncCommand = [
		"rsync -a --delete",
		...MULTIPASS_REPO_SYNC_EXCLUDES.flatMap((value) => ["--exclude", shellQuote(value)]),
		shellQuote(`${plan.guestMountedRepoPath}/`),
		shellQuote(`${plan.guestRepoPath}/`)
	].join(" ");
	const qaCommand = [
		...Object.entries(plan.forwardedEnv).filter(([key]) => key !== "CODEX_HOME" && key !== "OPENCLAW_CONFIG_PATH" && key !== "OPENCLAW_QA_LIVE_PROVIDER_CONFIG_PATH").map(([key, value]) => `${key}=${shellQuote(redactSecrets ? "<redacted>" : value)}`),
		...plan.guestCodexHomePath ? [`CODEX_HOME=${shellQuote(plan.guestCodexHomePath)}`] : [],
		...plan.guestLiveProviderConfigPath ? [`OPENCLAW_CONFIG_PATH=${shellQuote(plan.guestLiveProviderConfigPath)}`, `OPENCLAW_QA_LIVE_PROVIDER_CONFIG_PATH=${shellQuote(plan.guestLiveProviderConfigPath)}`] : [],
		plan.qaCommand.map(shellQuote).join(" ")
	].join(" ");
	return [
		"#!/usr/bin/env bash",
		"set -euo pipefail",
		"trap 'status=$?; echo \"guest failure (exit ${status})\" >&2; exit ${status}' ERR",
		"",
		"export DEBIAN_FRONTEND=noninteractive",
		`BOOTSTRAP_LOG=${shellQuote(plan.guestBootstrapLogPath)}`,
		": > \"$BOOTSTRAP_LOG\"",
		"",
		"ensure_guest_packages() {",
		"  sudo -E apt-get update >>\"$BOOTSTRAP_LOG\" 2>&1",
		"  sudo -E apt-get install -y \\",
		...MULTIPASS_GUEST_PACKAGES.map((value, index) => index === MULTIPASS_GUEST_PACKAGES.length - 1 ? `    ${value} >>"$BOOTSTRAP_LOG" 2>&1` : `    ${value} \\`),
		"}",
		"",
		"ensure_node() {",
		"  if command -v node >/dev/null; then",
		"    local node_major",
		"    node_major=\"$(node -p 'process.versions.node.split(\".\")[0]' 2>/dev/null || echo 0)\"",
		"    if [ \"${node_major}\" -ge 22 ]; then",
		"      return 0",
		"    fi",
		"  fi",
		"  local node_arch",
		"  case \"$(uname -m)\" in",
		"    x86_64) node_arch=\"x64\" ;;",
		"    aarch64|arm64) node_arch=\"arm64\" ;;",
		"    *) echo \"unsupported guest architecture for node bootstrap: $(uname -m)\" >&2; return 1 ;;",
		"  esac",
		"  local node_tmp_dir tarball_name extract_dir base_url",
		"  node_tmp_dir=\"$(mktemp -d)\"",
		"  trap 'rm -rf \"${node_tmp_dir}\"' RETURN",
		"  base_url=\"https://nodejs.org/dist/latest-v22.x\"",
		"  curl -fsSL \"${base_url}/SHASUMS256.txt\" -o \"${node_tmp_dir}/SHASUMS256.txt\" >>\"$BOOTSTRAP_LOG\" 2>&1",
		"  tarball_name=\"$(awk '/linux-'\"${node_arch}\"'\\.tar\\.xz$/ { print $2; exit }' \"${node_tmp_dir}/SHASUMS256.txt\")\"",
		"  [ -n \"${tarball_name}\" ] || { echo \"unable to resolve node tarball for ${node_arch}\" >&2; return 1; }",
		"  curl -fsSL \"${base_url}/${tarball_name}\" -o \"${node_tmp_dir}/${tarball_name}\" >>\"$BOOTSTRAP_LOG\" 2>&1",
		"  (cd \"${node_tmp_dir}\" && grep \" ${tarball_name}$\" SHASUMS256.txt | sha256sum -c -) >>\"$BOOTSTRAP_LOG\" 2>&1",
		"  extract_dir=\"${tarball_name%.tar.xz}\"",
		"  sudo mkdir -p /usr/local/lib/nodejs >>\"$BOOTSTRAP_LOG\" 2>&1",
		"  sudo rm -rf \"/usr/local/lib/nodejs/${extract_dir}\" >>\"$BOOTSTRAP_LOG\" 2>&1",
		"  sudo tar -xJf \"${node_tmp_dir}/${tarball_name}\" -C /usr/local/lib/nodejs >>\"$BOOTSTRAP_LOG\" 2>&1",
		"  sudo ln -sf \"/usr/local/lib/nodejs/${extract_dir}/bin/node\" /usr/local/bin/node >>\"$BOOTSTRAP_LOG\" 2>&1",
		"  sudo ln -sf \"/usr/local/lib/nodejs/${extract_dir}/bin/npm\" /usr/local/bin/npm >>\"$BOOTSTRAP_LOG\" 2>&1",
		"  sudo ln -sf \"/usr/local/lib/nodejs/${extract_dir}/bin/npx\" /usr/local/bin/npx >>\"$BOOTSTRAP_LOG\" 2>&1",
		"  sudo ln -sf \"/usr/local/lib/nodejs/${extract_dir}/bin/corepack\" /usr/local/bin/corepack >>\"$BOOTSTRAP_LOG\" 2>&1",
		"}",
		"",
		"ensure_pnpm() {",
		"  sudo env PATH=\"/usr/local/bin:/usr/bin:/bin\" corepack enable >>\"$BOOTSTRAP_LOG\" 2>&1",
		`  sudo env PATH="/usr/local/bin:/usr/bin:/bin" corepack prepare ${shellQuote(`pnpm@${plan.pnpmVersion}`)} --activate >>"$BOOTSTRAP_LOG" 2>&1`,
		"}",
		"",
		"command -v sudo >/dev/null || { echo \"missing sudo in guest\" >&2; exit 1; }",
		"ensure_guest_packages",
		"ensure_node",
		"ensure_pnpm",
		"command -v node >/dev/null || { echo \"missing node after guest bootstrap\" >&2; exit 1; }",
		"command -v pnpm >/dev/null || { echo \"missing pnpm after guest bootstrap\" >&2; exit 1; }",
		"command -v rsync >/dev/null || { echo \"missing rsync after guest bootstrap\" >&2; exit 1; }",
		"",
		`mkdir -p ${shellQuote(path.posix.dirname(plan.guestRepoPath))}`,
		`rm -rf ${shellQuote(plan.guestRepoPath)}`,
		`mkdir -p ${shellQuote(plan.guestRepoPath)}`,
		`mkdir -p ${shellQuote(plan.guestOutputDir)}`,
		rsyncCommand,
		`cd ${shellQuote(plan.guestRepoPath)}`,
		"pnpm install --frozen-lockfile >>\"$BOOTSTRAP_LOG\" 2>&1",
		"pnpm build >>\"$BOOTSTRAP_LOG\" 2>&1",
		qaCommand,
		""
	].join("\n");
}
async function appendMultipassLog(logPath, message) {
	await appendFile(logPath, message, "utf8");
}
async function runMultipassCommand(logPath, args, options = {}) {
	await appendMultipassLog(logPath, `$ ${["multipass", ...args].join(" ")}\n`);
	const result = await execFileAsync("multipass", args, options);
	if (result.stdout.trim()) await appendMultipassLog(logPath, `${result.stdout.trim()}\n`);
	if (result.stderr.trim()) await appendMultipassLog(logPath, `${result.stderr.trim()}\n`);
	await appendMultipassLog(logPath, "\n");
	return result;
}
async function waitForGuestReady(logPath, vmName) {
	let lastError;
	for (let attempt = 1; attempt <= 12; attempt += 1) try {
		await runMultipassCommand(logPath, [
			"exec",
			vmName,
			"--",
			"bash",
			"-lc",
			"echo guest-ready"
		]);
		return;
	} catch (error) {
		lastError = error;
		await appendMultipassLog(logPath, `guest-ready retry ${attempt}/12: ${error instanceof Error ? error.message : String(error)}\n\n`);
		if (attempt < 12) await sleep(2e3);
	}
	throw lastError instanceof Error ? lastError : new Error(String(lastError));
}
async function mountRepo(logPath, repoRoot, vmName) {
	let lastError;
	for (let attempt = 1; attempt <= 5; attempt += 1) try {
		await runMultipassCommand(logPath, [
			"mount",
			repoRoot,
			`${vmName}:${MULTIPASS_MOUNTED_REPO_PATH}`
		]);
		return;
	} catch (error) {
		lastError = error;
		await appendMultipassLog(logPath, `mount retry ${attempt}/5: ${error instanceof Error ? error.message : String(error)}\n\n`);
		if (attempt < 5) await sleep(2e3);
	}
	throw lastError instanceof Error ? lastError : new Error(String(lastError));
}
async function mountCodexHome(logPath, hostCodexHomePath, vmName) {
	let lastError;
	for (let attempt = 1; attempt <= 5; attempt += 1) try {
		await runMultipassCommand(logPath, [
			"mount",
			hostCodexHomePath,
			`${vmName}:${MULTIPASS_GUEST_CODEX_HOME_PATH}`
		]);
		return;
	} catch (error) {
		lastError = error;
		await appendMultipassLog(logPath, `codex-home mount retry ${attempt}/5: ${error instanceof Error ? error.message : String(error)}\n\n`);
		if (attempt < 5) await sleep(2e3);
	}
	throw lastError instanceof Error ? lastError : new Error(String(lastError));
}
async function transferLiveProviderConfig(plan) {
	if (!plan.hostLiveProviderConfigPath || !plan.guestLiveProviderConfigPath) return;
	await runMultipassCommand(plan.hostLogPath, [
		"transfer",
		plan.hostLiveProviderConfigPath,
		`${plan.vmName}:${plan.guestLiveProviderConfigPath}`
	]);
}
async function tryCopyGuestBootstrapLog(plan) {
	try {
		await runMultipassCommand(plan.hostLogPath, [
			"transfer",
			`${plan.vmName}:${plan.guestBootstrapLogPath}`,
			plan.hostBootstrapLogPath
		]);
	} catch (error) {
		await appendMultipassLog(plan.hostLogPath, `bootstrap log transfer skipped: ${error instanceof Error ? error.message : String(error)}\n\n`);
	}
}
async function runQaMultipass(params) {
	const plan = createQaMultipassPlan(params);
	await mkdir(plan.outputDir, { recursive: true });
	await writeFile(plan.hostLogPath, `# OpenClaw QA Multipass host log\nvmName=${plan.vmName}\noutputDir=${plan.outputDir}\n\n`, "utf8");
	await writeFile(plan.hostGuestScriptPath, renderQaMultipassGuestScript(plan, { redactSecrets: true }), {
		encoding: "utf8",
		mode: 384
	});
	try {
		await execFileAsync("multipass", ["version"]);
	} catch (error) {
		if (error.code !== "ENOENT") throw new Error(`Unable to verify Multipass availability: ${error instanceof Error ? error.message : String(error)}.`, { cause: error });
		throw new Error(`Multipass is not installed on this host. Install it with '${resolveMultipassInstallHint()}', then rerun 'pnpm openclaw qa suite --runner multipass'.`, { cause: error });
	}
	const hostTransferDirPath = await fs.promises.mkdtemp(path.join(resolvePreferredOpenClawTmpDir(), `${plan.vmName}-qa-suite-`));
	const hostTransferScriptPath = path.join(hostTransferDirPath, "guest-run.sh");
	await writeFile(hostTransferScriptPath, renderQaMultipassGuestScript(plan), {
		encoding: "utf8",
		mode: 384
	});
	let launched = false;
	try {
		await runMultipassCommand(plan.hostLogPath, [
			"launch",
			"--name",
			plan.vmName,
			"--cpus",
			String(plan.cpus),
			"--memory",
			plan.memory,
			"--disk",
			plan.disk,
			plan.image
		]);
		launched = true;
		await waitForGuestReady(plan.hostLogPath, plan.vmName);
		await mountRepo(plan.hostLogPath, plan.repoRoot, plan.vmName);
		if (plan.hostCodexHomePath) await mountCodexHome(plan.hostLogPath, plan.hostCodexHomePath, plan.vmName);
		await transferLiveProviderConfig(plan);
		await runMultipassCommand(plan.hostLogPath, [
			"transfer",
			hostTransferScriptPath,
			`${plan.vmName}:${plan.guestScriptPath}`
		]);
		await runMultipassCommand(plan.hostLogPath, [
			"exec",
			plan.vmName,
			"--",
			"chmod",
			"+x",
			plan.guestScriptPath
		]);
		await runMultipassCommand(plan.hostLogPath, [
			"exec",
			plan.vmName,
			"--",
			plan.guestScriptPath
		], { timeoutMs: MULTIPASS_GUEST_RUN_TIMEOUT_MS });
		await tryCopyGuestBootstrapLog(plan);
	} catch (error) {
		if (launched) await tryCopyGuestBootstrapLog(plan);
		throw new Error(`QA Multipass run failed: ${error instanceof Error ? error.message : String(error)}. See ${plan.hostLogPath}.`, { cause: error });
	} finally {
		await fs.promises.rm(hostTransferDirPath, {
			recursive: true,
			force: true
		});
		if (launched) try {
			await runMultipassCommand(plan.hostLogPath, [
				"delete",
				"--purge",
				plan.vmName
			]);
		} catch (error) {
			await appendMultipassLog(plan.hostLogPath, `cleanup error: ${error instanceof Error ? error.message : String(error)}\n\n`);
		}
	}
	await access(plan.reportPath);
	await access(plan.summaryPath);
	return {
		outputDir: plan.outputDir,
		reportPath: plan.reportPath,
		summaryPath: plan.summaryPath,
		hostLogPath: plan.hostLogPath,
		bootstrapLogPath: plan.hostBootstrapLogPath,
		guestScriptPath: plan.hostGuestScriptPath,
		vmName: plan.vmName,
		scenarioIds: plan.scenarioIds
	};
}
//#endregion
//#region extensions/qa-lab/src/cli.runtime.ts
function resolveQaManualLaneModels(opts) {
	const primaryModel = opts.primaryModel?.trim() || defaultQaModelForMode(opts.providerMode);
	const alternateModel = opts.alternateModel?.trim();
	return {
		primaryModel,
		alternateModel: alternateModel && alternateModel.length > 0 ? alternateModel : opts.primaryModel?.trim() ? primaryModel : defaultQaModelForMode(opts.providerMode, true)
	};
}
function parseQaThinkingLevel(label, value) {
	if (value === void 0) return;
	const normalized = normalizeQaThinkingLevel(value);
	if (!normalized) throw new Error(`${label} must be one of off, minimal, low, medium, high, xhigh, adaptive`);
	return normalized;
}
function parseQaModelThinkingOverrides(entries) {
	const overrides = {};
	for (const entry of entries ?? []) {
		const separatorIndex = entry.lastIndexOf("=");
		if (separatorIndex <= 0 || separatorIndex === entry.length - 1) throw new Error(`--model-thinking must use provider/model=level, got "${entry}"`);
		const model = entry.slice(0, separatorIndex).trim();
		const level = parseQaThinkingLevel("--model-thinking", entry.slice(separatorIndex + 1).trim());
		if (!model || !level) throw new Error(`--model-thinking must use provider/model=level, got "${entry}"`);
		overrides[model] = level;
	}
	return Object.keys(overrides).length > 0 ? overrides : void 0;
}
function parseQaBooleanModelOption(label, value) {
	switch (value.trim().toLowerCase()) {
		case "1":
		case "on":
		case "true":
		case "yes": return true;
		case "0":
		case "false":
		case "no":
		case "off": return false;
		default: throw new Error(`${label} fast must be one of true, false, on, off, yes, no, 1, 0`);
	}
}
function parseQaPositiveIntegerOption(label, value) {
	if (value === void 0) return;
	if (!Number.isFinite(value) || value < 1) throw new Error(`${label} must be a positive integer`);
	return Math.floor(value);
}
function parseQaCliBackendAuthMode(value) {
	const normalized = value?.trim().toLowerCase();
	if (!normalized) return;
	if (normalized === "auto" || normalized === "api-key" || normalized === "subscription") return normalized;
	throw new Error("--cli-auth-mode must be one of auto, api-key, subscription");
}
function parseQaModelSpecs(label, entries) {
	const models = [];
	const optionsByModel = {};
	for (const entry of entries ?? []) {
		const parts = entry.split(",").map((part) => part.trim());
		const model = parts[0];
		if (!model) throw new Error(`${label} must start with provider/model, got "${entry}"`);
		models.push(model);
		const options = {};
		for (const part of parts.slice(1)) {
			if (!part) throw new Error(`${label} option cannot be empty in "${entry}"`);
			if (part === "fast") {
				options.fastMode = true;
				continue;
			}
			if (part === "no-fast") {
				options.fastMode = false;
				continue;
			}
			const separatorIndex = part.indexOf("=");
			if (separatorIndex <= 0 || separatorIndex === part.length - 1) throw new Error(`${label} options must be thinking=<level>, fast, no-fast, or fast=<boolean>, got "${part}"`);
			const key = part.slice(0, separatorIndex).trim();
			const value = part.slice(separatorIndex + 1).trim();
			switch (key) {
				case "thinking": {
					const thinkingDefault = parseQaThinkingLevel(`${label} thinking`, value);
					if (!thinkingDefault) throw new Error(`${label} thinking must be one of off, minimal, low, medium, high, xhigh, adaptive`);
					options.thinkingDefault = thinkingDefault;
					break;
				}
				case "fast":
					options.fastMode = parseQaBooleanModelOption(label, value);
					break;
				default: throw new Error(`${label} does not support option "${key}" in "${entry}"`);
			}
		}
		if (Object.keys(options).length > 0) optionsByModel[model] = {
			...optionsByModel[model],
			...options
		};
	}
	return {
		models,
		optionsByModel: Object.keys(optionsByModel).length > 0 ? optionsByModel : void 0
	};
}
async function runInterruptibleServer(label, server) {
	process.stdout.write(`${label}: ${server.baseUrl}\n`);
	process.stdout.write("Press Ctrl+C to stop.\n");
	const shutdown = async () => {
		process.off("SIGINT", onSignal);
		process.off("SIGTERM", onSignal);
		await server.stop();
		process.exit(0);
	};
	const onSignal = () => {
		shutdown();
	};
	process.on("SIGINT", onSignal);
	process.on("SIGTERM", onSignal);
	await new Promise(() => void 0);
}
async function runQaLabSelfCheckCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const server = await startQaLabServer({
		repoRoot,
		outputPath: opts.output ? path.resolve(repoRoot, opts.output) : void 0
	});
	try {
		const result = await server.runSelfCheck();
		process.stdout.write(`QA self-check report: ${result.outputPath}\n`);
	} finally {
		await server.stop();
	}
}
async function runQaSuiteCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const runner = (opts.runner ?? "host").trim().toLowerCase();
	const scenarioIds = resolveQaParityPackScenarioIds({
		parityPack: opts.parityPack,
		scenarioIds: opts.scenarioIds
	});
	if (runner !== "host" && runner !== "multipass") throw new Error(`--runner must be one of host or multipass, got "${opts.runner}".`);
	const providerMode = normalizeQaProviderMode(opts.providerMode);
	const claudeCliAuthMode = parseQaCliBackendAuthMode(opts.cliAuthMode);
	if (runner === "host" && (opts.image !== void 0 || opts.cpus !== void 0 || opts.memory !== void 0 || opts.disk !== void 0)) throw new Error("--image, --cpus, --memory, and --disk require --runner multipass.");
	if (runner === "multipass" && opts.cliAuthMode !== void 0) throw new Error("--cli-auth-mode requires --runner host.");
	if (runner === "multipass") {
		const result = await runQaMultipass({
			repoRoot,
			outputDir: resolveRepoRelativeOutputDir(repoRoot, opts.outputDir),
			providerMode,
			primaryModel: opts.primaryModel,
			alternateModel: opts.alternateModel,
			fastMode: opts.fastMode,
			scenarioIds,
			...opts.concurrency !== void 0 ? { concurrency: parseQaPositiveIntegerOption("--concurrency", opts.concurrency) } : {},
			image: opts.image,
			cpus: parseQaPositiveIntegerOption("--cpus", opts.cpus),
			memory: opts.memory,
			disk: opts.disk
		});
		process.stdout.write(`QA Multipass dir: ${result.outputDir}\n`);
		process.stdout.write(`QA Multipass report: ${result.reportPath}\n`);
		process.stdout.write(`QA Multipass summary: ${result.summaryPath}\n`);
		process.stdout.write(`QA Multipass host log: ${result.hostLogPath}\n`);
		process.stdout.write(`QA Multipass bootstrap log: ${result.bootstrapLogPath}\n`);
		return;
	}
	const result = await runQaSuiteFromRuntime({
		repoRoot,
		outputDir: resolveRepoRelativeOutputDir(repoRoot, opts.outputDir),
		providerMode,
		primaryModel: opts.primaryModel,
		alternateModel: opts.alternateModel,
		fastMode: opts.fastMode,
		...claudeCliAuthMode ? { claudeCliAuthMode } : {},
		scenarioIds,
		...opts.concurrency !== void 0 ? { concurrency: parseQaPositiveIntegerOption("--concurrency", opts.concurrency) } : {}
	});
	process.stdout.write(`QA suite watch: ${result.watchUrl}\n`);
	process.stdout.write(`QA suite report: ${result.reportPath}\n`);
	process.stdout.write(`QA suite summary: ${result.summaryPath}\n`);
}
async function runQaParityReportCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const outputDir = resolveRepoRelativeOutputDir(repoRoot, opts.outputDir) ?? path.join(repoRoot, ".artifacts", "qa-e2e", `parity-${Date.now().toString(36)}`);
	await fs$1.mkdir(outputDir, { recursive: true });
	const candidateSummaryPath = path.resolve(repoRoot, opts.candidateSummary);
	const baselineSummaryPath = path.resolve(repoRoot, opts.baselineSummary);
	const candidateSummary = JSON.parse(await fs$1.readFile(candidateSummaryPath, "utf8"));
	const baselineSummary = JSON.parse(await fs$1.readFile(baselineSummaryPath, "utf8"));
	const comparison = buildQaAgenticParityComparison({
		candidateLabel: opts.candidateLabel?.trim() || "openai/gpt-5.4",
		baselineLabel: opts.baselineLabel?.trim() || "anthropic/claude-opus-4-6",
		candidateSummary,
		baselineSummary
	});
	const report = renderQaAgenticParityMarkdownReport(comparison);
	const reportPath = path.join(outputDir, "qa-agentic-parity-report.md");
	const summaryPath = path.join(outputDir, "qa-agentic-parity-summary.json");
	await fs$1.writeFile(reportPath, report, "utf8");
	await fs$1.writeFile(summaryPath, `${JSON.stringify(comparison, null, 2)}\n`, "utf8");
	process.stdout.write(`QA parity report: ${reportPath}\n`);
	process.stdout.write(`QA parity summary: ${summaryPath}\n`);
	process.stdout.write(`QA parity verdict: ${comparison.pass ? "pass" : "fail"}\n`);
	if (!comparison.pass) process.exitCode = 1;
}
async function runQaCharacterEvalCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const candidates = parseQaModelSpecs("--model", opts.model);
	const judges = parseQaModelSpecs("--judge-model", opts.judgeModel);
	const result = await runQaCharacterEval({
		repoRoot,
		outputDir: resolveRepoRelativeOutputDir(repoRoot, opts.outputDir),
		models: candidates.models,
		scenarioId: opts.scenario,
		candidateFastMode: opts.fast,
		candidateThinkingDefault: parseQaThinkingLevel("--thinking", opts.thinking),
		candidateThinkingByModel: parseQaModelThinkingOverrides(opts.modelThinking),
		candidateModelOptions: candidates.optionsByModel,
		judgeModels: judges.models.length > 0 ? judges.models : void 0,
		judgeModelOptions: judges.optionsByModel,
		judgeTimeoutMs: opts.judgeTimeoutMs,
		judgeBlindModels: opts.blindJudgeModels === true ? true : void 0,
		candidateConcurrency: parseQaPositiveIntegerOption("--concurrency", opts.concurrency),
		judgeConcurrency: parseQaPositiveIntegerOption("--judge-concurrency", opts.judgeConcurrency),
		progress: (message) => process.stderr.write(`${message}\n`)
	});
	process.stdout.write(`QA character eval report: ${result.reportPath}\n`);
	process.stdout.write(`QA character eval summary: ${result.summaryPath}\n`);
}
async function runQaManualLaneCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const providerMode = opts.providerMode === void 0 ? "live-frontier" : normalizeQaProviderMode(opts.providerMode);
	const models = resolveQaManualLaneModels({
		providerMode,
		primaryModel: opts.primaryModel,
		alternateModel: opts.alternateModel
	});
	const result = await runQaManualLane({
		repoRoot,
		providerMode,
		primaryModel: models.primaryModel,
		alternateModel: models.alternateModel,
		fastMode: opts.fastMode,
		message: opts.message,
		timeoutMs: opts.timeoutMs
	});
	process.stdout.write(JSON.stringify(result, null, 2));
	process.stdout.write("\n");
}
async function runQaLabUiCommand(opts) {
	await runInterruptibleServer("QA Lab UI", await startQaLabServer({
		repoRoot: path.resolve(opts.repoRoot ?? process.cwd()),
		host: opts.host,
		port: Number.isFinite(opts.port) ? opts.port : void 0,
		advertiseHost: opts.advertiseHost,
		advertisePort: Number.isFinite(opts.advertisePort) ? opts.advertisePort : void 0,
		controlUiUrl: opts.controlUiUrl,
		controlUiToken: opts.controlUiToken,
		controlUiProxyTarget: opts.controlUiProxyTarget,
		uiDistDir: opts.uiDistDir,
		autoKickoffTarget: opts.autoKickoffTarget,
		embeddedGateway: opts.embeddedGateway,
		sendKickoffOnStart: opts.sendKickoffOnStart
	}));
}
async function runQaDockerScaffoldCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const outputDir = resolveRepoRelativeOutputDir(repoRoot, opts.outputDir);
	if (!outputDir) throw new Error("--output-dir is required.");
	const result = await writeQaDockerHarnessFiles({
		outputDir,
		repoRoot,
		gatewayPort: Number.isFinite(opts.gatewayPort) ? opts.gatewayPort : void 0,
		qaLabPort: Number.isFinite(opts.qaLabPort) ? opts.qaLabPort : void 0,
		providerBaseUrl: opts.providerBaseUrl,
		imageName: opts.image,
		usePrebuiltImage: opts.usePrebuiltImage,
		bindUiDist: opts.bindUiDist
	});
	process.stdout.write(`QA docker scaffold: ${result.outputDir}\n`);
}
async function runQaDockerBuildImageCommand(opts) {
	const result = await buildQaDockerHarnessImage({
		repoRoot: path.resolve(opts.repoRoot ?? process.cwd()),
		imageName: opts.image
	});
	process.stdout.write(`QA docker image: ${result.imageName}\n`);
}
async function runQaDockerUpCommand(opts) {
	const repoRoot = path.resolve(opts.repoRoot ?? process.cwd());
	const result = await runQaDockerUp({
		repoRoot,
		outputDir: resolveRepoRelativeOutputDir(repoRoot, opts.outputDir),
		gatewayPort: Number.isFinite(opts.gatewayPort) ? opts.gatewayPort : void 0,
		qaLabPort: Number.isFinite(opts.qaLabPort) ? opts.qaLabPort : void 0,
		providerBaseUrl: opts.providerBaseUrl,
		image: opts.image,
		usePrebuiltImage: opts.usePrebuiltImage,
		bindUiDist: opts.bindUiDist,
		skipUiBuild: opts.skipUiBuild
	});
	process.stdout.write(`QA docker dir: ${result.outputDir}\n`);
	process.stdout.write(`QA Lab UI: ${result.qaLabUrl}\n`);
	process.stdout.write(`Gateway UI: ${result.gatewayUrl}\n`);
	process.stdout.write(`Stop: ${result.stopCommand}\n`);
}
async function runQaMockOpenAiCommand(opts) {
	await runInterruptibleServer("QA mock OpenAI", await startQaMockOpenAiServer({
		host: opts.host,
		port: Number.isFinite(opts.port) ? opts.port : void 0
	}));
}
const __testing = { resolveRepoRelativeOutputDir };
//#endregion
export { __testing, runQaCharacterEvalCommand, runQaDockerBuildImageCommand, runQaDockerScaffoldCommand, runQaDockerUpCommand, runQaLabSelfCheckCommand, runQaLabUiCommand, runQaManualLaneCommand, runQaMockOpenAiCommand, runQaParityReportCommand, runQaSuiteCommand };
