// openclaw/dist/cli.runtime-CMiWJfIX.js
import { i as formatErrorMessage } from "./errors-D8p6rxH8.js";
import { n as fetchWithSsrFGuard } from "./fetch-guard-DKe4b6qW.js";
import "./error-runtime-Da0X1TOQ.js";
import "./ssrf-runtime-DgTiz1oy.js";
import { i as normalizeQaProviderMode, r as defaultQaModelForMode } from "./run-config-B359hkWD.js";
import { a as buildLiveLaneArtifactsError, i as appendLiveLaneIssue, n as selectLiveTransportScenarios, o as printLiveTransportQaArtifacts, r as startQaLiveLaneGateway, s as resolveLiveTransportQaRunOptions, t as collectLiveTransportStandardScenarioCoverage } from "./live-transport-scenarios-Bb9v3b4q.js";
import path from "node:path";
import fs from "node:fs/promises";
import { randomUUID } from "node:crypto";
//#region extensions/qa-lab/src/live-transports/telegram/telegram-live.runtime.ts
var TelegramQaCanaryError = class extends Error {
	constructor(phase, message, context) {
		super(message);
		this.name = "TelegramQaCanaryError";
		this.phase = phase;
		this.context = context;
	}
};
function isTelegramQaCanaryError(error) {
	return error instanceof TelegramQaCanaryError || typeof error === "object" && error !== null && typeof error.phase === "string" && typeof error.context === "object" && error.context !== null;
}
const TELEGRAM_QA_SCENARIOS = [
	{
		id: "telegram-help-command",
		standardId: "help-command",
		title: "Telegram help command reply",
		timeoutMs: 45e3,
		buildRun: (sutUsername) => ({
			expectReply: true,
			input: `/help@${sutUsername}`,
			expectedTextIncludes: ["/new", "/commands for full list"]
		})
	},
	{
		id: "telegram-commands-command",
		title: "Telegram commands list reply",
		timeoutMs: 45e3,
		buildRun: (sutUsername) => ({
			expectReply: true,
			input: `/commands@${sutUsername}`,
			expectedTextIncludes: ["/help", "More: /tools for available capabilities"]
		})
	},
	{
		id: "telegram-tools-compact-command",
		title: "Telegram tools compact reply",
		timeoutMs: 45e3,
		buildRun: (sutUsername) => ({
			expectReply: true,
			input: `/tools@${sutUsername} compact`,
			expectedTextIncludes: ["exec", "Use /tools verbose for descriptions."]
		})
	},
	{
		id: "telegram-whoami-command",
		title: "Telegram whoami reply",
		timeoutMs: 45e3,
		buildRun: (sutUsername) => ({
			expectReply: true,
			input: `/whoami@${sutUsername}`,
			expectedTextIncludes: ["🧭 Identity", "Channel: telegram"]
		})
	},
	{
		id: "telegram-context-command",
		title: "Telegram context reply",
		timeoutMs: 45e3,
		buildRun: (sutUsername) => ({
			expectReply: true,
			input: `/context@${sutUsername}`,
			expectedTextIncludes: ["/context list", "Inline shortcut"]
		})
	},
	{
		id: "telegram-mentioned-message-reply",
		title: "Telegram mentioned message gets a reply",
		timeoutMs: 45e3,
		buildRun: (sutUsername) => {
			const token = `TELEGRAM_QA_REPLY_${randomUUID().slice(0, 8).toUpperCase()}`;
			return {
				expectReply: true,
				input: `@${sutUsername} reply with only this exact marker: ${token}`,
				expectedTextIncludes: [token],
				matchText: token
			};
		}
	},
	{
		id: "telegram-mention-gating",
		standardId: "mention-gating",
		title: "Telegram group message without mention does not trigger",
		timeoutMs: 8e3,
		buildRun: () => {
			const token = `TELEGRAM_QA_NOMENTION_${randomUUID().slice(0, 8).toUpperCase()}`;
			return {
				expectReply: false,
				input: `reply with only this exact marker: ${token}`,
				matchText: token
			};
		}
	}
];
collectLiveTransportStandardScenarioCoverage({
	alwaysOnStandardScenarioIds: ["canary"],
	scenarios: TELEGRAM_QA_SCENARIOS
});
function resolveEnvValue(env, key) {
	const value = env[key]?.trim();
	if (!value) throw new Error(`Missing ${key}.`);
	return value;
}
function resolveTelegramQaRuntimeEnv(env = process.env) {
	const groupId = resolveEnvValue(env, "OPENCLAW_QA_TELEGRAM_GROUP_ID");
	if (!/^-?\d+$/u.test(groupId)) throw new Error("OPENCLAW_QA_TELEGRAM_GROUP_ID must be a numeric Telegram chat id.");
	return {
		groupId,
		driverToken: resolveEnvValue(env, "OPENCLAW_QA_TELEGRAM_DRIVER_BOT_TOKEN"),
		sutToken: resolveEnvValue(env, "OPENCLAW_QA_TELEGRAM_SUT_BOT_TOKEN")
	};
}
function flattenInlineButtons(replyMarkup) {
	return (replyMarkup?.inline_keyboard ?? []).flat().map((button) => button.text?.trim()).filter((text) => Boolean(text));
}
function detectMediaKinds(message) {
	const kinds = [];
	if (Array.isArray(message.photo) && message.photo.length > 0) kinds.push("photo");
	if (message.document) kinds.push("document");
	if (message.audio) kinds.push("audio");
	if (message.video) kinds.push("video");
	if (message.voice) kinds.push("voice");
	if (message.sticker) kinds.push("sticker");
	return kinds;
}
function normalizeTelegramObservedMessage(update) {
	const message = update.message;
	if (!message?.from?.id) return null;
	return {
		updateId: update.update_id,
		messageId: message.message_id,
		chatId: message.chat.id,
		senderId: message.from.id,
		senderIsBot: message.from.is_bot === true,
		senderUsername: message.from.username,
		text: message.text ?? message.caption ?? "",
		caption: message.caption,
		replyToMessageId: message.reply_to_message?.message_id,
		timestamp: message.date * 1e3,
		inlineButtons: flattenInlineButtons(message.reply_markup),
		mediaKinds: detectMediaKinds(message)
	};
}
function buildTelegramQaConfig(baseCfg, params) {
	const pluginAllow = [...new Set([...baseCfg.plugins?.allow ?? [], "telegram"])];
	const pluginEntries = {
		...baseCfg.plugins?.entries,
		telegram: { enabled: true }
	};
	return {
		...baseCfg,
		plugins: {
			...baseCfg.plugins,
			allow: pluginAllow,
			entries: pluginEntries
		},
		channels: {
			...baseCfg.channels,
			telegram: {
				enabled: true,
				defaultAccount: params.sutAccountId,
				accounts: { [params.sutAccountId]: {
					enabled: true,
					botToken: params.sutToken,
					dmPolicy: "disabled",
					replyToMode: "first",
					groups: { [params.groupId]: {
						groupPolicy: "allowlist",
						allowFrom: [String(params.driverBotId)],
						requireMention: true
					} }
				} }
			}
		}
	};
}
async function callTelegramApi(token, method, body, timeoutMs = 15e3) {
	const { response, release } = await fetchWithSsrFGuard({
		url: `https://api.telegram.org/bot${token}/${method}`,
		init: {
			method: "POST",
			headers: { "content-type": "application/json" },
			body: JSON.stringify(body ?? {})
		},
		signal: AbortSignal.timeout(timeoutMs),
		policy: { hostnameAllowlist: ["api.telegram.org"] },
		auditContext: "qa-lab-telegram-live"
	});
	try {
		const payload = await response.json();
		if (!response.ok || !payload.ok || payload.result === void 0) throw new Error(payload.description?.trim() || `${method} failed with status ${response.status}`);
		return payload.result;
	} finally {
		await release();
	}
}
async function getBotIdentity(token) {
	return await callTelegramApi(token, "getMe");
}
async function flushTelegramUpdates(token) {
	const startedAt = Date.now();
	let offset = 0;
	while (Date.now() - startedAt < 15e3) {
		const updates = await callTelegramApi(token, "getUpdates", {
			offset,
			timeout: 0,
			allowed_updates: ["message"]
		}, 15e3);
		if (updates.length === 0) return offset;
		offset = (updates.at(-1)?.update_id ?? offset) + 1;
	}
	throw new Error("timed out after 15000ms draining Telegram updates");
}
async function sendGroupMessage(token, groupId, text) {
	return await callTelegramApi(token, "sendMessage", {
		chat_id: groupId,
		text,
		disable_notification: true
	});
}
async function waitForObservedMessage(params) {
	const startedAt = Date.now();
	let offset = params.initialOffset;
	while (Date.now() - startedAt < params.timeoutMs) {
		const remainingMs = Math.max(1e3, Math.min(1e4, params.timeoutMs - (Date.now() - startedAt)));
		const timeoutSeconds = Math.max(1, Math.min(10, Math.floor(remainingMs / 1e3)));
		const updates = await callTelegramApi(params.token, "getUpdates", {
			offset,
			timeout: timeoutSeconds,
			allowed_updates: ["message"]
		}, timeoutSeconds * 1e3 + 5e3);
		if (updates.length === 0) continue;
		offset = (updates.at(-1)?.update_id ?? offset) + 1;
		for (const update of updates) {
			const normalized = normalizeTelegramObservedMessage(update);
			if (!normalized) continue;
			params.observedMessages.push(normalized);
			if (params.predicate(normalized)) return {
				message: normalized,
				nextOffset: offset
			};
		}
	}
	throw new Error(`timed out after ${params.timeoutMs}ms waiting for Telegram message`);
}
async function waitForTelegramChannelRunning(gateway, accountId) {
	const startedAt = Date.now();
	while (Date.now() - startedAt < 45e3) {
		try {
			const match = ((await gateway.call("channels.status", {
				probe: false,
				timeoutMs: 2e3
			}, { timeoutMs: 5e3 })).channelAccounts?.telegram ?? []).find((entry) => entry.accountId === accountId);
			if (match?.running && match.restartPending !== true) return;
		} catch {}
		await new Promise((resolve) => setTimeout(resolve, 500));
	}
	throw new Error(`telegram account "${accountId}" did not become ready`);
}
function renderTelegramQaMarkdown(params) {
	const lines = [
		"# Telegram QA Report",
		"",
		`- Group: \`${params.groupId}\``,
		`- Started: ${params.startedAt}`,
		`- Finished: ${params.finishedAt}`,
		"",
		"## Scenarios",
		""
	];
	for (const scenario of params.scenarios) {
		lines.push(`### ${scenario.title}`);
		lines.push("");
		lines.push(`- Status: ${scenario.status}`);
		lines.push(`- Details: ${scenario.details}`);
		lines.push("");
	}
	if (params.cleanupIssues.length > 0) {
		lines.push("## Cleanup");
		lines.push("");
		for (const issue of params.cleanupIssues) lines.push(`- ${issue}`);
		lines.push("");
	}
	return lines.join("\n");
}
function buildObservedMessagesArtifact(params) {
	return params.observedMessages.map((message) => params.includeContent ? { ...message } : {
		updateId: message.updateId,
		messageId: message.messageId,
		chatId: message.chatId,
		senderId: message.senderId,
		senderIsBot: message.senderIsBot,
		senderUsername: message.senderUsername,
		replyToMessageId: message.replyToMessageId,
		timestamp: message.timestamp,
		inlineButtons: message.inlineButtons,
		mediaKinds: message.mediaKinds
	});
}
function findScenario(ids) {
	return selectLiveTransportScenarios({
		ids,
		laneLabel: "Telegram",
		scenarios: TELEGRAM_QA_SCENARIOS
	});
}
function matchesTelegramScenarioReply(params) {
	if (params.message.chatId !== Number(params.groupId) || params.message.senderId !== params.sutBotId) return false;
	if (params.message.replyToMessageId === params.sentMessageId) return true;
	return Boolean(params.matchText && params.message.text.includes(params.matchText));
}
function assertTelegramScenarioReply(params) {
	if (!params.message.text.trim()) throw new Error(`reply message ${params.message.messageId} was empty`);
	for (const expected of params.expectedTextIncludes ?? []) if (!params.message.text.includes(expected)) throw new Error(`reply message ${params.message.messageId} missing expected text: ${expected}`);
}
function classifyCanaryReply(params) {
	if (params.message.chatId !== Number(params.groupId) || params.message.senderId !== params.sutBotId) return "ignore";
	return params.message.replyToMessageId === params.driverMessageId ? "match" : "unthreaded";
}
async function runCanary(params) {
	const offset = await flushTelegramUpdates(params.driverToken);
	const driverMessage = await sendGroupMessage(params.driverToken, params.groupId, `/help@${params.sutUsername}`);
	let firstUnthreadedReply;
	let sutObserved;
	try {
		sutObserved = await waitForObservedMessage({
			token: params.driverToken,
			initialOffset: offset,
			timeoutMs: 3e4,
			observedMessages: params.observedMessages,
			predicate: (message) => {
				const classification = classifyCanaryReply({
					message,
					groupId: params.groupId,
					sutBotId: params.sutBotId,
					driverMessageId: driverMessage.message_id
				});
				if (classification === "ignore") return false;
				if (classification === "unthreaded") {
					firstUnthreadedReply ??= {
						messageId: message.messageId,
						replyToMessageId: message.replyToMessageId,
						text: message.text
					};
					return false;
				}
				return classification === "match";
			}
		});
	} catch (error) {
		if (firstUnthreadedReply) throw new TelegramQaCanaryError("sut_reply_not_threaded", "SUT bot replied, but not as a reply to the canary driver message.", {
			groupId: params.groupId,
			sutBotId: params.sutBotId,
			driverMessageId: driverMessage.message_id,
			sutMessageId: firstUnthreadedReply.messageId,
			sutReplyToMessageId: firstUnthreadedReply.replyToMessageId
		});
		throw new TelegramQaCanaryError("sut_reply_timeout", "SUT bot did not send any group reply after the canary command within 30s.", {
			groupId: params.groupId,
			sutBotId: params.sutBotId,
			driverMessageId: driverMessage.message_id,
			cause: formatErrorMessage(error)
		});
	}
	if (!sutObserved.message.text.trim()) throw new TelegramQaCanaryError("sut_reply_empty", "SUT bot replied to the canary message but the reply text was empty.", {
		groupId: params.groupId,
		sutBotId: params.sutBotId,
		driverMessageId: driverMessage.message_id,
		sutMessageId: sutObserved.message.messageId
	});
}
function canaryFailureMessage(params) {
	const error = params.error;
	const details = formatErrorMessage(error);
	const phase = isTelegramQaCanaryError(error) ? error.phase : "unknown";
	const canonicalContext = new Set([
		"groupId",
		"driverBotId",
		"driverUsername",
		"sutBotId",
		"sutUsername"
	]);
	const context = isTelegramQaCanaryError(error) ? Object.entries(error.context).filter(([key, value]) => value !== void 0 && value !== "" && !canonicalContext.has(key)).map(([key, value]) => `- ${key}: ${String(value)}`) : [];
	const remediation = (() => {
		switch (phase) {
			case "sut_reply_timeout": return [
				"1. Enable Bot-to-Bot Communication Mode for both the driver and SUT bots in @BotFather.",
				"2. Confirm the SUT bot is present in the target private group and can receive /help@BotUsername commands there.",
				"3. Confirm the QA child gateway started the SUT Telegram account with the expected token."
			];
			case "sut_reply_not_threaded": return [
				"1. Check whether the SUT bot is replying in the group without threading to the driver message.",
				"2. Confirm the Telegram native command path preserves reply-to behavior for group commands.",
				"3. Inspect the observed messages artifact for the mismatched SUT message id and reply target."
			];
			case "sut_reply_empty": return [
				"1. Inspect the observed messages artifact to confirm whether the SUT sent media-only or blank text.",
				"2. Check whether the Telegram native command response path produced an empty or suppressed reply.",
				"3. Confirm the SUT command completed successfully in gateway logs."
			];
			default: return [
				"1. Enable Bot-to-Bot Communication Mode for both the driver and SUT bots in @BotFather.",
				"2. Ensure the driver bot can observe bot traffic in the private group by making it admin or disabling privacy mode, then re-add it.",
				"3. Ensure both bots are members of the same private group.",
				"4. Confirm the SUT bot is allowed to receive /help@BotUsername commands in that group."
			];
		}
	})();
	return [
		"Telegram QA canary failed.",
		`Phase: ${phase}`,
		details,
		"Context:",
		`- groupId: ${params.groupId}`,
		`- driverBotId: ${params.driverBotId}`,
		`- driverUsername: ${params.driverUsername ?? "<none>"}`,
		`- sutBotId: ${params.sutBotId}`,
		`- sutUsername: ${params.sutUsername}`,
		...context,
		"Remediation:",
		...remediation
	].join("\n");
}
async function runTelegramQaLive(params) {
	const repoRoot = path.resolve(params.repoRoot ?? process.cwd());
	const outputDir = params.outputDir ?? path.join(repoRoot, ".artifacts", "qa-e2e", `telegram-${Date.now().toString(36)}`);
	await fs.mkdir(outputDir, { recursive: true });
	const runtimeEnv = resolveTelegramQaRuntimeEnv();
	const providerMode = normalizeQaProviderMode(params.providerMode ?? "live-frontier");
	const primaryModel = params.primaryModel?.trim() || defaultQaModelForMode(providerMode);
	const alternateModel = params.alternateModel?.trim() || defaultQaModelForMode(providerMode, true);
	const sutAccountId = params.sutAccountId?.trim() || "sut";
	const scenarios = findScenario(params.scenarioIds);
	const observedMessages = [];
	const includeObservedMessageContent = process.env.OPENCLAW_QA_TELEGRAM_CAPTURE_CONTENT === "1";
	const startedAt = (/* @__PURE__ */ new Date()).toISOString();
	const driverIdentity = await getBotIdentity(runtimeEnv.driverToken);
	const sutIdentity = await getBotIdentity(runtimeEnv.sutToken);
	const sutUsername = sutIdentity.username?.trim();
	if (new Set([driverIdentity.id, sutIdentity.id]).size !== 2) throw new Error("Telegram QA requires two distinct bots for driver and SUT.");
	if (!sutUsername) throw new Error("Telegram QA requires the SUT bot to have a Telegram username.");
	await Promise.all([flushTelegramUpdates(runtimeEnv.driverToken), flushTelegramUpdates(runtimeEnv.sutToken)]);
	const gatewayHarness = await startQaLiveLaneGateway({
		repoRoot,
		qaBusBaseUrl: "http://127.0.0.1:43123",
		providerMode,
		primaryModel,
		alternateModel,
		fastMode: params.fastMode,
		controlUiEnabled: false,
		mutateConfig: (cfg) => buildTelegramQaConfig(cfg, {
			groupId: runtimeEnv.groupId,
			sutToken: runtimeEnv.sutToken,
			driverBotId: driverIdentity.id,
			sutAccountId
		})
	});
	const scenarioResults = [];
	const cleanupIssues = [];
	let canaryFailure = null;
	try {
		await waitForTelegramChannelRunning(gatewayHarness.gateway, sutAccountId);
		try {
			await runCanary({
				driverToken: runtimeEnv.driverToken,
				groupId: runtimeEnv.groupId,
				sutUsername,
				sutBotId: sutIdentity.id,
				observedMessages
			});
		} catch (error) {
			canaryFailure = canaryFailureMessage({
				error,
				groupId: runtimeEnv.groupId,
				driverBotId: driverIdentity.id,
				driverUsername: driverIdentity.username,
				sutBotId: sutIdentity.id,
				sutUsername
			});
			scenarioResults.push({
				id: "telegram-canary",
				title: "Telegram canary",
				status: "fail",
				details: canaryFailure
			});
		}
		if (!canaryFailure) {
			let driverOffset = await flushTelegramUpdates(runtimeEnv.driverToken);
			for (const scenario of scenarios) {
				const scenarioRun = scenario.buildRun(sutUsername);
				try {
					const sent = await sendGroupMessage(runtimeEnv.driverToken, runtimeEnv.groupId, scenarioRun.input);
					const matched = await waitForObservedMessage({
						token: runtimeEnv.driverToken,
						initialOffset: driverOffset,
						timeoutMs: scenario.timeoutMs,
						observedMessages,
						predicate: (message) => matchesTelegramScenarioReply({
							groupId: runtimeEnv.groupId,
							matchText: scenarioRun.matchText,
							message,
							sentMessageId: sent.message_id,
							sutBotId: sutIdentity.id
						})
					});
					driverOffset = matched.nextOffset;
					if (!scenarioRun.expectReply) throw new Error(`unexpected reply message ${matched.message.messageId} matched`);
					assertTelegramScenarioReply({
						expectedTextIncludes: scenarioRun.expectedTextIncludes,
						message: matched.message
					});
					scenarioResults.push({
						id: scenario.id,
						title: scenario.title,
						status: "pass",
						details: `reply message ${matched.message.messageId} matched`
					});
				} catch (error) {
					if (!scenarioRun.expectReply) {
						if (formatErrorMessage(error) === `timed out after ${scenario.timeoutMs}ms waiting for Telegram message`) {
							scenarioResults.push({
								id: scenario.id,
								title: scenario.title,
								status: "pass",
								details: "no reply"
							});
							continue;
						}
					}
					scenarioResults.push({
						id: scenario.id,
						title: scenario.title,
						status: "fail",
						details: formatErrorMessage(error)
					});
				}
			}
		}
	} finally {
		try {
			await gatewayHarness.stop();
		} catch (error) {
			appendLiveLaneIssue(cleanupIssues, "live gateway cleanup", error);
		}
	}
	const finishedAt = (/* @__PURE__ */ new Date()).toISOString();
	const summary = {
		groupId: runtimeEnv.groupId,
		startedAt,
		finishedAt,
		cleanupIssues,
		counts: {
			total: scenarioResults.length,
			passed: scenarioResults.filter((entry) => entry.status === "pass").length,
			failed: scenarioResults.filter((entry) => entry.status === "fail").length
		},
		scenarios: scenarioResults
	};
	const reportPath = path.join(outputDir, "telegram-qa-report.md");
	const summaryPath = path.join(outputDir, "telegram-qa-summary.json");
	const observedMessagesPath = path.join(outputDir, "telegram-qa-observed-messages.json");
	await fs.writeFile(reportPath, `${renderTelegramQaMarkdown({
		cleanupIssues,
		groupId: runtimeEnv.groupId,
		startedAt,
		finishedAt,
		scenarios: scenarioResults
	})}\n`, {
		encoding: "utf8",
		mode: 384
	});
	await fs.writeFile(summaryPath, `${JSON.stringify(summary, null, 2)}\n`, {
		encoding: "utf8",
		mode: 384
	});
	await fs.writeFile(observedMessagesPath, `${JSON.stringify(buildObservedMessagesArtifact({
		observedMessages,
		includeContent: includeObservedMessageContent
	}), null, 2)}\n`, {
		encoding: "utf8",
		mode: 384
	});
	const artifactPaths = {
		report: reportPath,
		summary: summaryPath,
		observedMessages: observedMessagesPath
	};
	if (canaryFailure) throw new Error(buildLiveLaneArtifactsError({
		heading: canaryFailure,
		artifacts: artifactPaths
	}));
	if (cleanupIssues.length > 0) throw new Error(buildLiveLaneArtifactsError({
		heading: "Telegram QA cleanup failed after artifacts were written.",
		details: cleanupIssues,
		artifacts: artifactPaths
	}));
	return {
		outputDir,
		reportPath,
		summaryPath,
		observedMessagesPath,
		scenarios: scenarioResults
	};
}
//#endregion
//#region extensions/qa-lab/src/live-transports/telegram/cli.runtime.ts
async function runQaTelegramCommand(opts) {
	const result = await runTelegramQaLive(resolveLiveTransportQaRunOptions(opts));
	printLiveTransportQaArtifacts("Telegram QA", {
		report: result.reportPath,
		summary: result.summaryPath,
		"observed messages": result.observedMessagesPath
	});
}
//#endregion
export { runQaTelegramCommand };
