// openclaw/dist/cli.runtime-BBLS9wrA.js
import { i as formatErrorMessage } from "./errors-D8p6rxH8.js";
import "./error-runtime-Da0X1TOQ.js";
import { i as normalizeQaProviderMode, r as defaultQaModelForMode } from "./run-config-B359hkWD.js";
import { t as renderQaMarkdownReport } from "./report-fUz6OldI.js";
import { a as waitForDockerServiceHealth, i as resolveHostPort, n as fetchHealthUrl, o as waitForHealth, r as resolveComposeServiceUrl, t as execCommand } from "./docker-runtime-C6coNoAQ.js";
import { a as buildLiveLaneArtifactsError, i as appendLiveLaneIssue, n as selectLiveTransportScenarios, o as printLiveTransportQaArtifacts, r as startQaLiveLaneGateway, s as resolveLiveTransportQaRunOptions, t as collectLiveTransportStandardScenarioCoverage } from "./live-transport-scenarios-Bb9v3b4q.js";
import path from "node:path";
import fs from "node:fs/promises";
import { randomUUID } from "node:crypto";
import { setTimeout as setTimeout$1 } from "node:timers/promises";
//#region extensions/qa-lab/src/live-transports/matrix/matrix-driver-client.ts
function buildMatrixThreadRelation(threadRootEventId, replyToEventId) {
	return { "m.relates_to": {
		rel_type: "m.thread",
		event_id: threadRootEventId,
		is_falling_back: true,
		"m.in_reply_to": { event_id: replyToEventId?.trim() || threadRootEventId }
	} };
}
function buildMatrixReactionRelation(messageId, emoji) {
	const normalizedMessageId = messageId.trim();
	const normalizedEmoji = emoji.trim();
	if (!normalizedMessageId) throw new Error("Matrix reaction requires a messageId");
	if (!normalizedEmoji) throw new Error("Matrix reaction requires an emoji");
	return { "m.relates_to": {
		rel_type: "m.annotation",
		event_id: normalizedMessageId,
		key: normalizedEmoji
	} };
}
function escapeMatrixHtml(value) {
	return value.replace(/[&<>"']/g, (char) => {
		switch (char) {
			case "&": return "&amp;";
			case "<": return "&lt;";
			case ">": return "&gt;";
			case "\"": return "&quot;";
			case "'": return "&#39;";
			default: return char;
		}
	});
}
function buildMatrixMentionLink(userId) {
	return `<a href="${`https://matrix.to/#/${encodeURIComponent(userId)}`}">${escapeMatrixHtml(userId)}</a>`;
}
function buildMatrixQaMessageContent(params) {
	const body = params.body;
	const uniqueMentionUserIds = [...new Set(params.mentionUserIds?.filter(Boolean) ?? [])];
	const formattedParts = [];
	let cursor = 0;
	let usedFormattedMention = false;
	while (cursor < body.length) {
		let matchedUserId = null;
		for (const userId of uniqueMentionUserIds) if (body.startsWith(userId, cursor)) {
			matchedUserId = userId;
			break;
		}
		if (matchedUserId) {
			formattedParts.push(buildMatrixMentionLink(matchedUserId));
			cursor += matchedUserId.length;
			usedFormattedMention = true;
			continue;
		}
		formattedParts.push(escapeMatrixHtml(body[cursor] ?? ""));
		cursor += 1;
	}
	return {
		body,
		msgtype: "m.text",
		...usedFormattedMention ? {
			format: "org.matrix.custom.html",
			formatted_body: formattedParts.join("")
		} : {},
		...uniqueMentionUserIds.length > 0 ? { "m.mentions": { user_ids: uniqueMentionUserIds } } : {},
		...params.threadRootEventId ? buildMatrixThreadRelation(params.threadRootEventId, params.replyToEventId) : {}
	};
}
function normalizeMentionUserIds(value) {
	return Array.isArray(value) ? value.filter((entry) => typeof entry === "string" && entry.trim().length > 0) : void 0;
}
function normalizeMatrixQaObservedEvent(roomId, event) {
	const eventId = event.event_id?.trim();
	const type = event.type?.trim();
	if (!eventId || !type) return null;
	const content = event.content ?? {};
	const relatesToRaw = content["m.relates_to"];
	const relatesTo = typeof relatesToRaw === "object" && relatesToRaw !== null ? relatesToRaw : null;
	const inReplyToRaw = relatesTo?.["m.in_reply_to"];
	const inReplyTo = typeof inReplyToRaw === "object" && inReplyToRaw !== null ? inReplyToRaw : null;
	const mentionsRaw = content["m.mentions"];
	const mentions = typeof mentionsRaw === "object" && mentionsRaw !== null ? mentionsRaw : null;
	const mentionUserIds = normalizeMentionUserIds(mentions?.user_ids);
	const reactionKey = type === "m.reaction" && typeof relatesTo?.key === "string" ? relatesTo.key : void 0;
	const reactionEventId = type === "m.reaction" && typeof relatesTo?.event_id === "string" ? relatesTo.event_id : void 0;
	return {
		roomId,
		eventId,
		sender: typeof event.sender === "string" ? event.sender : void 0,
		stateKey: typeof event.state_key === "string" ? event.state_key : void 0,
		type,
		originServerTs: typeof event.origin_server_ts === "number" ? Math.floor(event.origin_server_ts) : void 0,
		body: typeof content.body === "string" ? content.body : void 0,
		formattedBody: typeof content.formatted_body === "string" ? content.formatted_body : void 0,
		msgtype: typeof content.msgtype === "string" ? content.msgtype : void 0,
		membership: typeof content.membership === "string" ? content.membership : void 0,
		...relatesTo ? { relatesTo: {
			eventId: typeof relatesTo.event_id === "string" ? relatesTo.event_id : void 0,
			inReplyToId: typeof inReplyTo?.event_id === "string" ? inReplyTo.event_id : void 0,
			isFallingBack: typeof relatesTo.is_falling_back === "boolean" ? relatesTo.is_falling_back : void 0,
			relType: typeof relatesTo.rel_type === "string" ? relatesTo.rel_type : void 0
		} } : {},
		...mentions ? { mentions: {
			...mentions.room === true ? { room: true } : {},
			...mentionUserIds ? { userIds: mentionUserIds } : {}
		} } : {},
		...reactionEventId || reactionKey ? { reaction: {
			...reactionEventId ? { eventId: reactionEventId } : {},
			...reactionKey ? { key: reactionKey } : {}
		} } : {}
	};
}
function resolveNextRegistrationAuth(params) {
	const session = params.response.session?.trim();
	if (!session) throw new Error("Matrix registration UIAA response did not include a session id.");
	const completed = new Set((params.response.completed ?? []).filter((stage) => stage === "m.login.dummy" || stage === "m.login.registration_token"));
	const supportedStages = new Set(["m.login.registration_token", "m.login.dummy"]);
	for (const flow of params.response.flows ?? []) {
		const flowStages = flow.stages ?? [];
		if (flowStages.length === 0 || flowStages.some((stage) => !supportedStages.has(stage))) continue;
		const nextStage = flowStages.find((stage) => !completed.has(stage));
		if (!nextStage) continue;
		if (nextStage === "m.login.registration_token") return {
			session,
			type: nextStage,
			token: params.registrationToken
		};
		return {
			session,
			type: nextStage
		};
	}
	throw new Error(`Matrix registration requires unsupported auth stages: ${JSON.stringify(params.response.flows ?? [])}`);
}
async function requestMatrixJson(params) {
	const url = new URL(params.endpoint, params.baseUrl);
	for (const [key, value] of Object.entries(params.query ?? {})) if (value !== void 0) url.searchParams.set(key, String(value));
	const response = await params.fetchImpl(url, {
		method: params.method,
		headers: {
			accept: "application/json",
			...params.body !== void 0 ? { "content-type": "application/json" } : {},
			...params.accessToken ? { authorization: `Bearer ${params.accessToken}` } : {}
		},
		...params.body !== void 0 ? { body: JSON.stringify(params.body) } : {},
		signal: AbortSignal.timeout(params.timeoutMs ?? 2e4)
	});
	let body = {};
	try {
		body = await response.json();
	} catch {
		body = {};
	}
	if (!(params.okStatuses ?? [200]).includes(response.status)) {
		const details = typeof body === "object" && body !== null && typeof body.error === "string" ? body.error : `${params.method} ${params.endpoint} failed with status ${response.status}`;
		throw new Error(details);
	}
	return {
		status: response.status,
		body
	};
}
function buildRegisteredAccount(params) {
	const userId = params.response.user_id?.trim();
	const accessToken = params.response.access_token?.trim();
	if (!userId || !accessToken) throw new Error("Matrix registration did not return both user_id and access_token.");
	return {
		accessToken,
		deviceId: params.response.device_id?.trim() || void 0,
		localpart: params.localpart,
		password: params.password,
		userId
	};
}
function createMatrixQaClient(params) {
	const fetchImpl = params.fetchImpl ?? fetch;
	async function waitForOptionalRoomEvent(opts) {
		const startedAt = Date.now();
		let since = opts.since;
		while (Date.now() - startedAt < opts.timeoutMs) {
			const remainingMs = Math.max(1e3, opts.timeoutMs - (Date.now() - startedAt));
			const response = await requestMatrixJson({
				accessToken: params.accessToken,
				baseUrl: params.baseUrl,
				endpoint: "/_matrix/client/v3/sync",
				fetchImpl,
				method: "GET",
				query: {
					...since ? { since } : {},
					timeout: Math.min(1e4, remainingMs)
				},
				timeoutMs: Math.min(15e3, remainingMs + 5e3)
			});
			since = response.body.next_batch?.trim() || since;
			const roomEvents = response.body.rooms?.join?.[opts.roomId]?.timeline?.events ?? [];
			let matchedEvent = null;
			for (const event of roomEvents) {
				const normalized = normalizeMatrixQaObservedEvent(opts.roomId, event);
				if (!normalized) continue;
				opts.observedEvents.push(normalized);
				if (matchedEvent === null && opts.predicate(normalized)) matchedEvent = normalized;
			}
			if (matchedEvent) return {
				event: matchedEvent,
				matched: true,
				since
			};
		}
		return {
			matched: false,
			since
		};
	}
	return {
		async createPrivateRoom(opts) {
			const roomId = (await requestMatrixJson({
				accessToken: params.accessToken,
				baseUrl: params.baseUrl,
				body: {
					creation_content: { "m.federate": false },
					initial_state: [{
						type: "m.room.history_visibility",
						state_key: "",
						content: { history_visibility: "joined" }
					}],
					invite: opts.inviteUserIds,
					is_direct: false,
					name: opts.name,
					preset: "private_chat"
				},
				endpoint: "/_matrix/client/v3/createRoom",
				fetchImpl,
				method: "POST"
			})).body.room_id?.trim();
			if (!roomId) throw new Error("Matrix createRoom did not return room_id.");
			return roomId;
		},
		async primeRoom() {
			return (await requestMatrixJson({
				accessToken: params.accessToken,
				baseUrl: params.baseUrl,
				endpoint: "/_matrix/client/v3/sync",
				fetchImpl,
				method: "GET",
				query: { timeout: 0 }
			})).body.next_batch?.trim() || void 0;
		},
		async registerWithToken(opts) {
			let auth;
			const baseBody = {
				inhibit_login: false,
				initial_device_display_name: opts.deviceName,
				password: opts.password,
				username: opts.localpart
			};
			for (let attempt = 0; attempt < 4; attempt += 1) {
				const response = await requestMatrixJson({
					baseUrl: params.baseUrl,
					body: {
						...baseBody,
						...auth ? { auth } : {}
					},
					endpoint: "/_matrix/client/v3/register",
					fetchImpl,
					method: "POST",
					okStatuses: [200, 401],
					timeoutMs: 3e4
				});
				if (response.status === 200) return buildRegisteredAccount({
					localpart: opts.localpart,
					password: opts.password,
					response: response.body
				});
				auth = resolveNextRegistrationAuth({
					registrationToken: opts.registrationToken,
					response: response.body
				});
			}
			throw new Error(`Matrix registration for ${opts.localpart} did not complete after 4 attempts.`);
		},
		async sendTextMessage(opts) {
			const txnId = randomUUID();
			const eventId = (await requestMatrixJson({
				accessToken: params.accessToken,
				baseUrl: params.baseUrl,
				body: buildMatrixQaMessageContent(opts),
				endpoint: `/_matrix/client/v3/rooms/${encodeURIComponent(opts.roomId)}/send/m.room.message/${encodeURIComponent(txnId)}`,
				fetchImpl,
				method: "PUT"
			})).body.event_id?.trim();
			if (!eventId) throw new Error("Matrix sendMessage did not return event_id.");
			return eventId;
		},
		async sendReaction(opts) {
			const txnId = randomUUID();
			const eventId = (await requestMatrixJson({
				accessToken: params.accessToken,
				baseUrl: params.baseUrl,
				body: buildMatrixReactionRelation(opts.messageId, opts.emoji),
				endpoint: `/_matrix/client/v3/rooms/${encodeURIComponent(opts.roomId)}/send/m.reaction/${encodeURIComponent(txnId)}`,
				fetchImpl,
				method: "PUT"
			})).body.event_id?.trim();
			if (!eventId) throw new Error("Matrix sendReaction did not return event_id.");
			return eventId;
		},
		async joinRoom(roomId) {
			return (await requestMatrixJson({
				accessToken: params.accessToken,
				baseUrl: params.baseUrl,
				body: {},
				endpoint: `/_matrix/client/v3/join/${encodeURIComponent(roomId)}`,
				fetchImpl,
				method: "POST"
			})).body.room_id?.trim() || roomId;
		},
		waitForOptionalRoomEvent,
		async waitForRoomEvent(opts) {
			const result = await waitForOptionalRoomEvent(opts);
			if (result.matched) return {
				event: result.event,
				since: result.since
			};
			throw new Error(`timed out after ${opts.timeoutMs}ms waiting for Matrix room event`);
		}
	};
}
async function joinRoomWithRetry(params) {
	const client = createMatrixQaClient({
		accessToken: params.accessToken,
		baseUrl: params.baseUrl,
		fetchImpl: params.fetchImpl
	});
	let lastError = null;
	for (let attempt = 1; attempt <= 10; attempt += 1) try {
		await client.joinRoom(params.roomId);
		return;
	} catch (error) {
		lastError = error;
		await new Promise((resolve) => setTimeout(resolve, 300 * attempt));
	}
	throw new Error(`Matrix join retry failed: ${formatErrorMessage(lastError)}`);
}
async function provisionMatrixQaRoom(params) {
	const anonClient = createMatrixQaClient({
		baseUrl: params.baseUrl,
		fetchImpl: params.fetchImpl
	});
	const driver = await anonClient.registerWithToken({
		deviceName: "OpenClaw Matrix QA Driver",
		localpart: params.driverLocalpart,
		password: `driver-${randomUUID()}`,
		registrationToken: params.registrationToken
	});
	const sut = await anonClient.registerWithToken({
		deviceName: "OpenClaw Matrix QA SUT",
		localpart: params.sutLocalpart,
		password: `sut-${randomUUID()}`,
		registrationToken: params.registrationToken
	});
	const observer = await anonClient.registerWithToken({
		deviceName: "OpenClaw Matrix QA Observer",
		localpart: params.observerLocalpart,
		password: `observer-${randomUUID()}`,
		registrationToken: params.registrationToken
	});
	const roomId = await createMatrixQaClient({
		accessToken: driver.accessToken,
		baseUrl: params.baseUrl,
		fetchImpl: params.fetchImpl
	}).createPrivateRoom({
		inviteUserIds: [sut.userId, observer.userId],
		name: params.roomName
	});
	await joinRoomWithRetry({
		accessToken: sut.accessToken,
		baseUrl: params.baseUrl,
		fetchImpl: params.fetchImpl,
		roomId
	});
	await joinRoomWithRetry({
		accessToken: observer.accessToken,
		baseUrl: params.baseUrl,
		fetchImpl: params.fetchImpl,
		roomId
	});
	return {
		driver,
		observer,
		roomId,
		sut
	};
}
//#endregion
//#region extensions/qa-lab/src/live-transports/matrix/matrix-harness.runtime.ts
const MATRIX_QA_DEFAULT_IMAGE = "ghcr.io/matrix-construct/tuwunel:v1.5.1";
const MATRIX_QA_DEFAULT_SERVER_NAME = "matrix-qa.test";
const MATRIX_QA_DEFAULT_PORT = 28008;
const MATRIX_QA_INTERNAL_PORT = 8008;
const MATRIX_QA_SERVICE = "matrix-qa-homeserver";
function buildVersionsUrl(baseUrl) {
	return `${baseUrl}_matrix/client/versions`;
}
async function isMatrixVersionsReachable(baseUrl, fetchImpl) {
	return await fetchImpl(buildVersionsUrl(baseUrl)).then((response) => response.ok).catch(() => false);
}
async function waitForReachableMatrixBaseUrl(params) {
	const timeoutMs = params.timeoutMs ?? 6e4;
	const pollMs = params.pollMs ?? 1e3;
	const startedAt = Date.now();
	while (Date.now() - startedAt < timeoutMs) {
		if (await isMatrixVersionsReachable(params.hostBaseUrl, params.fetchImpl)) return params.hostBaseUrl;
		if (params.containerBaseUrl && await isMatrixVersionsReachable(params.containerBaseUrl, params.fetchImpl)) return params.containerBaseUrl;
		await params.sleepImpl(pollMs);
	}
	const candidateLabel = params.containerBaseUrl ? `${params.hostBaseUrl} or ${params.containerBaseUrl}` : params.hostBaseUrl;
	throw new Error([
		`Matrix homeserver did not become healthy within ${Math.round(timeoutMs / 1e3)}s.`,
		`Last checked: ${candidateLabel}`,
		`Hint: check container logs with \`docker compose -f ${params.composeFile} logs ${MATRIX_QA_SERVICE}\`.`
	].join("\n"));
}
function resolveMatrixQaHarnessImage(image) {
	return image?.trim() || process.env.OPENCLAW_QA_MATRIX_TUWUNEL_IMAGE?.trim() || MATRIX_QA_DEFAULT_IMAGE;
}
function renderMatrixQaCompose(params) {
	return `services:
  ${MATRIX_QA_SERVICE}:
    image: ${params.image}
    ports:
      - "127.0.0.1:${params.homeserverPort}:${MATRIX_QA_INTERNAL_PORT}"
    environment:
      TUWUNEL_ADDRESS: "0.0.0.0"
      TUWUNEL_ALLOW_ENCRYPTION: "false"
      TUWUNEL_ALLOW_FEDERATION: "false"
      TUWUNEL_ALLOW_REGISTRATION: "true"
      TUWUNEL_DATABASE_PATH: "/var/lib/tuwunel"
      TUWUNEL_PORT: "${MATRIX_QA_INTERNAL_PORT}"
      TUWUNEL_REGISTRATION_TOKEN: "${params.registrationToken}"
      TUWUNEL_SERVER_NAME: "${params.serverName}"
    volumes:
      - ./data:/var/lib/tuwunel
`;
}
async function writeMatrixQaHarnessFiles(params) {
	const image = resolveMatrixQaHarnessImage(params.image);
	const registrationToken = params.registrationToken?.trim() || `matrix-qa-${randomUUID()}`;
	const serverName = params.serverName?.trim() || MATRIX_QA_DEFAULT_SERVER_NAME;
	const composeFile = path.join(params.outputDir, "docker-compose.matrix-qa.yml");
	const dataDir = path.join(params.outputDir, "data");
	const manifestPath = path.join(params.outputDir, "matrix-qa-harness.json");
	await fs.mkdir(dataDir, { recursive: true });
	await fs.writeFile(composeFile, `${renderMatrixQaCompose({
		homeserverPort: params.homeserverPort,
		image,
		registrationToken,
		serverName
	})}\n`, {
		encoding: "utf8",
		mode: 384
	});
	const manifest = {
		image,
		serverName,
		homeserverPort: params.homeserverPort,
		composeFile,
		dataDir
	};
	await fs.writeFile(manifestPath, `${JSON.stringify(manifest, null, 2)}\n`, {
		encoding: "utf8",
		mode: 384
	});
	return {
		outputDir: params.outputDir,
		composeFile,
		manifestPath,
		image,
		serverName,
		homeserverPort: params.homeserverPort,
		registrationToken
	};
}
async function startMatrixQaHarness(params, deps) {
	const repoRoot = path.resolve(params.repoRoot ?? process.cwd());
	const resolveHostPortImpl = deps?.resolveHostPortImpl ?? resolveHostPort;
	const runCommand = deps?.runCommand ?? execCommand;
	const fetchImpl = deps?.fetchImpl ?? fetchHealthUrl;
	const sleepImpl = deps?.sleepImpl ?? setTimeout$1;
	const homeserverPort = await resolveHostPortImpl(params.homeserverPort ?? MATRIX_QA_DEFAULT_PORT, params.homeserverPort != null);
	const files = await writeMatrixQaHarnessFiles({
		outputDir: path.resolve(params.outputDir),
		image: params.image,
		homeserverPort,
		serverName: params.serverName
	});
	try {
		await runCommand("docker", [
			"compose",
			"-f",
			files.composeFile,
			"down",
			"--remove-orphans"
		], repoRoot);
	} catch {}
	await runCommand("docker", [
		"compose",
		"-f",
		files.composeFile,
		"up",
		"-d"
	], repoRoot);
	await sleepImpl(1e3);
	await waitForDockerServiceHealth(MATRIX_QA_SERVICE, files.composeFile, repoRoot, runCommand, sleepImpl);
	const hostBaseUrl = `http://127.0.0.1:${homeserverPort}/`;
	let baseUrl = hostBaseUrl;
	if (!await isMatrixVersionsReachable(hostBaseUrl, fetchImpl)) {
		const containerBaseUrl = await resolveComposeServiceUrl(MATRIX_QA_SERVICE, MATRIX_QA_INTERNAL_PORT, files.composeFile, repoRoot, runCommand);
		baseUrl = await waitForReachableMatrixBaseUrl({
			composeFile: files.composeFile,
			containerBaseUrl,
			fetchImpl,
			hostBaseUrl,
			sleepImpl
		});
	}
	await waitForHealth(buildVersionsUrl(baseUrl), {
		label: "Matrix homeserver",
		composeFile: files.composeFile,
		fetchImpl,
		sleepImpl
	});
	return {
		...files,
		baseUrl,
		stopCommand: `docker compose -f ${files.composeFile} down --remove-orphans`,
		async stop() {
			await runCommand("docker", [
				"compose",
				"-f",
				files.composeFile,
				"down",
				"--remove-orphans"
			], repoRoot);
		}
	};
}
//#endregion
//#region extensions/qa-lab/src/live-transports/matrix/matrix-live-scenarios.ts
const NO_REPLY_WINDOW_MS = 8e3;
const MATRIX_QA_SCENARIOS = [
	{
		id: "matrix-thread-follow-up",
		standardId: "thread-follow-up",
		timeoutMs: 6e4,
		title: "Matrix thread follow-up reply"
	},
	{
		id: "matrix-thread-isolation",
		standardId: "thread-isolation",
		timeoutMs: 75e3,
		title: "Matrix top-level reply stays out of prior thread"
	},
	{
		id: "matrix-top-level-reply-shape",
		standardId: "top-level-reply-shape",
		timeoutMs: 45e3,
		title: "Matrix top-level reply keeps replyToMode off"
	},
	{
		id: "matrix-reaction-notification",
		standardId: "reaction-observation",
		timeoutMs: 45e3,
		title: "Matrix reactions on bot replies are observed"
	},
	{
		id: "matrix-restart-resume",
		standardId: "restart-resume",
		timeoutMs: 6e4,
		title: "Matrix lane resumes cleanly after gateway restart"
	},
	{
		id: "matrix-mention-gating",
		standardId: "mention-gating",
		timeoutMs: NO_REPLY_WINDOW_MS,
		title: "Matrix room message without mention does not trigger"
	},
	{
		id: "matrix-allowlist-block",
		standardId: "allowlist-block",
		timeoutMs: NO_REPLY_WINDOW_MS,
		title: "Matrix allowlist blocks non-driver replies"
	}
];
collectLiveTransportStandardScenarioCoverage({
	alwaysOnStandardScenarioIds: ["canary"],
	scenarios: MATRIX_QA_SCENARIOS
});
function findMatrixQaScenarios(ids) {
	return selectLiveTransportScenarios({
		ids,
		laneLabel: "Matrix",
		scenarios: MATRIX_QA_SCENARIOS
	});
}
function buildMentionPrompt(sutUserId, token) {
	return `${sutUserId} reply with only this exact marker: ${token}`;
}
function buildExactMarkerPrompt(token) {
	return `reply with only this exact marker: ${token}`;
}
function buildMatrixReplyArtifact(event, token) {
	const replyBody = event.body?.trim();
	return {
		bodyPreview: replyBody?.slice(0, 200),
		eventId: event.eventId,
		mentions: event.mentions,
		relatesTo: event.relatesTo,
		sender: event.sender,
		...token ? { tokenMatched: replyBody === token } : {}
	};
}
function buildMatrixReplyDetails(label, artifact) {
	return [
		`${label} event: ${artifact.eventId}`,
		`${label} token matched: ${artifact.tokenMatched === void 0 ? "n/a" : artifact.tokenMatched ? "yes" : "no"}`,
		`${label} rel_type: ${artifact.relatesTo?.relType ?? "<none>"}`,
		`${label} in_reply_to: ${artifact.relatesTo?.inReplyToId ?? "<none>"}`,
		`${label} is_falling_back: ${artifact.relatesTo?.isFallingBack === true ? "true" : "false"}`
	];
}
function assertTopLevelReplyArtifact(label, artifact) {
	if (!artifact.tokenMatched) throw new Error(`${label} did not contain the expected token`);
	if (artifact.relatesTo !== void 0) throw new Error(`${label} unexpectedly included relation metadata`);
}
function assertThreadReplyArtifact(artifact, params) {
	if (!artifact.tokenMatched) throw new Error(`${params.label} did not contain the expected token`);
	if (artifact.relatesTo?.relType !== "m.thread") throw new Error(`${params.label} did not use m.thread`);
	if (artifact.relatesTo.eventId !== params.expectedRootEventId) throw new Error(`${params.label} targeted ${artifact.relatesTo.eventId ?? "<none>"} instead of ${params.expectedRootEventId}`);
	if (artifact.relatesTo.isFallingBack !== true) throw new Error(`${params.label} did not set is_falling_back`);
	if (!artifact.relatesTo.inReplyToId) throw new Error(`${params.label} did not set m.in_reply_to`);
}
function readMatrixQaSyncCursor(syncState, actorId) {
	return syncState[actorId];
}
function writeMatrixQaSyncCursor(syncState, actorId, since) {
	if (since) syncState[actorId] = since;
}
async function primeMatrixQaActorCursor(params) {
	const client = createMatrixQaClient({
		accessToken: params.accessToken,
		baseUrl: params.baseUrl
	});
	const existingSince = readMatrixQaSyncCursor(params.syncState, params.actorId);
	if (existingSince) return {
		client,
		startSince: existingSince
	};
	const startSince = await client.primeRoom();
	if (!startSince) throw new Error(`Matrix ${params.actorId} /sync prime did not return a next_batch cursor`);
	return {
		client,
		startSince
	};
}
function advanceMatrixQaActorCursor(params) {
	writeMatrixQaSyncCursor(params.syncState, params.actorId, params.nextSince ?? params.startSince);
}
async function runTopLevelMentionScenario(params) {
	const { client, startSince } = await primeMatrixQaActorCursor({
		accessToken: params.accessToken,
		actorId: params.actorId,
		baseUrl: params.baseUrl,
		syncState: params.syncState
	});
	const token = `${params.tokenPrefix}_${randomUUID().slice(0, 8).toUpperCase()}`;
	const body = params.withMention === false ? buildExactMarkerPrompt(token) : buildMentionPrompt(params.sutUserId, token);
	const driverEventId = await client.sendTextMessage({
		body,
		...params.withMention === false ? {} : { mentionUserIds: [params.sutUserId] },
		roomId: params.roomId
	});
	const matched = await client.waitForRoomEvent({
		observedEvents: params.observedEvents,
		predicate: (event) => event.roomId === params.roomId && event.sender === params.sutUserId && event.type === "m.room.message" && (event.body ?? "").includes(token) && event.relatesTo === void 0,
		roomId: params.roomId,
		since: startSince,
		timeoutMs: params.timeoutMs
	});
	advanceMatrixQaActorCursor({
		actorId: params.actorId,
		syncState: params.syncState,
		nextSince: matched.since,
		startSince
	});
	return {
		body,
		driverEventId,
		reply: buildMatrixReplyArtifact(matched.event, token),
		since: matched.since,
		token
	};
}
async function runThreadScenario(params) {
	const { client, startSince } = await primeMatrixQaActorCursor({
		accessToken: params.driverAccessToken,
		actorId: "driver",
		baseUrl: params.baseUrl,
		syncState: params.syncState
	});
	const rootBody = `thread root ${randomUUID().slice(0, 8)}`;
	const rootEventId = await client.sendTextMessage({
		body: rootBody,
		roomId: params.roomId
	});
	const token = `MATRIX_QA_THREAD_${randomUUID().slice(0, 8).toUpperCase()}`;
	const driverEventId = await client.sendTextMessage({
		body: buildMentionPrompt(params.sutUserId, token),
		mentionUserIds: [params.sutUserId],
		replyToEventId: rootEventId,
		roomId: params.roomId,
		threadRootEventId: rootEventId
	});
	const matched = await client.waitForRoomEvent({
		observedEvents: params.observedEvents,
		predicate: (event) => event.roomId === params.roomId && event.sender === params.sutUserId && event.type === "m.room.message" && (event.body ?? "").includes(token) && event.relatesTo?.relType === "m.thread" && event.relatesTo.eventId === rootEventId,
		roomId: params.roomId,
		since: startSince,
		timeoutMs: params.timeoutMs
	});
	advanceMatrixQaActorCursor({
		actorId: "driver",
		syncState: params.syncState,
		nextSince: matched.since,
		startSince
	});
	return {
		driverEventId,
		reply: buildMatrixReplyArtifact(matched.event, token),
		rootEventId,
		since: matched.since,
		token
	};
}
async function runNoReplyExpectedScenario(params) {
	const { client, startSince } = await primeMatrixQaActorCursor({
		accessToken: params.accessToken,
		actorId: params.actorId,
		baseUrl: params.baseUrl,
		syncState: params.syncState
	});
	const driverEventId = await client.sendTextMessage({
		body: params.body,
		...params.mentionUserIds ? { mentionUserIds: params.mentionUserIds } : {},
		roomId: params.roomId
	});
	const result = await client.waitForOptionalRoomEvent({
		observedEvents: params.observedEvents,
		predicate: (event) => event.roomId === params.roomId && event.sender === params.sutUserId && event.type === "m.room.message",
		roomId: params.roomId,
		since: startSince,
		timeoutMs: params.timeoutMs
	});
	if (result.matched) {
		const unexpectedReply = buildMatrixReplyArtifact(result.event, params.token);
		throw new Error([
			`unexpected SUT reply from ${params.sutUserId}`,
			`trigger sender: ${params.actorUserId}`,
			...buildMatrixReplyDetails("unexpected reply", unexpectedReply)
		].join("\n"));
	}
	advanceMatrixQaActorCursor({
		actorId: params.actorId,
		syncState: params.syncState,
		nextSince: result.since,
		startSince
	});
	return {
		artifacts: {
			actorUserId: params.actorUserId,
			driverEventId,
			expectedNoReplyWindowMs: params.timeoutMs,
			token: params.token,
			triggerBody: params.body
		},
		details: [
			`trigger event: ${driverEventId}`,
			`trigger sender: ${params.actorUserId}`,
			`waited ${params.timeoutMs}ms with no SUT reply`
		].join("\n")
	};
}
async function runReactionNotificationScenario(context) {
	const reactionTargetEventId = context.canary?.reply.eventId?.trim();
	if (!reactionTargetEventId) throw new Error("Matrix reaction scenario requires a canary reply event id");
	const { client, startSince } = await primeMatrixQaActorCursor({
		accessToken: context.driverAccessToken,
		actorId: "driver",
		baseUrl: context.baseUrl,
		syncState: context.syncState
	});
	const reactionEmoji = "👍";
	const reactionEventId = await client.sendReaction({
		emoji: reactionEmoji,
		messageId: reactionTargetEventId,
		roomId: context.roomId
	});
	const matched = await client.waitForRoomEvent({
		observedEvents: context.observedEvents,
		predicate: (event) => event.roomId === context.roomId && event.sender === context.driverUserId && event.type === "m.reaction" && event.eventId === reactionEventId && event.reaction?.eventId === reactionTargetEventId && event.reaction?.key === reactionEmoji,
		roomId: context.roomId,
		since: startSince,
		timeoutMs: context.timeoutMs
	});
	advanceMatrixQaActorCursor({
		actorId: "driver",
		syncState: context.syncState,
		nextSince: matched.since,
		startSince
	});
	return {
		artifacts: {
			reactionEmoji,
			reactionEventId,
			reactionTargetEventId
		},
		details: [
			`reaction event: ${reactionEventId}`,
			`reaction target: ${reactionTargetEventId}`,
			`reaction emoji: ${reactionEmoji}`,
			`observed reaction key: ${matched.event.reaction?.key ?? "<none>"}`
		].join("\n")
	};
}
async function runRestartResumeScenario(context) {
	if (!context.restartGateway) throw new Error("Matrix restart scenario requires a gateway restart callback");
	await context.restartGateway();
	const result = await runTopLevelMentionScenario({
		accessToken: context.driverAccessToken,
		actorId: "driver",
		baseUrl: context.baseUrl,
		observedEvents: context.observedEvents,
		roomId: context.roomId,
		syncState: context.syncState,
		sutUserId: context.sutUserId,
		timeoutMs: context.timeoutMs,
		tokenPrefix: "MATRIX_QA_RESTART"
	});
	assertTopLevelReplyArtifact("post-restart reply", result.reply);
	return {
		artifacts: {
			driverEventId: result.driverEventId,
			reply: result.reply,
			restartSignal: "SIGUSR1",
			token: result.token
		},
		details: [
			"restart signal: SIGUSR1",
			`post-restart driver event: ${result.driverEventId}`,
			...buildMatrixReplyDetails("reply", result.reply)
		].join("\n")
	};
}
async function runMatrixQaCanary(params) {
	const canary = await runTopLevelMentionScenario({
		accessToken: params.driverAccessToken,
		actorId: "driver",
		baseUrl: params.baseUrl,
		observedEvents: params.observedEvents,
		roomId: params.roomId,
		syncState: params.syncState,
		sutUserId: params.sutUserId,
		timeoutMs: params.timeoutMs,
		tokenPrefix: "MATRIX_QA_CANARY"
	});
	assertTopLevelReplyArtifact("canary reply", canary.reply);
	return canary;
}
async function runMatrixQaScenario(scenario, context) {
	switch (scenario.id) {
		case "matrix-thread-follow-up": {
			const result = await runThreadScenario(context);
			assertThreadReplyArtifact(result.reply, {
				expectedRootEventId: result.rootEventId,
				label: "thread reply"
			});
			return {
				artifacts: {
					driverEventId: result.driverEventId,
					reply: result.reply,
					rootEventId: result.rootEventId,
					token: result.token
				},
				details: [
					`root event: ${result.rootEventId}`,
					`driver thread event: ${result.driverEventId}`,
					...buildMatrixReplyDetails("reply", result.reply)
				].join("\n")
			};
		}
		case "matrix-thread-isolation": {
			const threadPhase = await runThreadScenario(context);
			assertThreadReplyArtifact(threadPhase.reply, {
				expectedRootEventId: threadPhase.rootEventId,
				label: "thread isolation reply"
			});
			const topLevelPhase = await runTopLevelMentionScenario({
				accessToken: context.driverAccessToken,
				actorId: "driver",
				baseUrl: context.baseUrl,
				observedEvents: context.observedEvents,
				roomId: context.roomId,
				syncState: context.syncState,
				sutUserId: context.sutUserId,
				timeoutMs: context.timeoutMs,
				tokenPrefix: "MATRIX_QA_TOPLEVEL"
			});
			assertTopLevelReplyArtifact("top-level follow-up reply", topLevelPhase.reply);
			return {
				artifacts: {
					threadDriverEventId: threadPhase.driverEventId,
					threadReply: threadPhase.reply,
					threadRootEventId: threadPhase.rootEventId,
					threadToken: threadPhase.token,
					topLevelDriverEventId: topLevelPhase.driverEventId,
					topLevelReply: topLevelPhase.reply,
					topLevelToken: topLevelPhase.token
				},
				details: [
					`thread root event: ${threadPhase.rootEventId}`,
					`thread driver event: ${threadPhase.driverEventId}`,
					...buildMatrixReplyDetails("thread reply", threadPhase.reply),
					`top-level driver event: ${topLevelPhase.driverEventId}`,
					...buildMatrixReplyDetails("top-level reply", topLevelPhase.reply)
				].join("\n")
			};
		}
		case "matrix-top-level-reply-shape": {
			const result = await runTopLevelMentionScenario({
				accessToken: context.driverAccessToken,
				actorId: "driver",
				baseUrl: context.baseUrl,
				observedEvents: context.observedEvents,
				roomId: context.roomId,
				syncState: context.syncState,
				sutUserId: context.sutUserId,
				timeoutMs: context.timeoutMs,
				tokenPrefix: "MATRIX_QA_TOPLEVEL"
			});
			assertTopLevelReplyArtifact("top-level reply", result.reply);
			return {
				artifacts: {
					driverEventId: result.driverEventId,
					reply: result.reply,
					token: result.token
				},
				details: [`driver event: ${result.driverEventId}`, ...buildMatrixReplyDetails("reply", result.reply)].join("\n")
			};
		}
		case "matrix-reaction-notification": return await runReactionNotificationScenario(context);
		case "matrix-restart-resume": return await runRestartResumeScenario(context);
		case "matrix-mention-gating": {
			const token = `MATRIX_QA_NOMENTION_${randomUUID().slice(0, 8).toUpperCase()}`;
			return await runNoReplyExpectedScenario({
				accessToken: context.driverAccessToken,
				actorId: "driver",
				actorUserId: context.driverUserId,
				baseUrl: context.baseUrl,
				body: buildExactMarkerPrompt(token),
				observedEvents: context.observedEvents,
				roomId: context.roomId,
				syncState: context.syncState,
				sutUserId: context.sutUserId,
				timeoutMs: context.timeoutMs,
				token
			});
		}
		case "matrix-allowlist-block": {
			const token = `MATRIX_QA_ALLOWLIST_${randomUUID().slice(0, 8).toUpperCase()}`;
			return await runNoReplyExpectedScenario({
				accessToken: context.observerAccessToken,
				actorId: "observer",
				actorUserId: context.observerUserId,
				baseUrl: context.baseUrl,
				body: buildMentionPrompt(context.sutUserId, token),
				mentionUserIds: [context.sutUserId],
				observedEvents: context.observedEvents,
				roomId: context.roomId,
				syncState: context.syncState,
				sutUserId: context.sutUserId,
				timeoutMs: context.timeoutMs,
				token
			});
		}
		default: return scenario.id;
	}
}
//#endregion
//#region extensions/qa-lab/src/live-transports/matrix/matrix-live.runtime.ts
function buildMatrixQaSummary(params) {
	return {
		checks: params.checks,
		counts: {
			total: params.checks.length + params.scenarios.length,
			passed: params.checks.filter((check) => check.status === "pass").length + params.scenarios.filter((scenario) => scenario.status === "pass").length,
			failed: params.checks.filter((check) => check.status === "fail").length + params.scenarios.filter((scenario) => scenario.status === "fail").length
		},
		finishedAt: params.finishedAt,
		harness: params.harness,
		canary: params.canary,
		observedEventCount: params.observedEventCount,
		observedEventsPath: params.artifactPaths.observedEvents,
		reportPath: params.artifactPaths.report,
		scenarios: params.scenarios,
		startedAt: params.startedAt,
		summaryPath: params.artifactPaths.summary,
		sutAccountId: params.sutAccountId,
		userIds: params.userIds
	};
}
function buildMatrixQaConfig(baseCfg, params) {
	const pluginAllow = [...new Set([...baseCfg.plugins?.allow ?? [], "matrix"])];
	return {
		...baseCfg,
		plugins: {
			...baseCfg.plugins,
			allow: pluginAllow,
			entries: {
				...baseCfg.plugins?.entries,
				matrix: { enabled: true }
			}
		},
		channels: {
			...baseCfg.channels,
			matrix: {
				enabled: true,
				defaultAccount: params.sutAccountId,
				accounts: { [params.sutAccountId]: {
					accessToken: params.sutAccessToken,
					...params.sutDeviceId ? { deviceId: params.sutDeviceId } : {},
					dm: { enabled: false },
					enabled: true,
					encryption: false,
					groupAllowFrom: [params.driverUserId],
					groupPolicy: "allowlist",
					groups: { [params.roomId]: {
						enabled: true,
						requireMention: true
					} },
					homeserver: params.homeserver,
					network: { dangerouslyAllowPrivateNetwork: true },
					replyToMode: "off",
					threadReplies: "inbound",
					userId: params.sutUserId
				} }
			}
		}
	};
}
function buildObservedEventsArtifact(params) {
	return params.observedEvents.map((event) => params.includeContent ? event : {
		roomId: event.roomId,
		eventId: event.eventId,
		sender: event.sender,
		stateKey: event.stateKey,
		type: event.type,
		originServerTs: event.originServerTs,
		msgtype: event.msgtype,
		membership: event.membership,
		relatesTo: event.relatesTo,
		mentions: event.mentions,
		reaction: event.reaction
	});
}
function isMatrixAccountReady(entry) {
	return entry?.running === true && entry.connected === true && entry.restartPending !== true && (entry.healthState === void 0 || entry.healthState === "healthy");
}
async function waitForMatrixChannelReady(gateway, accountId, opts) {
	const pollMs = opts?.pollMs ?? 500;
	const timeoutMs = opts?.timeoutMs ?? 6e4;
	const startedAt = Date.now();
	while (Date.now() - startedAt < timeoutMs) {
		try {
			if (isMatrixAccountReady(((await gateway.call("channels.status", {
				probe: false,
				timeoutMs: 2e3
			}, { timeoutMs: 5e3 })).channelAccounts?.matrix ?? []).find((entry) => entry.accountId === accountId))) return;
		} catch {}
		await setTimeout$1(pollMs);
	}
	throw new Error(`matrix account "${accountId}" did not become ready`);
}
async function runMatrixQaLive(params) {
	const repoRoot = path.resolve(params.repoRoot ?? process.cwd());
	const outputDir = params.outputDir ?? path.join(repoRoot, ".artifacts", "qa-e2e", `matrix-${Date.now().toString(36)}`);
	await fs.mkdir(outputDir, { recursive: true });
	const providerMode = normalizeQaProviderMode(params.providerMode ?? "live-frontier");
	const primaryModel = params.primaryModel?.trim() || defaultQaModelForMode(providerMode);
	const alternateModel = params.alternateModel?.trim() || defaultQaModelForMode(providerMode, true);
	const sutAccountId = params.sutAccountId?.trim() || "sut";
	const scenarios = findMatrixQaScenarios(params.scenarioIds);
	const observedEvents = [];
	const includeObservedEventContent = process.env.OPENCLAW_QA_MATRIX_CAPTURE_CONTENT === "1";
	const startedAtDate = /* @__PURE__ */ new Date();
	const startedAt = startedAtDate.toISOString();
	const runSuffix = randomUUID().slice(0, 8);
	const harness = await startMatrixQaHarness({
		outputDir: path.join(outputDir, "matrix-harness"),
		repoRoot
	});
	const provisioning = await (async () => {
		try {
			return await provisionMatrixQaRoom({
				baseUrl: harness.baseUrl,
				driverLocalpart: `qa-driver-${runSuffix}`,
				observerLocalpart: `qa-observer-${runSuffix}`,
				registrationToken: harness.registrationToken,
				roomName: `OpenClaw Matrix QA ${runSuffix}`,
				sutLocalpart: `qa-sut-${runSuffix}`
			});
		} catch (error) {
			await harness.stop().catch(() => {});
			throw error;
		}
	})();
	const checks = [{
		name: "Matrix harness ready",
		status: "pass",
		details: [
			`image: ${harness.image}`,
			`baseUrl: ${harness.baseUrl}`,
			`serverName: ${harness.serverName}`,
			`roomId: ${provisioning.roomId}`
		].join("\n")
	}];
	const scenarioResults = [];
	const cleanupErrors = [];
	let canaryArtifact;
	let gatewayHarness = null;
	let canaryFailed = false;
	const syncState = {};
	try {
		gatewayHarness = await startQaLiveLaneGateway({
			repoRoot,
			qaBusBaseUrl: "http://127.0.0.1:43123",
			providerMode,
			primaryModel,
			alternateModel,
			fastMode: params.fastMode,
			controlUiEnabled: false,
			mutateConfig: (cfg) => buildMatrixQaConfig(cfg, {
				driverUserId: provisioning.driver.userId,
				homeserver: harness.baseUrl,
				roomId: provisioning.roomId,
				sutAccessToken: provisioning.sut.accessToken,
				sutAccountId,
				sutDeviceId: provisioning.sut.deviceId,
				sutUserId: provisioning.sut.userId
			})
		});
		await waitForMatrixChannelReady(gatewayHarness.gateway, sutAccountId);
		checks.push({
			name: "Matrix channel ready",
			status: "pass",
			details: `accountId: ${sutAccountId}\nuserId: ${provisioning.sut.userId}`
		});
		try {
			const canary = await runMatrixQaCanary({
				baseUrl: harness.baseUrl,
				driverAccessToken: provisioning.driver.accessToken,
				observedEvents,
				roomId: provisioning.roomId,
				syncState,
				sutUserId: provisioning.sut.userId,
				timeoutMs: 45e3
			});
			canaryArtifact = {
				driverEventId: canary.driverEventId,
				reply: canary.reply,
				token: canary.token
			};
			checks.push({
				name: "Matrix canary",
				status: "pass",
				details: buildMatrixReplyDetails("reply", canary.reply).join("\n")
			});
		} catch (error) {
			canaryFailed = true;
			checks.push({
				name: "Matrix canary",
				status: "fail",
				details: formatErrorMessage(error)
			});
		}
		if (!canaryFailed) for (const scenario of scenarios) try {
			const result = await runMatrixQaScenario(scenario, {
				baseUrl: harness.baseUrl,
				canary: canaryArtifact,
				driverAccessToken: provisioning.driver.accessToken,
				driverUserId: provisioning.driver.userId,
				observedEvents,
				observerAccessToken: provisioning.observer.accessToken,
				observerUserId: provisioning.observer.userId,
				restartGateway: async () => {
					if (!gatewayHarness) throw new Error("Matrix restart scenario requires a live gateway");
					await gatewayHarness.gateway.restart();
					await waitForMatrixChannelReady(gatewayHarness.gateway, sutAccountId);
				},
				roomId: provisioning.roomId,
				syncState,
				sutUserId: provisioning.sut.userId,
				timeoutMs: scenario.timeoutMs
			});
			scenarioResults.push({
				artifacts: result.artifacts,
				id: scenario.id,
				title: scenario.title,
				status: "pass",
				details: result.details
			});
		} catch (error) {
			scenarioResults.push({
				id: scenario.id,
				title: scenario.title,
				status: "fail",
				details: formatErrorMessage(error)
			});
		}
	} finally {
		if (gatewayHarness) try {
			await gatewayHarness.stop();
		} catch (error) {
			appendLiveLaneIssue(cleanupErrors, "live gateway cleanup", error);
		}
		try {
			await harness.stop();
		} catch (error) {
			appendLiveLaneIssue(cleanupErrors, "Matrix harness cleanup", error);
		}
	}
	if (cleanupErrors.length > 0) checks.push({
		name: "Matrix cleanup",
		status: "fail",
		details: cleanupErrors.join("\n")
	});
	const finishedAtDate = /* @__PURE__ */ new Date();
	const finishedAt = finishedAtDate.toISOString();
	const reportPath = path.join(outputDir, "matrix-qa-report.md");
	const summaryPath = path.join(outputDir, "matrix-qa-summary.json");
	const observedEventsPath = path.join(outputDir, "matrix-qa-observed-events.json");
	const artifactPaths = {
		observedEvents: observedEventsPath,
		report: reportPath,
		summary: summaryPath
	};
	const report = renderQaMarkdownReport({
		title: "Matrix QA Report",
		startedAt: startedAtDate,
		finishedAt: finishedAtDate,
		checks,
		scenarios: scenarioResults.map((scenario) => ({
			details: scenario.details,
			name: scenario.title,
			status: scenario.status
		})),
		notes: [
			`roomId: ${provisioning.roomId}`,
			`driver: ${provisioning.driver.userId}`,
			`observer: ${provisioning.observer.userId}`,
			`sut: ${provisioning.sut.userId}`,
			`homeserver: ${harness.baseUrl}`,
			`image: ${harness.image}`
		]
	});
	const summary = buildMatrixQaSummary({
		artifactPaths,
		canary: canaryArtifact,
		checks,
		finishedAt,
		harness: {
			baseUrl: harness.baseUrl,
			composeFile: harness.composeFile,
			image: harness.image,
			roomId: provisioning.roomId,
			serverName: harness.serverName
		},
		observedEventCount: observedEvents.length,
		scenarios: scenarioResults,
		startedAt,
		sutAccountId,
		userIds: {
			driver: provisioning.driver.userId,
			observer: provisioning.observer.userId,
			sut: provisioning.sut.userId
		}
	});
	await fs.writeFile(reportPath, `${report}\n`, {
		encoding: "utf8",
		mode: 384
	});
	await fs.writeFile(summaryPath, `${JSON.stringify(summary, null, 2)}\n`, {
		encoding: "utf8",
		mode: 384
	});
	await fs.writeFile(observedEventsPath, `${JSON.stringify(buildObservedEventsArtifact({
		includeContent: includeObservedEventContent,
		observedEvents
	}), null, 2)}\n`, {
		encoding: "utf8",
		mode: 384
	});
	const failedChecks = checks.filter((check) => check.status === "fail" && check.name !== "Matrix cleanup");
	const failedScenarios = scenarioResults.filter((scenario) => scenario.status === "fail");
	if (failedChecks.length > 0 || failedScenarios.length > 0) throw new Error(buildLiveLaneArtifactsError({
		heading: "Matrix QA failed.",
		details: [
			...failedChecks.map((check) => `check ${check.name}: ${check.details ?? "failed"}`),
			...failedScenarios.map((scenario) => `scenario ${scenario.id}: ${scenario.details}`),
			...cleanupErrors.map((error) => `cleanup: ${error}`)
		],
		artifacts: artifactPaths
	}));
	if (cleanupErrors.length > 0) throw new Error(buildLiveLaneArtifactsError({
		heading: "Matrix QA cleanup failed after artifacts were written.",
		details: cleanupErrors,
		artifacts: artifactPaths
	}));
	return {
		observedEventsPath,
		outputDir,
		reportPath,
		scenarios: scenarioResults,
		summaryPath
	};
}
//#endregion
//#region extensions/qa-lab/src/live-transports/matrix/cli.runtime.ts
async function runQaMatrixCommand(opts) {
	const result = await runMatrixQaLive(resolveLiveTransportQaRunOptions(opts));
	printLiveTransportQaArtifacts("Matrix QA", {
		report: result.reportPath,
		summary: result.summaryPath,
		"observed events": result.observedEventsPath
	});
}
//#endregion
export { runQaMatrixCommand };
