// openclaw/dist/docker-runtime-C6coNoAQ.js
import { n as fetchWithSsrFGuard } from "./fetch-guard-DKe4b6qW.js";
import "./ssrf-runtime-DgTiz1oy.js";
import { execFile } from "node:child_process";
import { createServer } from "node:net";
//#region extensions/qa-lab/src/docker-runtime.ts
async function fetchHealthUrl(url) {
	const { response, release } = await fetchWithSsrFGuard({
		url,
		init: { signal: AbortSignal.timeout(2e3) },
		policy: { allowPrivateNetwork: true },
		auditContext: "qa-lab-docker-health-check"
	});
	try {
		return { ok: response.ok };
	} finally {
		await release();
	}
}
function describeError(error) {
	if (error instanceof Error) return error.message;
	if (typeof error === "string") return error;
	return JSON.stringify(error);
}
async function isPortFree(port) {
	return await new Promise((resolve) => {
		const server = createServer();
		server.once("error", () => resolve(false));
		server.listen(port, "127.0.0.1", () => {
			server.close(() => resolve(true));
		});
	});
}
async function findFreePort() {
	return await new Promise((resolve, reject) => {
		const server = createServer();
		server.once("error", reject);
		server.listen(0, () => {
			const address = server.address();
			if (!address || typeof address === "string") {
				server.close();
				reject(/* @__PURE__ */ new Error("failed to find free port"));
				return;
			}
			server.close((error) => {
				if (error) {
					reject(error);
					return;
				}
				resolve(address.port);
			});
		});
	});
}
async function resolveHostPort(preferredPort, pinned) {
	if (pinned || await isPortFree(preferredPort)) return preferredPort;
	return await findFreePort();
}
function trimCommandOutput(output) {
	const trimmed = output.trim();
	if (!trimmed) return "";
	const lines = trimmed.split("\n");
	return lines.length <= 120 ? trimmed : lines.slice(-120).join("\n");
}
async function execCommand(command, args, cwd) {
	return await new Promise((resolve, reject) => {
		execFile(command, args, {
			cwd,
			encoding: "utf8",
			maxBuffer: 10 * 1024 * 1024
		}, (error, stdout, stderr) => {
			if (error) {
				const renderedStdout = trimCommandOutput(stdout);
				const renderedStderr = trimCommandOutput(stderr);
				reject(new Error([
					`Command failed: ${[command, ...args].join(" ")}`,
					renderedStderr ? `stderr:\n${renderedStderr}` : "",
					renderedStdout ? `stdout:\n${renderedStdout}` : ""
				].filter(Boolean).join("\n\n")));
				return;
			}
			resolve({
				stdout,
				stderr
			});
		});
	});
}
async function waitForHealth(url, deps) {
	const timeoutMs = deps.timeoutMs ?? 36e4;
	const pollMs = deps.pollMs ?? 1e3;
	const startMs = Date.now();
	const deadline = startMs + timeoutMs;
	let lastError = null;
	while (Date.now() < deadline) {
		try {
			if ((await deps.fetchImpl(url)).ok) return;
			lastError = /* @__PURE__ */ new Error(`Health check returned non-OK for ${url}`);
		} catch (error) {
			lastError = error;
		}
		await deps.sleepImpl(pollMs);
	}
	const elapsedSec = Math.round((Date.now() - startMs) / 1e3);
	const lines = [
		`${deps.label ?? url} did not become healthy within ${elapsedSec}s (limit ${Math.round(timeoutMs / 1e3)}s).`,
		lastError ? `Last error: ${describeError(lastError)}` : "",
		`Hint: check container logs with \`docker compose -f ${deps.composeFile ?? "<compose-file>"} logs\` and verify the port is not already in use.`
	];
	throw new Error(lines.filter(Boolean).join("\n"));
}
async function isHealthy(url, fetchImpl) {
	try {
		return (await fetchImpl(url)).ok;
	} catch {
		return false;
	}
}
function normalizeDockerServiceStatus(row) {
	const health = row?.Health?.trim();
	if (health) return health;
	const state = row?.State?.trim();
	if (state) return state;
	return "unknown";
}
function parseDockerComposePsRows(stdout) {
	const trimmed = stdout.trim();
	if (!trimmed) return [];
	try {
		const parsed = JSON.parse(trimmed);
		if (Array.isArray(parsed)) return parsed;
		return [parsed];
	} catch {
		return trimmed.split("\n").map((line) => line.trim()).filter(Boolean).map((line) => JSON.parse(line));
	}
}
async function waitForDockerServiceHealth(service, composeFile, repoRoot, runCommand, sleepImpl, timeoutMs = 36e4, pollMs = 1e3) {
	const startMs = Date.now();
	const deadline = startMs + timeoutMs;
	let lastStatus = "unknown";
	while (Date.now() < deadline) {
		try {
			const { stdout } = await runCommand("docker", [
				"compose",
				"-f",
				composeFile,
				"ps",
				"--format",
				"json",
				service
			], repoRoot);
			const row = parseDockerComposePsRows(stdout)[0];
			lastStatus = normalizeDockerServiceStatus(row);
			if (lastStatus === "healthy" || lastStatus === "running") return;
		} catch (error) {
			lastStatus = describeError(error);
		}
		await sleepImpl(pollMs);
	}
	const elapsedSec = Math.round((Date.now() - startMs) / 1e3);
	throw new Error([
		`${service} did not become healthy within ${elapsedSec}s (limit ${Math.round(timeoutMs / 1e3)}s).`,
		`Last status: ${lastStatus}`,
		`Hint: check container logs with \`docker compose -f ${composeFile} logs ${service}\`.`
	].join("\n"));
}
async function resolveComposeServiceUrl(service, port, composeFile, repoRoot, runCommand, fetchImpl) {
	const { stdout: containerStdout } = await runCommand("docker", [
		"compose",
		"-f",
		composeFile,
		"ps",
		"-q",
		service
	], repoRoot);
	const containerId = containerStdout.trim();
	if (!containerId) return null;
	const { stdout: ipStdout } = await runCommand("docker", [
		"inspect",
		"--format",
		"{{range .NetworkSettings.Networks}}{{.IPAddress}}{{end}}",
		containerId
	], repoRoot);
	const ip = ipStdout.trim();
	if (!ip) return null;
	const baseUrl = `http://${ip}:${port}/`;
	if (!fetchImpl) return baseUrl;
	return await isHealthy(`${baseUrl}healthz`, fetchImpl) ? baseUrl : null;
}
//#endregion
export { waitForDockerServiceHealth as a, resolveHostPort as i, fetchHealthUrl as n, waitForHealth as o, resolveComposeServiceUrl as r, execCommand as t };
