// openclaw/dist/extensionAPI.js
import { n as DEFAULT_MODEL, r as DEFAULT_PROVIDER } from "./defaults-BckzHLk4.js";
import { d as ensureAgentWorkspace } from "./workspace-_11V2h5W.js";
import { a as resolveAgentDir, m as resolveAgentWorkspaceDir } from "./agent-scope-C7LiYj5W.js";
import { x as resolveThinkingDefault } from "./model-selection-DP92S76A.js";
import { i as resolveSessionFilePath, u as resolveStorePath } from "./paths-CrHqfnMp.js";
import { n as resolveAgentIdentity } from "./identity-LLtDOuch.js";
import { o as saveSessionStore } from "./store-Cw5RrVCw.js";
import "./sessions-BWz-crFz.js";
import { t as loadSessionStore } from "./store-load-DsLWM8wp.js";
import { t as runEmbeddedPiAgent } from "./pi-embedded-runner-DXh-tqVs.js";
import { t as resolveAgentTimeoutMs } from "./timeout-Be5OnPsr.js";
import "./pi-embedded-HwLIt9IG.js";
//#region src/extensionAPI.ts
if (process.env.VITEST !== "true" && process.env.OPENCLAW_SUPPRESS_EXTENSION_API_WARNING !== "1") process.emitWarning("openclaw/extension-api is deprecated. Migrate to api.runtime.agent.* or focused openclaw/plugin-sdk/<subpath> imports. See https://docs.openclaw.ai/plugins/sdk-migration", {
	code: "OPENCLAW_EXTENSION_API_DEPRECATED",
	detail: "This compatibility bridge is temporary. Bundled plugins should use the injected plugin runtime instead of importing host-side agent helpers directly. Migration guide: https://docs.openclaw.ai/plugins/sdk-migration"
});
//#endregion
export { DEFAULT_MODEL, DEFAULT_PROVIDER, ensureAgentWorkspace, loadSessionStore, resolveAgentDir, resolveAgentIdentity, resolveAgentTimeoutMs, resolveAgentWorkspaceDir, resolveSessionFilePath, resolveStorePath, resolveThinkingDefault, runEmbeddedPiAgent, saveSessionStore };
