// openclaw/dist/gateway-child-Du_AnSFD.js
import { i as formatErrorMessage } from "./errors-D8p6rxH8.js";
import { n as resolvePreferredOpenClawTmpDir } from "./tmp-openclaw-dir-DgmFzeYV.js";
import { a as upsertAuthProfile } from "./profiles-DT2g1T-s.js";
import { n as fetchWithSsrFGuard } from "./fetch-guard-DKe4b6qW.js";
import "./temp-path-B39Iubqg.js";
import "./error-runtime-Da0X1TOQ.js";
import { t as applyAuthProfileConfig } from "./provider-auth-helpers-BYWHtohY.js";
import "./ssrf-runtime-DgTiz1oy.js";
import { n as callGatewayFromCli } from "./gateway-rpc-BGzygo8Y.js";
import { n as validateAnthropicSetupToken } from "./provider-auth-token-DAwigjn4.js";
import "./provider-auth-CzzkrCf0.js";
import "./runtime-api-pvMSXsR4.js";
import { i as readQaBootstrapScenarioCatalog, l as readQaScenarioPackMarkdown, t as DEFAULT_QA_AGENT_IDENTITY_MARKDOWN } from "./scenario-catalog-DQ-BbL_S.js";
import { t as closeQaHttpServer } from "./bus-server-2ab_NN8Y.js";
import { a as splitQaModelRef } from "./model-selection-ClMg2tje.js";
import { n as buildQaGatewayConfig } from "./qa-gateway-config-DunCiJBX.js";
import { createWriteStream, existsSync } from "node:fs";
import path from "node:path";
import { spawn } from "node:child_process";
import os from "node:os";
import fs$1 from "node:fs/promises";
import { randomUUID } from "node:crypto";
import net from "node:net";
import { setTimeout } from "node:timers/promises";
import { createServer as createServer$1 } from "node:http";
//#region extensions/qa-lab/src/qa-agent-bootstrap.ts
function readQaAgentIdentityMarkdown() {
	return readQaBootstrapScenarioCatalog().agentIdentityMarkdown || DEFAULT_QA_AGENT_IDENTITY_MARKDOWN;
}
function buildQaScenarioPlanMarkdown() {
	const catalog = readQaBootstrapScenarioCatalog();
	const lines = ["# QA Scenario Plan", ""];
	for (const scenario of catalog.scenarios) {
		lines.push(`## ${scenario.title}`);
		lines.push("");
		lines.push(`- id: ${scenario.id}`);
		lines.push(`- surface: ${scenario.surface}`);
		lines.push(`- objective: ${scenario.objective}`);
		if (scenario.execution?.summary) lines.push(`- execution: ${scenario.execution.summary}`);
		lines.push("- success criteria:");
		for (const criterion of scenario.successCriteria) lines.push(`  - ${criterion}`);
		if (scenario.docsRefs?.length) {
			lines.push("- docs:");
			for (const ref of scenario.docsRefs) lines.push(`  - ${ref}`);
		}
		if (scenario.codeRefs?.length) {
			lines.push("- code:");
			for (const ref of scenario.codeRefs) lines.push(`  - ${ref}`);
		}
		lines.push("");
	}
	return lines.join("\n");
}
//#endregion
//#region extensions/qa-lab/src/qa-agent-workspace.ts
async function seedQaAgentWorkspace(params) {
	const catalog = readQaBootstrapScenarioCatalog();
	await fs$1.mkdir(params.workspaceDir, { recursive: true });
	const kickoffTask = catalog.kickoffTask || "QA mission unavailable.";
	const files = new Map([
		["IDENTITY.md", readQaAgentIdentityMarkdown()],
		["QA_KICKOFF_TASK.md", kickoffTask],
		["QA_SCENARIO_PLAN.md", buildQaScenarioPlanMarkdown()],
		["QA_SCENARIOS.md", readQaScenarioPackMarkdown()]
	]);
	if (params.repoRoot) files.set("README.md", `# QA Workspace

- repo: ./repo/
- kickoff: ./QA_KICKOFF_TASK.md
- scenario plan: ./QA_SCENARIO_PLAN.md
- scenario pack: ./QA_SCENARIOS.md
- identity: ./IDENTITY.md

The mounted repo source should be available read-only under \`./repo/\`.
`);
	await Promise.all([...files.entries()].map(async ([name, body]) => {
		await fs$1.writeFile(path.join(params.workspaceDir, name), `${body.trim()}\n`, "utf8");
	}));
	if (params.repoRoot) {
		const repoLinkPath = path.join(params.workspaceDir, "repo");
		await fs$1.rm(repoLinkPath, {
			force: true,
			recursive: true
		});
		await fs$1.symlink(params.repoRoot, repoLinkPath, "dir");
	}
}
//#endregion
//#region extensions/qa-lab/src/mock-openai-server.ts
const TINY_PNG_BASE64 = "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8/x8AAwMCAO7Z0nQAAAAASUVORK5CYII=";
let subagentFanoutPhase = 0;
function readBody(req) {
	return new Promise((resolve, reject) => {
		const chunks = [];
		req.on("data", (chunk) => chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk)));
		req.on("end", () => resolve(Buffer.concat(chunks).toString("utf8")));
		req.on("error", reject);
	});
}
function writeJson(res, status, body) {
	const text = JSON.stringify(body);
	res.writeHead(status, {
		"content-type": "application/json; charset=utf-8",
		"content-length": Buffer.byteLength(text),
		"cache-control": "no-store"
	});
	res.end(text);
}
function writeSse(res, events) {
	const body = `${events.map((event) => `data: ${JSON.stringify(event)}\n\n`).join("")}data: [DONE]\n\n`;
	res.writeHead(200, {
		"content-type": "text/event-stream",
		"cache-control": "no-store",
		connection: "keep-alive",
		"content-length": Buffer.byteLength(body)
	});
	res.end(body);
}
function countApproxTokens(text) {
	const trimmed = text.trim();
	if (!trimmed) return 0;
	return Math.max(1, Math.ceil(trimmed.length / 4));
}
function extractEmbeddingInputTexts(input) {
	if (typeof input === "string") return [input];
	if (Array.isArray(input)) return input.flatMap((entry) => extractEmbeddingInputTexts(entry));
	if (input && typeof input === "object" && typeof input.text === "string") return [input.text];
	return [];
}
function buildDeterministicEmbedding(text, dimensions = 16) {
	const values = Array.from({ length: dimensions }, () => 0);
	for (let index = 0; index < text.length; index += 1) values[index % dimensions] += text.charCodeAt(index) / 255;
	const magnitude = Math.hypot(...values) || 1;
	return values.map((value) => Number((value / magnitude).toFixed(8)));
}
function extractLastUserText(input) {
	for (let index = input.length - 1; index >= 0; index -= 1) {
		const item = input[index];
		if (item.role !== "user" || !Array.isArray(item.content)) continue;
		const text = extractInputText(item.content);
		if (text) return text;
	}
	return "";
}
function findLastUserIndex(input) {
	for (let index = input.length - 1; index >= 0; index -= 1) {
		const item = input[index];
		if (item.role === "user" && Array.isArray(item.content)) return index;
	}
	return -1;
}
function extractToolOutput(input) {
	const lastUserIndex = findLastUserIndex(input);
	for (let index = input.length - 1; index > lastUserIndex; index -= 1) {
		const item = input[index];
		if (item.type === "function_call_output" && typeof item.output === "string" && item.output) return item.output;
	}
	return "";
}
function extractInputText(content) {
	return content.filter((entry) => !!entry && typeof entry === "object" && entry.type === "input_text" && typeof entry.text === "string").map((entry) => entry.text).join("\n").trim();
}
function extractAllUserTexts(input) {
	const texts = [];
	for (const item of input) {
		if (item.role !== "user" || !Array.isArray(item.content)) continue;
		const text = extractInputText(item.content);
		if (text) texts.push(text);
	}
	return texts;
}
function extractAllInputTexts(input) {
	const texts = [];
	for (const item of input) {
		if (typeof item.output === "string" && item.output.trim()) texts.push(item.output.trim());
		if (!Array.isArray(item.content)) continue;
		const text = extractInputText(item.content);
		if (text) texts.push(text);
	}
	return texts.join("\n");
}
function countImageInputs(input) {
	let count = 0;
	for (const item of input) {
		if (!Array.isArray(item.content)) continue;
		for (const entry of item.content) if (entry && typeof entry === "object" && entry.type === "input_image") count += 1;
	}
	return count;
}
function parseToolOutputJson(toolOutput) {
	if (!toolOutput.trim()) return null;
	try {
		return JSON.parse(toolOutput);
	} catch {
		return null;
	}
}
function normalizePromptPathCandidate(candidate) {
	const trimmed = candidate.trim().replace(/^`+|`+$/g, "");
	if (!trimmed) return null;
	const normalized = trimmed.replace(/^\.\//, "");
	if (normalized.includes("/") || /\.(?:md|json|ts|tsx|js|mjs|cjs|txt|yaml|yml)$/i.test(normalized)) return normalized;
	return null;
}
function readTargetFromPrompt(prompt) {
	const backtickedMatches = Array.from(prompt.matchAll(/`([^`]+)`/g)).map((match) => normalizePromptPathCandidate(match[1] ?? "")).filter((value) => !!value);
	if (backtickedMatches.length > 0) return backtickedMatches[0];
	const quotedMatches = Array.from(prompt.matchAll(/"([^"]+)"/g)).map((match) => normalizePromptPathCandidate(match[1] ?? "")).filter((value) => !!value);
	if (quotedMatches.length > 0) return quotedMatches[0];
	const repoScoped = /\b(?:repo\/[^\s`",)]+|QA_[A-Z_]+\.md)\b/.exec(prompt)?.[0]?.trim();
	if (repoScoped) return repoScoped;
	if (/\bdocs?\b/i.test(prompt)) return "repo/docs/help/testing.md";
	if (/\bscenario|kickoff|qa\b/i.test(prompt)) return "QA_KICKOFF_TASK.md";
	return "repo/package.json";
}
function buildToolCallEventsWithArgs(name, args) {
	const callId = `call_mock_${name}_1`;
	const serialized = JSON.stringify(args);
	return [
		{
			type: "response.output_item.added",
			item: {
				type: "function_call",
				id: `fc_mock_${name}_1`,
				call_id: callId,
				name,
				arguments: ""
			}
		},
		{
			type: "response.function_call_arguments.delta",
			delta: serialized
		},
		{
			type: "response.output_item.done",
			item: {
				type: "function_call",
				id: `fc_mock_${name}_1`,
				call_id: callId,
				name,
				arguments: serialized
			}
		},
		{
			type: "response.completed",
			response: {
				id: `resp_mock_${name}_1`,
				status: "completed",
				output: [{
					type: "function_call",
					id: `fc_mock_${name}_1`,
					call_id: callId,
					name,
					arguments: serialized
				}],
				usage: {
					input_tokens: 64,
					output_tokens: 16,
					total_tokens: 80
				}
			}
		}
	];
}
function extractRememberedFact(userTexts) {
	for (const text of userTexts) {
		const qaCanaryMatch = /\bqa canary code is\s+([A-Za-z0-9-]+)/i.exec(text);
		if (qaCanaryMatch?.[1]) return qaCanaryMatch[1];
	}
	for (const text of userTexts) {
		const match = /remember(?: this fact for later)?:\s*([A-Za-z0-9-]+)/i.exec(text);
		if (match?.[1]) return match[1];
	}
	return null;
}
function extractOrbitCode(text) {
	return /\bORBIT-\d+\b/i.exec(text)?.[0]?.toUpperCase() ?? null;
}
function extractLastCapture(text, pattern) {
	let lastMatch = null;
	const flags = pattern.flags.includes("g") ? pattern.flags : `${pattern.flags}g`;
	const globalPattern = new RegExp(pattern.source, flags);
	for (let match = globalPattern.exec(text); match; match = globalPattern.exec(text)) lastMatch = match;
	return lastMatch?.[1]?.trim() || null;
}
function extractExactReplyDirective(text) {
	const colonMatch = extractLastCapture(text, /reply(?: with)? exactly:\s*([^\n]+)/i);
	if (colonMatch) return colonMatch;
	return extractLastCapture(text, /reply(?: with)? exactly\s+`([^`]+)`/i);
}
function extractExactMarkerDirective(text) {
	const backtickedMatch = extractLastCapture(text, /exact marker:\s*`([^`]+)`/i);
	if (backtickedMatch) return backtickedMatch;
	return extractLastCapture(text, /exact marker:\s*([^\s`.,;:!?]+(?:-[^\s`.,;:!?]+)*)/i);
}
function isHeartbeatPrompt(text) {
	return /Read HEARTBEAT\.md if it exists/i.test(text);
}
function buildAssistantText(input, body) {
	const prompt = extractLastUserText(input);
	const toolOutput = extractToolOutput(input);
	const toolJson = parseToolOutputJson(toolOutput);
	const userTexts = extractAllUserTexts(input);
	const allInputText = extractAllInputTexts(input);
	const rememberedFact = extractRememberedFact(userTexts);
	const model = typeof body.model === "string" ? body.model : "";
	const orbitCode = extractOrbitCode(typeof toolJson?.text === "string" ? toolJson.text : Array.isArray(toolJson?.results) ? JSON.stringify(toolJson.results) : toolOutput);
	const mediaPath = /MEDIA:([^\n]+)/.exec(toolOutput)?.[1]?.trim();
	const exactReplyDirective = extractExactReplyDirective(allInputText);
	const exactMarkerDirective = extractExactMarkerDirective(allInputText);
	const imageInputCount = countImageInputs(input);
	if (/what was the qa canary code/i.test(prompt) && rememberedFact) return `Protocol note: the QA canary code was ${rememberedFact}.`;
	if (/remember this fact/i.test(prompt) && exactReplyDirective) return exactReplyDirective;
	if (/remember this fact/i.test(prompt) && rememberedFact) return `Protocol note: acknowledged. I will remember ${rememberedFact}.`;
	if (/memory unavailable check/i.test(prompt)) return "Protocol note: I checked the available runtime context but could not confirm the hidden memory-only fact, so I will not guess.";
	if (isHeartbeatPrompt(prompt)) return "HEARTBEAT_OK";
	if (/\bmarker\b/i.test(prompt) && exactReplyDirective) return exactReplyDirective;
	if (/\bmarker\b/i.test(prompt) && exactMarkerDirective) return exactMarkerDirective;
	if (/visible skill marker/i.test(prompt)) return "VISIBLE-SKILL-OK";
	if (/hot install marker/i.test(prompt)) return "HOT-INSTALL-OK";
	if (/memory tools check/i.test(prompt) && orbitCode) return `Protocol note: I checked memory and the project codename is ${orbitCode}.`;
	if (/tool continuity check/i.test(prompt) && toolOutput) return `Protocol note: model switch handoff confirmed on ${model || "the requested model"}. QA mission from QA_KICKOFF_TASK.md still applies: understand this OpenClaw repo from source + docs before acting.`;
	if (/session memory ranking check/i.test(prompt) && orbitCode) return `Protocol note: I checked memory and the current Project Nebula codename is ${orbitCode}.`;
	if (/thread memory check/i.test(prompt) && orbitCode) return `Protocol note: I checked memory in-thread and the hidden thread codename is ${orbitCode}.`;
	if (/switch(?:ing)? models?/i.test(prompt)) return `Protocol note: model switch acknowledged. Continuing on ${model || "the requested model"}.`;
	if (/(image generation check|capability flip image check)/i.test(prompt) && mediaPath) return `Protocol note: generated the QA lighthouse image successfully.\nMEDIA:${mediaPath}`;
	if (/roundtrip image inspection check/i.test(prompt) && imageInputCount > 0) return "Protocol note: the generated attachment shows the same QA lighthouse scene from the previous step.";
	if (/image understanding check/i.test(prompt) && imageInputCount > 0) return "Protocol note: the attached image is split horizontally, with red on top and blue on the bottom.";
	if (/interrupted by a gateway reload/i.test(prompt) && /subagent recovery worker/i.test(allInputText)) return "RECOVERED-SUBAGENT-OK";
	if (/subagent recovery worker/i.test(prompt)) return "RECOVERED-SUBAGENT-OK";
	if (/fanout worker alpha/i.test(prompt)) return "ALPHA-OK";
	if (/fanout worker beta/i.test(prompt)) return "BETA-OK";
	if (/subagent fanout synthesis check/i.test(prompt) && toolOutput && subagentFanoutPhase >= 2) return "Protocol note: delegated fanout complete. Alpha=ALPHA-OK. Beta=BETA-OK.";
	if (toolOutput && (/\bdelegate\b/i.test(prompt) || /subagent handoff/i.test(prompt))) return `Delegated task:\n- Inspect the QA workspace via a bounded subagent.\nResult:\n- ${toolOutput.replace(/\s+/g, " ").trim() || "no delegated output"}\nEvidence:\n- The child result was folded back into the main thread exactly once.`;
	if (toolOutput && /worked, failed, blocked|worked\/failed\/blocked|follow-up/i.test(prompt)) return `Worked:\n- Read seeded QA material.\n- Expanded the report structure.\nFailed:\n- None observed in mock mode.\nBlocked:\n- No live provider evidence in this lane.\nFollow-up:\n- Re-run with a real model for qualitative coverage.`;
	if (toolOutput && /lobster invaders/i.test(prompt)) {
		if (toolOutput.includes("QA mission") || toolOutput.includes("Testing")) return "";
		return `Protocol note: Lobster Invaders built at lobster-invaders.html.`;
	}
	if (toolOutput && /compaction retry mutating tool check/i.test(prompt)) {
		if (toolOutput.includes("Replay safety: unsafe after write.") || /compaction-retry-summary\.txt/i.test(toolOutput) || /successfully (?:wrote|replaced)/i.test(toolOutput) || /\bwrote\b.*\bcompaction-retry-summary\.txt\b/i.test(toolOutput)) return "Protocol note: replay unsafe after write.";
		return "";
	}
	if (toolOutput) return `Protocol note: I reviewed the requested material. Evidence snippet: ${toolOutput.replace(/\s+/g, " ").trim().slice(0, 220) || "no content"}`;
	if (prompt) return `Protocol note: acknowledged. Continue with the QA scenario plan and report worked, failed, and blocked items.`;
	return "Protocol note: mock OpenAI server ready.";
}
function buildToolCallEvents(prompt) {
	return buildToolCallEventsWithArgs("read", { path: readTargetFromPrompt(prompt) });
}
function extractPlannedToolName(events) {
	for (const event of events) {
		if (event.type !== "response.output_item.done") continue;
		const item = event.item;
		if (item.type === "function_call" && typeof item.name === "string") return item.name;
	}
}
function buildAssistantEvents(text) {
	const outputItem = {
		type: "message",
		id: "msg_mock_1",
		role: "assistant",
		status: "completed",
		content: [{
			type: "output_text",
			text,
			annotations: []
		}]
	};
	return [
		{
			type: "response.output_item.added",
			item: {
				type: "message",
				id: "msg_mock_1",
				role: "assistant",
				content: [],
				status: "in_progress"
			}
		},
		{
			type: "response.output_item.done",
			item: outputItem
		},
		{
			type: "response.completed",
			response: {
				id: "resp_mock_msg_1",
				status: "completed",
				output: [outputItem],
				usage: {
					input_tokens: 64,
					output_tokens: 24,
					total_tokens: 88
				}
			}
		}
	];
}
async function buildResponsesPayload(body) {
	const input = Array.isArray(body.input) ? body.input : [];
	const prompt = extractLastUserText(input);
	const toolOutput = extractToolOutput(input);
	const toolJson = parseToolOutputJson(toolOutput);
	const isGroupChat = extractAllInputTexts(input).includes("\"is_group_chat\": true");
	const isBaselineUnmentionedChannelChatter = /\bno bot ping here\b/i.test(prompt);
	if (isHeartbeatPrompt(prompt)) return buildAssistantEvents("HEARTBEAT_OK");
	if (/lobster invaders/i.test(prompt)) {
		if (!toolOutput) return buildToolCallEventsWithArgs("read", { path: "QA_KICKOFF_TASK.md" });
		if (toolOutput.includes("QA mission") || toolOutput.includes("Testing")) return buildToolCallEventsWithArgs("write", {
			path: "lobster-invaders.html",
			content: `<!doctype html>
<html lang="en">
  <head><meta charset="utf-8" /><title>Lobster Invaders</title></head>
  <body><h1>Lobster Invaders</h1><p>Tiny playable stub.</p></body>
</html>`
		});
	}
	if (/compaction retry mutating tool check/i.test(prompt)) {
		if (!toolOutput) return buildToolCallEventsWithArgs("read", { path: "COMPACTION_RETRY_CONTEXT.md" });
		if (toolOutput.includes("compaction retry evidence")) return buildToolCallEventsWithArgs("write", {
			path: "compaction-retry-summary.txt",
			content: "Replay safety: unsafe after write.\n"
		});
	}
	if (/memory tools check/i.test(prompt)) {
		if (!toolOutput) return buildToolCallEventsWithArgs("memory_search", {
			query: "project codename ORBIT-9",
			maxResults: 3
		});
		const first = (Array.isArray(toolJson?.results) ? toolJson.results : [])[0];
		if (typeof first?.path === "string" && (typeof first.startLine === "number" || typeof first.endLine === "number")) {
			const from = typeof first.startLine === "number" ? Math.max(1, first.startLine) : typeof first.endLine === "number" ? Math.max(1, first.endLine) : 1;
			return buildToolCallEventsWithArgs("memory_get", {
				path: first.path,
				from,
				lines: 4
			});
		}
	}
	if (/session memory ranking check/i.test(prompt)) {
		if (!toolOutput) return buildToolCallEventsWithArgs("memory_search", {
			query: "current Project Nebula codename ORBIT-10",
			maxResults: 3
		});
		const first = (Array.isArray(toolJson?.results) ? toolJson.results : [])[0];
		const firstPath = typeof first?.path === "string" ? first.path : void 0;
		if (first?.source === "sessions" || firstPath?.startsWith("sessions/")) return buildAssistantEvents("Protocol note: I checked memory and the current Project Nebula codename is ORBIT-10.");
		if (typeof first?.path === "string" && (typeof first.startLine === "number" || typeof first.endLine === "number")) {
			const from = typeof first.startLine === "number" ? Math.max(1, first.startLine) : typeof first.endLine === "number" ? Math.max(1, first.endLine) : 1;
			return buildToolCallEventsWithArgs("memory_get", {
				path: first.path,
				from,
				lines: 4
			});
		}
	}
	if (/thread memory check/i.test(prompt)) {
		if (!toolOutput) return buildToolCallEventsWithArgs("memory_search", {
			query: "hidden thread codename ORBIT-22",
			maxResults: 3
		});
		const first = (Array.isArray(toolJson?.results) ? toolJson.results : [])[0];
		if (typeof first?.path === "string" && (typeof first.startLine === "number" || typeof first.endLine === "number")) {
			const from = typeof first.startLine === "number" ? Math.max(1, first.startLine) : typeof first.endLine === "number" ? Math.max(1, first.endLine) : 1;
			return buildToolCallEventsWithArgs("memory_get", {
				path: first.path,
				from,
				lines: 4
			});
		}
	}
	if (/(image generation check|capability flip image check)/i.test(prompt) && !toolOutput) return buildToolCallEventsWithArgs("image_generate", {
		prompt: "A QA lighthouse on a dark sea with a tiny protocol droid silhouette.",
		filename: "qa-lighthouse.png",
		size: "1024x1024"
	});
	if (/subagent fanout synthesis check/i.test(prompt)) {
		if (!toolOutput && subagentFanoutPhase === 0) {
			subagentFanoutPhase = 1;
			return buildToolCallEventsWithArgs("sessions_spawn", {
				task: "Fanout worker alpha: inspect the QA workspace and finish with exactly ALPHA-OK.",
				label: "qa-fanout-alpha",
				thread: false
			});
		}
		if (toolOutput && subagentFanoutPhase === 1) {
			subagentFanoutPhase = 2;
			return buildToolCallEventsWithArgs("sessions_spawn", {
				task: "Fanout worker beta: inspect the QA workspace and finish with exactly BETA-OK.",
				label: "qa-fanout-beta",
				thread: false
			});
		}
	}
	if (/tool continuity check/i.test(prompt) && !toolOutput) return buildToolCallEventsWithArgs("read", { path: "QA_KICKOFF_TASK.md" });
	if ((/\bdelegate\b/i.test(prompt) || /subagent handoff/i.test(prompt)) && !toolOutput) return buildToolCallEventsWithArgs("sessions_spawn", {
		task: "Inspect the QA workspace and return one concise protocol note.",
		label: "qa-sidecar",
		thread: false
	});
	if (/(worked, failed, blocked|worked\/failed\/blocked|source and docs)/i.test(prompt) && !toolOutput) return buildToolCallEventsWithArgs("read", { path: "QA_SCENARIO_PLAN.md" });
	if (!toolOutput && /\b(read|inspect|repo|docs|scenario|kickoff)\b/i.test(prompt)) return buildToolCallEvents(prompt);
	if (/visible skill marker/i.test(prompt) && !toolOutput) return buildAssistantEvents("VISIBLE-SKILL-OK");
	if (/hot install marker/i.test(prompt) && !toolOutput) return buildAssistantEvents("HOT-INSTALL-OK");
	if (isGroupChat && isBaselineUnmentionedChannelChatter && !toolOutput) return buildAssistantEvents("NO_REPLY");
	if (/subagent recovery worker/i.test(prompt) && !/interrupted by a gateway reload/i.test(prompt)) await setTimeout(6e4);
	return buildAssistantEvents(buildAssistantText(input, body));
}
async function startQaMockOpenAiServer(params) {
	const host = params?.host ?? "127.0.0.1";
	subagentFanoutPhase = 0;
	let lastRequest = null;
	const requests = [];
	const imageGenerationRequests = [];
	const server = createServer$1(async (req, res) => {
		const url = new URL(req.url ?? "/", "http://127.0.0.1");
		if (req.method === "GET" && (url.pathname === "/healthz" || url.pathname === "/readyz")) {
			writeJson(res, 200, {
				ok: true,
				status: "live"
			});
			return;
		}
		if (req.method === "GET" && url.pathname === "/v1/models") {
			writeJson(res, 200, { data: [
				{
					id: "gpt-5.4",
					object: "model"
				},
				{
					id: "gpt-5.4-alt",
					object: "model"
				},
				{
					id: "gpt-image-1",
					object: "model"
				},
				{
					id: "text-embedding-3-small",
					object: "model"
				}
			] });
			return;
		}
		if (req.method === "GET" && url.pathname === "/debug/last-request") {
			writeJson(res, 200, lastRequest ?? {
				ok: false,
				error: "no request recorded"
			});
			return;
		}
		if (req.method === "GET" && url.pathname === "/debug/requests") {
			writeJson(res, 200, requests);
			return;
		}
		if (req.method === "GET" && url.pathname === "/debug/image-generations") {
			writeJson(res, 200, imageGenerationRequests);
			return;
		}
		if (req.method === "POST" && url.pathname === "/v1/images/generations") {
			const raw = await readBody(req);
			const body = raw ? JSON.parse(raw) : {};
			imageGenerationRequests.push(body);
			if (imageGenerationRequests.length > 20) imageGenerationRequests.splice(0, imageGenerationRequests.length - 20);
			writeJson(res, 200, { data: [{
				b64_json: TINY_PNG_BASE64,
				revised_prompt: "A QA lighthouse with protocol droid silhouette."
			}] });
			return;
		}
		if (req.method === "POST" && url.pathname === "/v1/embeddings") {
			const raw = await readBody(req);
			const body = raw ? JSON.parse(raw) : {};
			const inputs = extractEmbeddingInputTexts(body.input);
			writeJson(res, 200, {
				object: "list",
				data: inputs.map((text, index) => ({
					object: "embedding",
					index,
					embedding: buildDeterministicEmbedding(text)
				})),
				model: typeof body.model === "string" && body.model.trim() ? body.model : "text-embedding-3-small",
				usage: {
					prompt_tokens: inputs.reduce((sum, text) => sum + countApproxTokens(text), 0),
					total_tokens: inputs.reduce((sum, text) => sum + countApproxTokens(text), 0)
				}
			});
			return;
		}
		if (req.method === "POST" && url.pathname === "/v1/responses") {
			const raw = await readBody(req);
			const body = raw ? JSON.parse(raw) : {};
			const input = Array.isArray(body.input) ? body.input : [];
			const events = await buildResponsesPayload(body);
			lastRequest = {
				raw,
				body,
				prompt: extractLastUserText(input),
				allInputText: extractAllInputTexts(input),
				toolOutput: extractToolOutput(input),
				model: typeof body.model === "string" ? body.model : "",
				imageInputCount: countImageInputs(input),
				plannedToolName: extractPlannedToolName(events)
			};
			requests.push(lastRequest);
			if (requests.length > 50) requests.splice(0, requests.length - 50);
			if (body.stream === false) {
				const completion = events.at(-1);
				if (!completion || completion.type !== "response.completed") {
					writeJson(res, 500, { error: "mock completion failed" });
					return;
				}
				writeJson(res, 200, completion.response);
				return;
			}
			writeSse(res, events);
			return;
		}
		writeJson(res, 404, { error: "not found" });
	});
	await new Promise((resolve, reject) => {
		server.once("error", reject);
		server.listen(params?.port ?? 0, host, () => resolve());
	});
	const address = server.address();
	if (!address || typeof address === "string") throw new Error("qa mock openai failed to bind");
	return {
		baseUrl: `http://${host}:${address.port}`,
		async stop() {
			await closeQaHttpServer(server);
		}
	};
}
//#endregion
//#region extensions/qa-lab/src/gateway-rpc-client.ts
function formatQaGatewayRpcError(error, logs) {
	const details = formatErrorMessage(error);
	return /* @__PURE__ */ new Error(`${details}\nGateway logs:\n${logs()}`);
}
let qaGatewayRpcQueue = Promise.resolve();
async function runQueuedQaGatewayRpc(task) {
	const run = qaGatewayRpcQueue.then(task, task);
	qaGatewayRpcQueue = run.then(() => void 0, () => void 0);
	return await run;
}
async function startQaGatewayRpcClient(params) {
	const wrapError = (error) => formatQaGatewayRpcError(error, params.logs);
	let stopped = false;
	return {
		async request(method, rpcParams, opts) {
			if (stopped) throw wrapError(/* @__PURE__ */ new Error("gateway rpc client already stopped"));
			try {
				return await runQueuedQaGatewayRpc(async () => await callGatewayFromCli(method, {
					url: params.wsUrl,
					token: params.token,
					timeout: String(opts?.timeoutMs ?? 2e4),
					expectFinal: opts?.expectFinal,
					json: true
				}, rpcParams ?? {}, {
					expectFinal: opts?.expectFinal,
					progress: false
				}));
			} catch (error) {
				throw wrapError(error);
			}
		},
		async stop() {
			stopped = true;
		}
	};
}
//#endregion
//#region extensions/qa-lab/src/gateway-child.ts
const QA_LIVE_ENV_ALIASES = Object.freeze([
	{
		liveVar: "OPENCLAW_LIVE_OPENAI_KEY",
		providerVar: "OPENAI_API_KEY"
	},
	{
		liveVar: "OPENCLAW_LIVE_ANTHROPIC_KEY",
		providerVar: "ANTHROPIC_API_KEY"
	},
	{
		liveVar: "OPENCLAW_LIVE_GEMINI_KEY",
		providerVar: "GEMINI_API_KEY"
	}
]);
const QA_MOCK_BLOCKED_ENV_VARS = Object.freeze([
	"ANTHROPIC_API_KEY",
	"ANTHROPIC_OAUTH_TOKEN",
	"AWS_ACCESS_KEY_ID",
	"AWS_BEARER_TOKEN_BEDROCK",
	"AWS_REGION",
	"AWS_SECRET_ACCESS_KEY",
	"AWS_SESSION_TOKEN",
	"GEMINI_API_KEY",
	"GEMINI_API_KEYS",
	"GOOGLE_API_KEY",
	"MISTRAL_API_KEY",
	"OPENAI_API_KEY",
	"OPENAI_API_KEYS",
	"OPENAI_BASE_URL",
	"CODEX_HOME",
	"OPENCLAW_LIVE_ANTHROPIC_KEY",
	"OPENCLAW_LIVE_ANTHROPIC_KEYS",
	"OPENCLAW_LIVE_GEMINI_KEY",
	"OPENCLAW_LIVE_OPENAI_KEY",
	"VOYAGE_API_KEY"
]);
const QA_MOCK_BLOCKED_ENV_KEY_PATTERNS = Object.freeze([
	/^DISCORD_/i,
	/^TELEGRAM_/i,
	/^SLACK_/i,
	/^MATRIX_/i,
	/^SIGNAL_/i,
	/^WHATSAPP_/i,
	/^IMESSAGE_/i,
	/^ZALO/i,
	/^TWILIO_/i,
	/^PLIVO_/i,
	/^NGROK_/i
]);
const QA_LIVE_PROVIDER_CONFIG_PATH_ENV = "OPENCLAW_QA_LIVE_PROVIDER_CONFIG_PATH";
const QA_LIVE_ANTHROPIC_SETUP_TOKEN_ENV = "OPENCLAW_QA_LIVE_ANTHROPIC_SETUP_TOKEN";
const QA_LIVE_SETUP_TOKEN_VALUE_ENV = "OPENCLAW_LIVE_SETUP_TOKEN_VALUE";
const QA_LIVE_ANTHROPIC_SETUP_TOKEN_PROFILE_ENV = "OPENCLAW_QA_LIVE_ANTHROPIC_SETUP_TOKEN_PROFILE";
const QA_LIVE_ANTHROPIC_SETUP_TOKEN_PROFILE_ID = "anthropic:qa-setup-token";
const QA_OPENAI_PLUGIN_ID = "openai";
const QA_LIVE_CLI_BACKEND_PRESERVE_ENV = "OPENCLAW_LIVE_CLI_BACKEND_PRESERVE_ENV";
const QA_LIVE_CLI_BACKEND_AUTH_MODE_ENV = "OPENCLAW_LIVE_CLI_BACKEND_AUTH_MODE";
async function getFreePort() {
	return await new Promise((resolve, reject) => {
		const server = net.createServer();
		server.once("error", reject);
		server.listen(0, "127.0.0.1", () => {
			const address = server.address();
			if (!address || typeof address === "string") {
				reject(/* @__PURE__ */ new Error("failed to allocate port"));
				return;
			}
			server.close((error) => error ? reject(error) : resolve(address.port));
		});
	});
}
function normalizeQaProviderModeEnv(env, providerMode) {
	if (providerMode === "mock-openai") {
		for (const key of QA_MOCK_BLOCKED_ENV_VARS) delete env[key];
		for (const key of Object.keys(env)) if (QA_MOCK_BLOCKED_ENV_KEY_PATTERNS.some((pattern) => pattern.test(key))) delete env[key];
		return env;
	}
	if (providerMode === "live-frontier") for (const { liveVar, providerVar } of QA_LIVE_ENV_ALIASES) {
		const liveValue = env[liveVar]?.trim();
		if (!liveValue || env[providerVar]?.trim()) continue;
		env[providerVar] = liveValue;
	}
	return env;
}
function resolveQaLiveCliAuthEnv(baseEnv, opts) {
	const parsePreservedCliEnv = () => {
		const raw = baseEnv[QA_LIVE_CLI_BACKEND_PRESERVE_ENV]?.trim();
		if (raw?.startsWith("[")) try {
			const parsed = JSON.parse(raw);
			return Array.isArray(parsed) ? parsed.filter((entry) => typeof entry === "string") : [];
		} catch {
			return [];
		}
		return (raw ?? "").split(/[,\s]+/).filter((entry) => entry.length > 0);
	};
	const renderPreservedCliEnv = (values) => JSON.stringify([...new Set(values)]);
	const authMode = opts?.claudeCliAuthMode ?? "auto";
	const hasAnthropicKey = Boolean(baseEnv.ANTHROPIC_API_KEY?.trim() || baseEnv.OPENCLAW_LIVE_ANTHROPIC_KEY?.trim());
	if (opts?.forwardHostHomeForClaudeCli && authMode === "api-key" && !hasAnthropicKey) throw new Error("Claude CLI API-key QA mode requires ANTHROPIC_API_KEY or OPENCLAW_LIVE_ANTHROPIC_KEY");
	const preserveEnvValues = (() => {
		if (!opts?.forwardHostHomeForClaudeCli) return;
		const values = parsePreservedCliEnv().filter((entry) => entry !== "ANTHROPIC_API_KEY");
		if (authMode === "api-key" || authMode === "auto" && hasAnthropicKey) values.push("ANTHROPIC_API_KEY");
		return renderPreservedCliEnv(values);
	})();
	const claudeCliEnv = opts?.forwardHostHomeForClaudeCli ? {
		[QA_LIVE_CLI_BACKEND_AUTH_MODE_ENV]: authMode,
		...preserveEnvValues ? { [QA_LIVE_CLI_BACKEND_PRESERVE_ENV]: preserveEnvValues } : {}
	} : {};
	const configuredCodexHome = baseEnv.CODEX_HOME?.trim();
	if (configuredCodexHome) return {
		CODEX_HOME: configuredCodexHome,
		...claudeCliEnv,
		...opts?.forwardHostHomeForClaudeCli && baseEnv.HOME?.trim() ? { HOME: baseEnv.HOME.trim() } : {}
	};
	const hostHome = baseEnv.HOME?.trim();
	if (!hostHome) return {};
	const codexHome = path.join(hostHome, ".codex");
	return {
		...existsSync(codexHome) ? { CODEX_HOME: codexHome } : {},
		...claudeCliEnv,
		...opts?.forwardHostHomeForClaudeCli ? { HOME: hostHome } : {}
	};
}
function buildQaRuntimeEnv(params) {
	const baseEnv = params.baseEnv ?? process.env;
	const normalizedEnv = normalizeQaProviderModeEnv({
		...baseEnv,
		HOME: params.homeDir,
		...params.providerMode === "live-frontier" ? resolveQaLiveCliAuthEnv(baseEnv, {
			forwardHostHomeForClaudeCli: params.forwardHostHomeForClaudeCli,
			claudeCliAuthMode: params.claudeCliAuthMode
		}) : {},
		OPENCLAW_HOME: params.homeDir,
		OPENCLAW_CONFIG_PATH: params.configPath,
		OPENCLAW_STATE_DIR: params.stateDir,
		OPENCLAW_OAUTH_DIR: path.join(params.stateDir, "credentials"),
		OPENCLAW_GATEWAY_TOKEN: params.gatewayToken,
		OPENCLAW_SKIP_BROWSER_CONTROL_SERVER: "1",
		OPENCLAW_SKIP_GMAIL_WATCHER: "1",
		OPENCLAW_SKIP_CANVAS_HOST: "1",
		OPENCLAW_NO_RESPAWN: "1",
		OPENCLAW_TEST_FAST: "1",
		OPENCLAW_QA_ALLOW_LOCAL_IMAGE_PROVIDER: "1",
		OPENCLAW_ALLOW_SLOW_REPLY_TESTS: "1",
		XDG_CONFIG_HOME: params.xdgConfigHome,
		XDG_DATA_HOME: params.xdgDataHome,
		XDG_CACHE_HOME: params.xdgCacheHome,
		...params.bundledPluginsDir ? { OPENCLAW_BUNDLED_PLUGINS_DIR: params.bundledPluginsDir } : {},
		...params.compatibilityHostVersion ? { OPENCLAW_COMPATIBILITY_HOST_VERSION: params.compatibilityHostVersion } : {}
	}, params.providerMode);
	delete normalizedEnv[QA_LIVE_ANTHROPIC_SETUP_TOKEN_ENV];
	delete normalizedEnv[QA_LIVE_SETUP_TOKEN_VALUE_ENV];
	return normalizedEnv;
}
function resolveQaLiveAnthropicSetupToken(env = process.env) {
	const token = (env[QA_LIVE_ANTHROPIC_SETUP_TOKEN_ENV]?.trim() || env[QA_LIVE_SETUP_TOKEN_VALUE_ENV]?.trim() || "").replaceAll(/\s+/g, "");
	if (!token) return null;
	const tokenError = validateAnthropicSetupToken(token);
	if (tokenError) throw new Error(`Invalid QA Anthropic setup-token: ${tokenError}`);
	return {
		token,
		profileId: env[QA_LIVE_ANTHROPIC_SETUP_TOKEN_PROFILE_ENV]?.trim() || QA_LIVE_ANTHROPIC_SETUP_TOKEN_PROFILE_ID
	};
}
async function stageQaLiveAnthropicSetupToken(params) {
	const resolved = resolveQaLiveAnthropicSetupToken(params.env);
	if (!resolved) return params.cfg;
	const agentDir = path.join(params.stateDir, "agents", "main", "agent");
	await fs$1.mkdir(agentDir, { recursive: true });
	upsertAuthProfile({
		profileId: resolved.profileId,
		credential: {
			type: "token",
			provider: "anthropic",
			token: resolved.token
		},
		agentDir
	});
	return applyAuthProfileConfig(params.cfg, {
		profileId: resolved.profileId,
		provider: "anthropic",
		mode: "token",
		displayName: "QA setup-token"
	});
}
function isRetryableGatewayCallError(details) {
	return details.includes("handshake timeout") || details.includes("gateway closed (1000") || details.includes("gateway closed (1012)") || details.includes("gateway closed (1006") || details.includes("abnormal closure") || details.includes("service restart");
}
async function fetchLocalGatewayHealth(params) {
	const { response, release } = await fetchWithSsrFGuard({
		url: `${params.baseUrl}${params.healthPath}`,
		init: { signal: AbortSignal.timeout(2e3) },
		policy: { allowPrivateNetwork: true },
		auditContext: "qa-lab-gateway-child-health"
	});
	try {
		return response.ok;
	} finally {
		await release();
	}
}
const __testing = {
	buildQaRuntimeEnv,
	fetchLocalGatewayHealth,
	isRetryableGatewayCallError,
	readQaLiveProviderConfigOverrides,
	resolveQaLiveAnthropicSetupToken,
	stageQaLiveAnthropicSetupToken,
	resolveQaLiveCliAuthEnv,
	resolveQaOwnerPluginIdsForProviderIds,
	resolveQaBundledPluginsSourceRoot,
	resolveQaRuntimeHostVersion,
	createQaBundledPluginsDir,
	stopQaGatewayChildProcessTree
};
function hasChildExited(child) {
	return child.exitCode !== null || child.signalCode !== null;
}
function signalQaGatewayChildProcessTree(child, signal) {
	if (!child.pid) return;
	try {
		if (process.platform === "win32") {
			child.kill(signal);
			return;
		}
		process.kill(-child.pid, signal);
	} catch {
		try {
			child.kill(signal);
		} catch {}
	}
}
async function waitForQaGatewayChildExit(child, timeoutMs) {
	if (hasChildExited(child)) return true;
	return await Promise.race([new Promise((resolve) => child.once("exit", () => resolve(true))), setTimeout(timeoutMs).then(() => false)]);
}
async function stopQaGatewayChildProcessTree(child, opts) {
	if (hasChildExited(child)) return;
	signalQaGatewayChildProcessTree(child, "SIGTERM");
	if (await waitForQaGatewayChildExit(child, opts?.gracefulTimeoutMs ?? 5e3)) return;
	signalQaGatewayChildProcessTree(child, "SIGKILL");
	await waitForQaGatewayChildExit(child, opts?.forceTimeoutMs ?? 2e3);
}
function resolveQaBundledPluginsSourceRoot(repoRoot) {
	const candidates = [
		path.join(repoRoot, "dist", "extensions"),
		path.join(repoRoot, "dist-runtime", "extensions"),
		path.join(repoRoot, "extensions")
	];
	for (const candidate of candidates) if (existsSync(candidate)) return candidate;
	throw new Error("failed to resolve qa bundled plugins source root");
}
async function resolveQaOwnerPluginIdsForProviderIds(params) {
	const providerIds = [...new Set(params.providerIds.map((providerId) => providerId.trim()))].filter((providerId) => providerId.length > 0);
	if (providerIds.length === 0) return [];
	const remainingProviderIds = new Set(providerIds);
	const ownerPluginIds = /* @__PURE__ */ new Set();
	const sourceRoot = resolveQaBundledPluginsSourceRoot(params.repoRoot);
	for (const entry of await fs$1.readdir(sourceRoot, { withFileTypes: true })) {
		if (!entry.isDirectory()) continue;
		const manifestPath = path.join(sourceRoot, entry.name, "openclaw.plugin.json");
		if (!existsSync(manifestPath)) continue;
		const manifest = JSON.parse(await fs$1.readFile(manifestPath, "utf8"));
		const pluginId = typeof manifest.id === "string" ? manifest.id.trim() : entry.name;
		if (!pluginId) continue;
		const ownedIds = new Set([
			pluginId,
			...Array.isArray(manifest.providers) ? manifest.providers : [],
			...Array.isArray(manifest.cliBackends) ? manifest.cliBackends : []
		].filter((ownedId) => typeof ownedId === "string"));
		for (const providerId of providerIds) {
			if (!ownedIds.has(providerId)) continue;
			ownerPluginIds.add(pluginId);
			remainingProviderIds.delete(providerId);
		}
	}
	for (const providerId of remainingProviderIds) {
		const providerConfig = params.providerConfigs?.[providerId];
		if (providerConfig && isQaOpenAiResponsesProviderConfig(providerConfig)) {
			ownerPluginIds.add(QA_OPENAI_PLUGIN_ID);
			continue;
		}
		ownerPluginIds.add(providerId);
	}
	return [...ownerPluginIds];
}
function resolveQaUserPath(value, env = process.env) {
	if (value === "~") return env.HOME ?? os.homedir();
	if (value.startsWith("~/")) return path.join(env.HOME ?? os.homedir(), value.slice(2));
	return path.resolve(value);
}
function resolveQaLiveProviderConfigPath(env = process.env) {
	const explicit = env[QA_LIVE_PROVIDER_CONFIG_PATH_ENV]?.trim() || env.OPENCLAW_CONFIG_PATH?.trim();
	return explicit ? {
		path: resolveQaUserPath(explicit, env),
		explicit: true
	} : {
		path: path.join(os.homedir(), ".openclaw", "openclaw.json"),
		explicit: false
	};
}
function isRecord(value) {
	return Boolean(value) && typeof value === "object" && !Array.isArray(value);
}
function isQaModelProviderConfig(value) {
	return isRecord(value) && typeof value.baseUrl === "string" && Array.isArray(value.models);
}
function isQaOpenAiResponsesProviderConfig(config) {
	return config.api === "openai-responses" || config.models.some((model) => model.api === "openai-responses");
}
async function readQaLiveProviderConfigOverrides(params) {
	const providerIds = [...new Set(params.providerIds.map((providerId) => providerId.trim()))].filter((providerId) => providerId.length > 0);
	if (providerIds.length === 0) return {};
	const configPath = resolveQaLiveProviderConfigPath(params.env);
	if (!existsSync(configPath.path)) return {};
	try {
		const raw = await fs$1.readFile(configPath.path, "utf8");
		const parsed = JSON.parse(raw);
		const providers = isRecord(parsed) ? isRecord(parsed.models) ? isRecord(parsed.models.providers) ? parsed.models.providers : {} : {} : {};
		const selected = {};
		for (const providerId of providerIds) {
			const providerConfig = providers[providerId];
			if (isQaModelProviderConfig(providerConfig)) selected[providerId] = providerConfig;
		}
		return selected;
	} catch (error) {
		if (configPath.explicit) throw new Error(`failed to read ${QA_LIVE_PROVIDER_CONFIG_PATH_ENV} provider config: ${formatErrorMessage(error)}`, { cause: error });
		return {};
	}
}
function parseStableSemverFloor(value) {
	if (!value) return null;
	const match = value.trim().match(/(\d+)\.(\d+)\.(\d+)/);
	if (!match) return null;
	return {
		major: Number.parseInt(match[1] ?? "", 10),
		minor: Number.parseInt(match[2] ?? "", 10),
		patch: Number.parseInt(match[3] ?? "", 10),
		label: `${match[1]}.${match[2]}.${match[3]}`
	};
}
function compareSemverFloors(left, right) {
	if (!left && !right) return 0;
	if (!left) return -1;
	if (!right) return 1;
	if (left.major !== right.major) return left.major - right.major;
	if (left.minor !== right.minor) return left.minor - right.minor;
	return left.patch - right.patch;
}
async function resolveQaRuntimeHostVersion(params) {
	const rootPackageRaw = await fs$1.readFile(path.join(params.repoRoot, "package.json"), "utf8");
	let selected = parseStableSemverFloor(JSON.parse(rootPackageRaw).version);
	for (const pluginId of params.allowedPluginIds) {
		const packagePath = path.join(params.bundledPluginsSourceRoot, pluginId, "package.json");
		if (!existsSync(packagePath)) continue;
		const packageRaw = await fs$1.readFile(packagePath, "utf8");
		const candidate = parseStableSemverFloor(JSON.parse(packageRaw).openclaw?.install?.minHostVersion);
		if (compareSemverFloors(candidate, selected) > 0) selected = candidate;
	}
	return selected?.label;
}
async function createQaBundledPluginsDir(params) {
	const sourceRoot = resolveQaBundledPluginsSourceRoot(params.repoRoot);
	const sourceTreeRoot = path.dirname(sourceRoot);
	if (sourceTreeRoot === path.join(params.repoRoot, "dist") || sourceTreeRoot === path.join(params.repoRoot, "dist-runtime")) {
		const stagedRoot = path.join(params.repoRoot, ".artifacts", "qa-runtime", path.basename(params.tempRoot));
		await fs$1.rm(stagedRoot, {
			recursive: true,
			force: true
		});
		await fs$1.mkdir(stagedRoot, { recursive: true });
		const stagedTreeRoot = path.join(stagedRoot, path.basename(sourceTreeRoot));
		await fs$1.mkdir(stagedTreeRoot, { recursive: true });
		for (const entry of await fs$1.readdir(sourceTreeRoot, { withFileTypes: true })) {
			const sourcePath = path.join(sourceTreeRoot, entry.name);
			const targetPath = path.join(stagedTreeRoot, entry.name);
			if (entry.name === "extensions") {
				await fs$1.mkdir(targetPath, { recursive: true });
				for (const pluginId of params.allowedPluginIds) {
					const sourceDir = path.join(sourceRoot, pluginId);
					if (!existsSync(sourceDir)) throw new Error(`qa bundled plugin not found: ${pluginId} (${sourceDir})`);
					await fs$1.cp(sourceDir, path.join(targetPath, pluginId), { recursive: true });
				}
				continue;
			}
			await fs$1.symlink(sourcePath, targetPath);
		}
		return {
			bundledPluginsDir: path.join(stagedTreeRoot, "extensions"),
			stagedRoot
		};
	}
	const bundledPluginsDir = path.join(params.tempRoot, "bundled-plugins");
	await fs$1.mkdir(bundledPluginsDir, { recursive: true });
	for (const pluginId of params.allowedPluginIds) {
		const sourceDir = path.join(sourceRoot, pluginId);
		if (!existsSync(sourceDir)) throw new Error(`qa bundled plugin not found: ${pluginId} (${sourceDir})`);
		await fs$1.cp(sourceDir, path.join(bundledPluginsDir, pluginId), { recursive: true });
	}
	return {
		bundledPluginsDir,
		stagedRoot: null
	};
}
async function waitForGatewayReady(params) {
	const startedAt = Date.now();
	while (Date.now() - startedAt < (params.timeoutMs ?? 6e4)) {
		if (params.child.exitCode !== null || params.child.signalCode !== null) throw new Error(`gateway exited before becoming healthy (exitCode=${String(params.child.exitCode)}, signal=${String(params.child.signalCode)}):\n${params.logs()}`);
		for (const healthPath of ["/readyz", "/healthz"]) try {
			if (await fetchLocalGatewayHealth({
				baseUrl: params.baseUrl,
				healthPath
			})) return;
		} catch {}
		await setTimeout(250);
	}
	throw new Error(`gateway failed to become healthy:\n${params.logs()}`);
}
function isRetryableRpcStartupError(error) {
	const details = formatErrorMessage(error);
	return details.includes("handshake timeout") || details.includes("gateway closed (1000") || details.includes("gateway closed (1006") || details.includes("gateway closed (1012)");
}
function resolveQaControlUiRoot(params) {
	if (params.controlUiEnabled === false) return;
	const controlUiRoot = path.join(params.repoRoot, "dist", "control-ui");
	return existsSync(path.join(controlUiRoot, "index.html")) ? controlUiRoot : void 0;
}
async function startQaGatewayChild(params) {
	const tempRoot = await fs$1.mkdtemp(path.join(resolvePreferredOpenClawTmpDir(), "openclaw-qa-suite-"));
	const runtimeCwd = tempRoot;
	const distEntryPath = path.join(params.repoRoot, "dist", "index.js");
	const workspaceDir = path.join(tempRoot, "workspace");
	const stateDir = path.join(tempRoot, "state");
	const homeDir = path.join(tempRoot, "home");
	const xdgConfigHome = path.join(tempRoot, "xdg-config");
	const xdgDataHome = path.join(tempRoot, "xdg-data");
	const xdgCacheHome = path.join(tempRoot, "xdg-cache");
	const configPath = path.join(tempRoot, "openclaw.json");
	const gatewayPort = await getFreePort();
	const gatewayToken = `qa-suite-${randomUUID()}`;
	await seedQaAgentWorkspace({
		workspaceDir,
		repoRoot: params.repoRoot
	});
	await Promise.all([
		fs$1.mkdir(stateDir, { recursive: true }),
		fs$1.mkdir(homeDir, { recursive: true }),
		fs$1.mkdir(xdgConfigHome, { recursive: true }),
		fs$1.mkdir(xdgDataHome, { recursive: true }),
		fs$1.mkdir(xdgCacheHome, { recursive: true })
	]);
	const liveProviderIds = params.providerMode === "live-frontier" ? [params.primaryModel, params.alternateModel].map((modelRef) => typeof modelRef === "string" ? splitQaModelRef(modelRef)?.provider : void 0).filter((providerId) => Boolean(providerId)) : [];
	const liveProviderConfigs = await readQaLiveProviderConfigOverrides({ providerIds: liveProviderIds });
	const enabledPluginIds = liveProviderIds.length > 0 ? await resolveQaOwnerPluginIdsForProviderIds({
		repoRoot: params.repoRoot,
		providerIds: liveProviderIds,
		providerConfigs: liveProviderConfigs
	}) : void 0;
	let cfg = buildQaGatewayConfig({
		bind: "loopback",
		gatewayPort,
		gatewayToken,
		providerBaseUrl: params.providerBaseUrl,
		qaBusBaseUrl: params.qaBusBaseUrl,
		includeQaChannel: params.includeQaChannel,
		workspaceDir,
		controlUiRoot: resolveQaControlUiRoot({
			repoRoot: params.repoRoot,
			controlUiEnabled: params.controlUiEnabled
		}),
		controlUiAllowedOrigins: params.controlUiAllowedOrigins,
		providerMode: params.providerMode,
		primaryModel: params.primaryModel,
		alternateModel: params.alternateModel,
		enabledPluginIds,
		liveProviderConfigs,
		fastMode: params.fastMode,
		thinkingDefault: params.thinkingDefault,
		controlUiEnabled: params.controlUiEnabled
	});
	cfg = await stageQaLiveAnthropicSetupToken({
		cfg,
		stateDir
	});
	cfg = params.mutateConfig ? params.mutateConfig(cfg) : cfg;
	await fs$1.writeFile(configPath, `${JSON.stringify(cfg, null, 2)}\n`, {
		encoding: "utf8",
		mode: 384
	});
	const allowedPluginIds = [...cfg.plugins?.allow ?? [], "openai"].filter((pluginId, index, array) => {
		return typeof pluginId === "string" && pluginId.length > 0 && array.indexOf(pluginId) === index;
	});
	const bundledPluginsSourceRoot = resolveQaBundledPluginsSourceRoot(params.repoRoot);
	const { bundledPluginsDir, stagedRoot: stagedBundledPluginsRoot } = await createQaBundledPluginsDir({
		repoRoot: params.repoRoot,
		tempRoot,
		allowedPluginIds
	});
	const runtimeHostVersion = await resolveQaRuntimeHostVersion({
		repoRoot: params.repoRoot,
		bundledPluginsSourceRoot,
		allowedPluginIds
	});
	const stdout = [];
	const stderr = [];
	const stdoutLogPath = path.join(tempRoot, "gateway.stdout.log");
	const stderrLogPath = path.join(tempRoot, "gateway.stderr.log");
	const stdoutLog = createWriteStream(stdoutLogPath, { flags: "a" });
	const stderrLog = createWriteStream(stderrLogPath, { flags: "a" });
	const env = buildQaRuntimeEnv({
		configPath,
		gatewayToken,
		homeDir,
		stateDir,
		xdgConfigHome,
		xdgDataHome,
		xdgCacheHome,
		bundledPluginsDir,
		compatibilityHostVersion: runtimeHostVersion,
		providerMode: params.providerMode,
		forwardHostHomeForClaudeCli: liveProviderIds.includes("claude-cli"),
		claudeCliAuthMode: params.claudeCliAuthMode
	});
	const child = spawn(process.execPath, [
		distEntryPath,
		"gateway",
		"run",
		"--port",
		String(gatewayPort),
		"--bind",
		"loopback",
		"--allow-unconfigured"
	], {
		cwd: runtimeCwd,
		env,
		detached: process.platform !== "win32",
		stdio: [
			"ignore",
			"pipe",
			"pipe"
		]
	});
	child.stdout.on("data", (chunk) => {
		const buffer = Buffer.from(chunk);
		stdout.push(buffer);
		stdoutLog.write(buffer);
	});
	child.stderr.on("data", (chunk) => {
		const buffer = Buffer.from(chunk);
		stderr.push(buffer);
		stderrLog.write(buffer);
	});
	const baseUrl = `http://127.0.0.1:${gatewayPort}`;
	const wsUrl = `ws://127.0.0.1:${gatewayPort}`;
	const logs = () => `${Buffer.concat(stdout).toString("utf8")}\n${Buffer.concat(stderr).toString("utf8")}`.trim();
	const keepTemp = process.env.OPENCLAW_QA_KEEP_TEMP === "1";
	let rpcClient;
	try {
		await waitForGatewayReady({
			baseUrl,
			logs,
			child,
			timeoutMs: 12e4
		});
		let lastRpcError = null;
		for (let attempt = 1; attempt <= 4; attempt += 1) try {
			rpcClient = await startQaGatewayRpcClient({
				wsUrl,
				token: gatewayToken,
				logs
			});
			break;
		} catch (error) {
			lastRpcError = error;
			if (attempt >= 4 || !isRetryableRpcStartupError(error)) throw error;
			await setTimeout(500 * attempt);
			await waitForGatewayReady({
				baseUrl,
				logs,
				child,
				timeoutMs: 15e3
			});
		}
		if (!rpcClient) throw lastRpcError ?? /* @__PURE__ */ new Error("qa gateway rpc client failed to start");
	} catch (error) {
		stdoutLog.end();
		stderrLog.end();
		await stopQaGatewayChildProcessTree(child, { gracefulTimeoutMs: 1e3 }).catch(() => {});
		if (!keepTemp && stagedBundledPluginsRoot) await fs$1.rm(stagedBundledPluginsRoot, {
			recursive: true,
			force: true
		}).catch(() => {});
		throw error;
	}
	return {
		cfg,
		baseUrl,
		wsUrl,
		pid: child.pid ?? null,
		token: gatewayToken,
		workspaceDir,
		tempRoot,
		configPath,
		runtimeEnv: env,
		logs,
		async restart(signal = "SIGUSR1") {
			if (!child.pid) throw new Error("qa gateway child has no pid");
			process.kill(child.pid, signal);
		},
		async call(method, rpcParams, opts) {
			const timeoutMs = opts?.timeoutMs ?? 2e4;
			const retryOnRestart = opts?.retryOnRestart !== false;
			let lastDetails = "";
			for (let attempt = 1; attempt <= 3; attempt += 1) try {
				return await rpcClient.request(method, rpcParams, {
					...opts,
					timeoutMs
				});
			} catch (error) {
				const details = formatErrorMessage(error);
				lastDetails = details;
				if (attempt >= 3 || !retryOnRestart || !isRetryableGatewayCallError(details)) throw new Error(`${details}\nGateway logs:\n${logs()}`, { cause: error });
				await waitForGatewayReady({
					baseUrl,
					logs,
					child,
					timeoutMs: Math.max(1e4, timeoutMs)
				});
			}
			throw new Error(`${lastDetails}\nGateway logs:\n${logs()}`);
		},
		async stop(opts) {
			await rpcClient.stop().catch(() => {});
			stdoutLog.end();
			stderrLog.end();
			await stopQaGatewayChildProcessTree(child);
			if (!(opts?.keepTemp ?? keepTemp)) {
				await fs$1.rm(tempRoot, {
					recursive: true,
					force: true
				});
				if (stagedBundledPluginsRoot) await fs$1.rm(stagedBundledPluginsRoot, {
					recursive: true,
					force: true
				});
			}
		}
	};
}
//#endregion
export { stageQaLiveAnthropicSetupToken as a, seedQaAgentWorkspace as c, resolveQaControlUiRoot as i, buildQaScenarioPlanMarkdown as l, buildQaRuntimeEnv as n, startQaGatewayChild as o, normalizeQaProviderModeEnv as r, startQaMockOpenAiServer as s, __testing as t, readQaAgentIdentityMarkdown as u };
