// openclaw/dist/googlechat-DKeMrYKw.js
import "./types.secrets-B4TpMTxm.js";
import "./config-schema-CZtpPFHc.js";
import "./registry-DR0Inr-7.js";
import "./zod-schema.providers-core-Bof37NVV.js";
import "./fetch-guard-DKe4b6qW.js";
import "./common-BTJs9jR5.js";
import { n as resolveChannelGroupRequireMention } from "./group-policy-YqA7ta9g.js";
import "./fetch-B85tpt48.js";
import "./setup-helpers-B_T9A0B8.js";
import "./channel-policy-no84IHKf.js";
import "./dm-policy-shared-D_6CpFGk.js";
import "./setup-wizard-helpers-BD6nhg53.js";
import "./channel-reply-pipeline-CJZpAxBU.js";
import "./channel-pairing-rPdRacuP.js";
import "./status-helpers-Csb7hAJZ.js";
import "./webhook-ingress-B4Mrq-gj.js";
import { t as createOptionalChannelSetupSurface } from "./channel-setup-DTFU7XjL.js";
import "./web-media-BErkJWj2.js";
import "./outbound-media-CDUyXUeY.js";
//#region src/plugin-sdk/googlechat.ts
function resolveGoogleChatGroupRequireMention(params) {
	return resolveChannelGroupRequireMention({
		cfg: params.cfg,
		channel: "googlechat",
		groupId: params.groupId,
		accountId: params.accountId
	});
}
const googlechatSetup = createOptionalChannelSetupSurface({
	channel: "googlechat",
	label: "Google Chat",
	npmSpec: "@openclaw/googlechat",
	docsPath: "/channels/googlechat"
});
const googlechatSetupAdapter = googlechatSetup.setupAdapter;
const googlechatSetupWizard = googlechatSetup.setupWizard;
//#endregion
export { googlechatSetupWizard as n, resolveGoogleChatGroupRequireMention as r, googlechatSetupAdapter as t };
