// openclaw/dist/infra-runtime-DENIjfhx.js
import "./errors-D8p6rxH8.js";
import "./tmp-openclaw-dir-DgmFzeYV.js";
import "./env-jndHmrmK.js";
import "./file-lock-B-BZH3IM.js";
import "./undici-global-dispatcher-D8nttVY6.js";
import "./fetch-guard-DKe4b6qW.js";
import "./ssrf-tCiyDHOt.js";
import "./exec-approval-reply-BhpFcpTV.js";
import "./exec-approvals-BIBEOnML.js";
import "./fs-safe-x0_RgCMh.js";
import "./system-events-BwxCjOwe.js";
import { n as drainPendingDeliveries$1 } from "./delivery-queue-DlULVbWl.js";
import "./proxy-fetch-C9aK_HTJ.js";
import "./retry-BQn5Qrea.js";
import "./secret-file-BgdDO4o1.js";
import "./http-body-idixF2ZF.js";
import "./approval-native-runtime-DABPBY_X.js";
import "./exec-approval-session-target-IjPSb9mx.js";
import "./heartbeat-visibility-B5n9-bz_.js";
import "./transport-ready-CIbz1r5V.js";
import "./identity-BUsSaT8c.js";
import "./retry-policy-kLd2Xctb.js";
import "./ssrf-policy-B9ZyjgPE.js";
//#region src/plugin-sdk/infra-runtime.ts
function normalizeWhatsAppReconnectAccountId(accountId) {
	return (accountId ?? "").trim() || "default";
}
const WHATSAPP_NO_LISTENER_ERROR_RE = /No active WhatsApp Web listener/i;
let outboundDeliverRuntimePromise = null;
async function loadOutboundDeliverRuntime() {
	outboundDeliverRuntimePromise ??= import("./deliver-runtime-3u-4wn9F.js");
	return await outboundDeliverRuntimePromise;
}
async function drainPendingDeliveries(opts) {
	const deliver = opts.deliver ?? (await loadOutboundDeliverRuntime()).deliverOutboundPayloads;
	await drainPendingDeliveries$1({
		...opts,
		deliver
	});
}
/**
* @deprecated Prefer plugin-owned reconnect policy wired through
* `drainPendingDeliveries(...)`. This compatibility shim preserves the
* historical public SDK symbol for existing plugin callers.
*/
async function drainReconnectQueue(opts) {
	const normalizedAccountId = normalizeWhatsAppReconnectAccountId(opts.accountId);
	await drainPendingDeliveries({
		drainKey: `whatsapp:${normalizedAccountId}`,
		logLabel: "WhatsApp reconnect drain",
		cfg: opts.cfg,
		log: opts.log,
		stateDir: opts.stateDir,
		deliver: opts.deliver,
		selectEntry: (entry) => ({
			match: entry.channel === "whatsapp" && normalizeWhatsAppReconnectAccountId(entry.accountId) === normalizedAccountId && typeof entry.lastError === "string" && WHATSAPP_NO_LISTENER_ERROR_RE.test(entry.lastError),
			bypassBackoff: true
		})
	});
}
//#endregion
export { drainReconnectQueue as n, drainPendingDeliveries as t };
