// openclaw/dist/lab-server-Dq4QthbC.js
import { i as formatErrorMessage } from "./errors-D8p6rxH8.js";
import { i as normalizeLowercaseStringOrEmpty, o as normalizeOptionalLowercaseString } from "./string-coerce-BUSzWgUA.js";
import { c as getDebugProxyCaptureStore, d as resolveDebugProxySettings } from "./runtime-BIHuBcHI.js";
import { t as extractToolPayload } from "./tool-payload-BX5uK_8i.js";
import "./text-runtime-3VPC1sGh.js";
import "./error-runtime-Da0X1TOQ.js";
import "./proxy-capture-ChXXhV_K.js";
import { d as qaChannelPlugin, g as setQaChannelRuntime } from "./qa-channel-UvOFzZEG.js";
import "./runtime-api-pvMSXsR4.js";
import { i as readQaBootstrapScenarioCatalog } from "./scenario-catalog-DQ-BbL_S.js";
import { a as writeError, o as writeJson, r as handleQaBusRequest, t as closeQaHttpServer } from "./bus-server-2ab_NN8Y.js";
import { a as normalizeQaRunSelection, n as createQaRunOutputDir, t as createIdleQaRunnerSnapshot } from "./run-config-B359hkWD.js";
import { t as renderQaMarkdownReport } from "./report-fUz6OldI.js";
import { fileURLToPath } from "node:url";
import fs from "node:fs";
import path from "node:path";
import fs$1 from "node:fs/promises";
import { createHash, randomUUID } from "node:crypto";
import net from "node:net";
import { createServer as createServer$1, request } from "node:http";
import { request as request$1 } from "node:https";
import tls from "node:tls";
//#region extensions/qa-lab/src/bus-queries.ts
const DEFAULT_ACCOUNT_ID = "default";
function normalizeAccountId(raw) {
	return raw?.trim() || "default";
}
function normalizeConversationFromTarget(target) {
	const trimmed = target.trim();
	if (trimmed.startsWith("thread:")) {
		const rest = trimmed.slice(7);
		const slash = rest.indexOf("/");
		if (slash > 0) return {
			conversation: {
				id: rest.slice(0, slash),
				kind: "channel"
			},
			threadId: rest.slice(slash + 1)
		};
	}
	if (trimmed.startsWith("channel:")) return { conversation: {
		id: trimmed.slice(8),
		kind: "channel"
	} };
	if (trimmed.startsWith("dm:")) return { conversation: {
		id: trimmed.slice(3),
		kind: "direct"
	} };
	return { conversation: {
		id: trimmed,
		kind: "direct"
	} };
}
function cloneMessage(message) {
	return {
		...message,
		conversation: { ...message.conversation },
		attachments: (message.attachments ?? []).map((attachment) => cloneAttachment(attachment)),
		reactions: message.reactions.map((reaction) => ({ ...reaction }))
	};
}
function cloneAttachment(attachment) {
	return { ...attachment };
}
function cloneEvent(event) {
	switch (event.kind) {
		case "inbound-message":
		case "outbound-message":
		case "message-edited":
		case "message-deleted":
		case "reaction-added": return {
			...event,
			message: cloneMessage(event.message)
		};
		case "thread-created": return {
			...event,
			thread: { ...event.thread }
		};
	}
	throw new Error("Unsupported QA bus event kind");
}
function buildQaBusSnapshot(params) {
	return {
		cursor: params.cursor,
		conversations: Array.from(params.conversations.values()).map((conversation) => ({ ...conversation })),
		threads: Array.from(params.threads.values()).map((thread) => ({ ...thread })),
		messages: Array.from(params.messages.values()).map((message) => cloneMessage(message)),
		events: params.events.map((event) => cloneEvent(event))
	};
}
function readQaBusMessage(params) {
	const message = params.messages.get(params.input.messageId);
	if (!message) throw new Error(`qa-bus message not found: ${params.input.messageId}`);
	return cloneMessage(message);
}
function searchQaBusMessages(params) {
	const accountId = normalizeAccountId(params.input.accountId);
	const limit = Math.max(1, Math.min(params.input.limit ?? 20, 100));
	const query = normalizeOptionalLowercaseString(params.input.query);
	return Array.from(params.messages.values()).filter((message) => message.accountId === accountId).filter((message) => params.input.conversationId ? message.conversation.id === params.input.conversationId : true).filter((message) => params.input.threadId ? message.threadId === params.input.threadId : true).filter((message) => {
		if (!query) return true;
		const searchableAttachmentText = (message.attachments ?? []).flatMap((attachment) => [
			attachment.fileName,
			attachment.altText,
			attachment.transcript,
			attachment.mimeType
		]).filter((value) => Boolean(value)).join(" ").toLowerCase();
		return `${normalizeOptionalLowercaseString(message.text) ?? ""} ${searchableAttachmentText}`.includes(query);
	}).slice(-limit).map((message) => cloneMessage(message));
}
function pollQaBusEvents(params) {
	const accountId = normalizeAccountId(params.input?.accountId);
	const startCursor = params.input?.cursor ?? 0;
	const effectiveStartCursor = params.cursor < startCursor ? 0 : startCursor;
	const limit = Math.max(1, Math.min(params.input?.limit ?? 100, 500));
	const matches = params.events.filter((event) => event.accountId === accountId && event.cursor > effectiveStartCursor).slice(0, limit).map((event) => cloneEvent(event));
	return {
		cursor: params.cursor,
		events: matches
	};
}
//#endregion
//#region extensions/qa-lab/src/bus-waiters.ts
const DEFAULT_WAIT_TIMEOUT_MS = 5e3;
function createQaBusMatcher(input) {
	return (snapshot) => {
		if (input.kind === "event-kind") return snapshot.events.find((event) => event.kind === input.eventKind) ?? null;
		if (input.kind === "thread-id") return snapshot.threads.find((thread) => thread.id === input.threadId) ?? null;
		return snapshot.messages.find((message) => (!input.direction || message.direction === input.direction) && message.text.includes(input.textIncludes)) ?? null;
	};
}
function createQaBusWaiterStore(getSnapshot) {
	const waiters = /* @__PURE__ */ new Set();
	return {
		reset(reason = "qa-bus reset") {
			for (const waiter of waiters) {
				clearTimeout(waiter.timer);
				waiter.reject(new Error(reason));
			}
			waiters.clear();
		},
		settle() {
			if (waiters.size === 0) return;
			const snapshot = getSnapshot();
			for (const waiter of Array.from(waiters)) {
				const match = waiter.matcher(snapshot);
				if (!match) continue;
				clearTimeout(waiter.timer);
				waiters.delete(waiter);
				waiter.resolve(match);
			}
		},
		async waitFor(input) {
			const matcher = createQaBusMatcher(input);
			const immediate = matcher(getSnapshot());
			if (immediate) return immediate;
			return await new Promise((resolve, reject) => {
				const timeoutMs = input.timeoutMs ?? 5e3;
				const waiter = {
					resolve,
					reject,
					matcher,
					timer: setTimeout(() => {
						waiters.delete(waiter);
						reject(/* @__PURE__ */ new Error(`qa-bus wait timeout after ${timeoutMs}ms`));
					}, timeoutMs)
				};
				waiters.add(waiter);
			});
		}
	};
}
//#endregion
//#region extensions/qa-lab/src/bus-state.ts
const DEFAULT_BOT_ID = "openclaw";
const DEFAULT_BOT_NAME = "OpenClaw QA";
function createQaBusState() {
	const conversations = /* @__PURE__ */ new Map();
	const threads = /* @__PURE__ */ new Map();
	const messages = /* @__PURE__ */ new Map();
	const events = [];
	let cursor = 0;
	const waiters = createQaBusWaiterStore(() => buildQaBusSnapshot({
		cursor,
		conversations,
		threads,
		messages,
		events
	}));
	const pushEvent = (event) => {
		cursor += 1;
		const next = typeof event === "function" ? event(cursor) : event;
		const finalized = {
			cursor,
			...next
		};
		events.push(finalized);
		waiters.settle();
		return finalized;
	};
	const ensureConversation = (conversation) => {
		const existing = conversations.get(conversation.id);
		if (existing) {
			if (!existing.title && conversation.title) existing.title = conversation.title;
			return existing;
		}
		const created = { ...conversation };
		conversations.set(created.id, created);
		return created;
	};
	const createMessage = (params) => {
		const conversation = ensureConversation(params.conversation);
		const message = {
			id: randomUUID(),
			accountId: params.accountId,
			direction: params.direction,
			conversation,
			senderId: params.senderId,
			senderName: params.senderName,
			text: params.text,
			timestamp: params.timestamp ?? Date.now(),
			threadId: params.threadId,
			threadTitle: params.threadTitle,
			replyToId: params.replyToId,
			attachments: params.attachments?.map((attachment) => ({ ...attachment })) ?? [],
			reactions: []
		};
		messages.set(message.id, message);
		return message;
	};
	return {
		reset() {
			conversations.clear();
			threads.clear();
			messages.clear();
			events.length = 0;
			waiters.reset();
		},
		getSnapshot() {
			return buildQaBusSnapshot({
				cursor,
				conversations,
				threads,
				messages,
				events
			});
		},
		addInboundMessage(input) {
			const accountId = normalizeAccountId(input.accountId);
			const message = createMessage({
				direction: "inbound",
				accountId,
				conversation: input.conversation,
				senderId: input.senderId,
				senderName: input.senderName,
				text: input.text,
				timestamp: input.timestamp,
				threadId: input.threadId,
				threadTitle: input.threadTitle,
				replyToId: input.replyToId,
				attachments: input.attachments
			});
			pushEvent({
				kind: "inbound-message",
				accountId,
				message: cloneMessage(message)
			});
			return cloneMessage(message);
		},
		addOutboundMessage(input) {
			const accountId = normalizeAccountId(input.accountId);
			const { conversation, threadId } = normalizeConversationFromTarget(input.to);
			const message = createMessage({
				direction: "outbound",
				accountId,
				conversation,
				senderId: input.senderId?.trim() || DEFAULT_BOT_ID,
				senderName: input.senderName?.trim() || DEFAULT_BOT_NAME,
				text: input.text,
				timestamp: input.timestamp,
				threadId: input.threadId ?? threadId,
				replyToId: input.replyToId,
				attachments: input.attachments
			});
			pushEvent({
				kind: "outbound-message",
				accountId,
				message: cloneMessage(message)
			});
			return cloneMessage(message);
		},
		createThread(input) {
			const accountId = normalizeAccountId(input.accountId);
			const thread = {
				id: `thread-${randomUUID()}`,
				accountId,
				conversationId: input.conversationId,
				title: input.title,
				createdAt: input.timestamp ?? Date.now(),
				createdBy: input.createdBy?.trim() || DEFAULT_BOT_ID
			};
			threads.set(thread.id, thread);
			ensureConversation({
				id: input.conversationId,
				kind: "channel"
			});
			pushEvent({
				kind: "thread-created",
				accountId,
				thread: { ...thread }
			});
			return { ...thread };
		},
		reactToMessage(input) {
			const accountId = normalizeAccountId(input.accountId);
			const message = messages.get(input.messageId);
			if (!message) throw new Error(`qa-bus message not found: ${input.messageId}`);
			const reaction = {
				emoji: input.emoji,
				senderId: input.senderId?.trim() || DEFAULT_BOT_ID,
				timestamp: input.timestamp ?? Date.now()
			};
			message.reactions.push(reaction);
			pushEvent({
				kind: "reaction-added",
				accountId,
				message: cloneMessage(message),
				emoji: reaction.emoji,
				senderId: reaction.senderId
			});
			return cloneMessage(message);
		},
		editMessage(input) {
			const accountId = normalizeAccountId(input.accountId);
			const message = messages.get(input.messageId);
			if (!message) throw new Error(`qa-bus message not found: ${input.messageId}`);
			message.text = input.text;
			message.editedAt = input.timestamp ?? Date.now();
			pushEvent({
				kind: "message-edited",
				accountId,
				message: cloneMessage(message)
			});
			return cloneMessage(message);
		},
		deleteMessage(input) {
			const accountId = normalizeAccountId(input.accountId);
			const message = messages.get(input.messageId);
			if (!message) throw new Error(`qa-bus message not found: ${input.messageId}`);
			message.deleted = true;
			pushEvent({
				kind: "message-deleted",
				accountId,
				message: cloneMessage(message)
			});
			return cloneMessage(message);
		},
		readMessage(input) {
			return readQaBusMessage({
				messages,
				input
			});
		},
		searchMessages(input) {
			return searchQaBusMessages({
				messages,
				input
			});
		},
		poll(input = {}) {
			return pollQaBusEvents({
				events,
				cursor,
				input
			});
		},
		async waitFor(input) {
			return await waiters.waitFor(input);
		}
	};
}
//#endregion
//#region extensions/qa-lab/src/harness-runtime.ts
function createQaRunnerRuntime() {
	const sessions = /* @__PURE__ */ new Map();
	return { channel: {
		routing: { resolveAgentRoute({ accountId, peer }) {
			return {
				agentId: "qa-agent",
				accountId: accountId ?? "default",
				sessionKey: `qa-agent:${peer?.kind ?? "direct"}:${peer?.id ?? "default"}`,
				mainSessionKey: "qa-agent:main",
				lastRoutePolicy: "session",
				matchedBy: "default",
				channel: "qa-channel"
			};
		} },
		session: {
			resolveStorePath(_store, { agentId }) {
				return agentId;
			},
			readSessionUpdatedAt({ sessionKey }) {
				return sessions.has(sessionKey) ? Date.now() : void 0;
			},
			recordInboundSession({ sessionKey, ctx }) {
				sessions.set(sessionKey, {
					sessionKey,
					body: ctx.BodyForAgent ?? ctx.Body ?? ""
				});
			}
		},
		reply: {
			resolveEnvelopeFormatOptions() {
				return {};
			},
			formatAgentEnvelope({ body }) {
				return body;
			},
			finalizeInboundContext(ctx) {
				return ctx;
			},
			async dispatchReplyWithBufferedBlockDispatcher({ ctx, dispatcherOptions }) {
				await dispatcherOptions.deliver({ text: `qa-echo: ${ctx.BodyForAgent ?? ctx.Body ?? ""}` });
			}
		}
	} };
}
//#endregion
//#region extensions/qa-lab/src/scenario.ts
async function runQaScenario(definition, ctx) {
	const steps = [];
	for (const step of definition.steps) try {
		const details = await step.run(ctx);
		steps.push({
			name: step.name,
			status: "pass",
			...details ? { details } : {}
		});
	} catch (error) {
		const details = formatErrorMessage(error);
		steps.push({
			name: step.name,
			status: "fail",
			details
		});
		return {
			name: definition.name,
			status: "fail",
			steps,
			details
		};
	}
	return {
		name: definition.name,
		status: "pass",
		steps
	};
}
//#endregion
//#region extensions/qa-lab/src/self-check-scenario.ts
function createQaSelfCheckScenario(cfg) {
	return {
		name: "Synthetic Slack-class roundtrip",
		steps: [
			{
				name: "DM echo roundtrip",
				async run({ state }) {
					state.addInboundMessage({
						conversation: {
							id: "alice",
							kind: "direct"
						},
						senderId: "alice",
						senderName: "Alice",
						text: "hello from qa"
					});
					await state.waitFor({
						kind: "message-text",
						textIncludes: "qa-echo: hello from qa",
						direction: "outbound",
						timeoutMs: 5e3
					});
				}
			},
			{
				name: "Thread create and threaded echo",
				async run({ state }) {
					const threadId = extractToolPayload(await qaChannelPlugin.actions?.handleAction?.({
						channel: "qa-channel",
						action: "thread-create",
						cfg,
						accountId: "default",
						params: {
							channelId: "qa-room",
							title: "QA thread"
						}
					}))?.thread?.id;
					if (!threadId) throw new Error("thread-create did not return thread id");
					state.addInboundMessage({
						conversation: {
							id: "qa-room",
							kind: "channel",
							title: "QA Room"
						},
						senderId: "alice",
						senderName: "Alice",
						text: "inside thread",
						threadId,
						threadTitle: "QA thread"
					});
					await state.waitFor({
						kind: "message-text",
						textIncludes: "qa-echo: inside thread",
						direction: "outbound",
						timeoutMs: 5e3
					});
					return threadId;
				}
			},
			{
				name: "Reaction, edit, delete lifecycle",
				async run({ state }) {
					const outbound = state.searchMessages({
						query: "qa-echo: inside thread",
						conversationId: "qa-room"
					}).at(-1);
					if (!outbound) throw new Error("threaded outbound message not found");
					await qaChannelPlugin.actions?.handleAction?.({
						channel: "qa-channel",
						action: "react",
						cfg,
						accountId: "default",
						params: {
							messageId: outbound.id,
							emoji: "white_check_mark"
						}
					});
					if (state.readMessage({ messageId: outbound.id }).reactions.length === 0) throw new Error("reaction not recorded");
					await qaChannelPlugin.actions?.handleAction?.({
						channel: "qa-channel",
						action: "edit",
						cfg,
						accountId: "default",
						params: {
							messageId: outbound.id,
							text: "qa-echo: inside thread (edited)"
						}
					});
					if (!state.readMessage({ messageId: outbound.id }).text.includes("(edited)")) throw new Error("edit not recorded");
					await qaChannelPlugin.actions?.handleAction?.({
						channel: "qa-channel",
						action: "delete",
						cfg,
						accountId: "default",
						params: { messageId: outbound.id }
					});
					if (!state.readMessage({ messageId: outbound.id }).deleted) throw new Error("delete not recorded");
				}
			}
		]
	};
}
//#endregion
//#region extensions/qa-lab/src/self-check.ts
function resolveQaSelfCheckOutputPath(params) {
	if (params?.outputPath) return params.outputPath;
	const repoRoot = path.resolve(params?.repoRoot ?? process.cwd());
	return path.join(repoRoot, ".artifacts", "qa-e2e", "self-check.md");
}
async function runQaSelfCheckAgainstState(params) {
	const startedAt = /* @__PURE__ */ new Date();
	params.state.reset();
	const scenarioResult = await runQaScenario(createQaSelfCheckScenario(params.cfg), { state: params.state });
	const checks = [{
		name: "QA self-check scenario",
		status: scenarioResult.status,
		details: `${scenarioResult.steps.filter((step) => step.status === "pass").length}/${scenarioResult.steps.length} steps passed`
	}];
	const finishedAt = /* @__PURE__ */ new Date();
	const timeline = params.state.getSnapshot().events.map((event) => {
		switch (event.kind) {
			case "thread-created": return `${event.cursor}. ${event.kind} ${event.thread.conversationId}/${event.thread.id}`;
			case "reaction-added": return `${event.cursor}. ${event.kind} ${event.message.id} ${event.emoji}`;
			default: return `${event.cursor}. ${event.kind} ${"message" in event ? event.message.id : ""}`.trim();
		}
	});
	const report = renderQaMarkdownReport({
		title: "OpenClaw QA E2E Self-Check",
		startedAt,
		finishedAt,
		checks,
		scenarios: [{
			name: scenarioResult.name,
			status: scenarioResult.status,
			details: scenarioResult.details,
			steps: scenarioResult.steps
		}],
		timeline,
		notes: params.notes ?? ["Vertical slice: qa-channel + qa-lab bus + private debugger surface.", "Docker orchestration, matrix runs, and auto-fix loops remain follow-up work."]
	});
	const outputPath = resolveQaSelfCheckOutputPath({
		outputPath: params.outputPath,
		repoRoot: params.repoRoot
	});
	await fs$1.mkdir(path.dirname(outputPath), { recursive: true });
	await fs$1.writeFile(outputPath, report, "utf8");
	return {
		outputPath,
		report,
		checks,
		scenarioResult
	};
}
//#endregion
//#region extensions/qa-lab/src/lab-server.ts
const CAPTURE_QUERY_PRESETS = new Set([
	"double-sends",
	"retry-storms",
	"cache-busting",
	"ws-duplicate-frames",
	"missing-ack",
	"error-bursts"
]);
function isCaptureQueryPreset(value) {
	return CAPTURE_QUERY_PRESETS.has(value);
}
function parseCaptureMeta(metaJson) {
	if (typeof metaJson !== "string" || metaJson.trim().length === 0) return null;
	try {
		const parsed = JSON.parse(metaJson);
		return parsed && typeof parsed === "object" ? parsed : null;
	} catch {
		return null;
	}
}
function readCaptureMetaString(meta, key) {
	const value = meta?.[key];
	return typeof value === "string" && value.trim().length > 0 ? value.trim() : void 0;
}
function mapCaptureEventForQa(row) {
	const meta = parseCaptureMeta(row.metaJson);
	return {
		...row,
		payloadPreview: typeof row.dataText === "string" ? row.dataText : void 0,
		provider: readCaptureMetaString(meta, "provider"),
		api: readCaptureMetaString(meta, "api"),
		model: readCaptureMetaString(meta, "model"),
		captureOrigin: readCaptureMetaString(meta, "captureOrigin")
	};
}
function defaultPortForProtocol(protocol) {
	if (protocol === "https:") return 443;
	if (protocol === "http:") return 80;
	return 0;
}
async function probeTcpReachability(rawUrl, timeoutMs = 700) {
	let parsed;
	try {
		parsed = new URL(rawUrl);
	} catch {
		return {
			label: rawUrl,
			url: rawUrl,
			ok: false,
			error: "invalid url"
		};
	}
	const host = parsed.hostname;
	const port = parsed.port ? Number(parsed.port) : defaultPortForProtocol(parsed.protocol);
	if (!host || !Number.isFinite(port) || port <= 0) return {
		label: parsed.origin,
		url: parsed.toString(),
		ok: false,
		error: "missing host or port"
	};
	try {
		await new Promise((resolve, reject) => {
			const socket = net.createConnection({
				host,
				port
			});
			const onError = (error) => {
				socket.destroy();
				reject(error);
			};
			socket.setTimeout(timeoutMs, () => {
				socket.destroy(/* @__PURE__ */ new Error("timeout"));
			});
			socket.once("connect", () => {
				socket.end();
				resolve();
			});
			socket.once("error", onError);
			socket.once("timeout", () => onError(/* @__PURE__ */ new Error("timeout")));
		});
		return {
			label: parsed.host,
			url: parsed.toString(),
			ok: true
		};
	} catch (error) {
		return {
			label: parsed.host,
			url: parsed.toString(),
			ok: false,
			error: formatErrorMessage(error)
		};
	}
}
function countQaLabScenarioRun(scenarios) {
	return {
		total: scenarios.length,
		pending: scenarios.filter((scenario) => scenario.status === "pending").length,
		running: scenarios.filter((scenario) => scenario.status === "running").length,
		passed: scenarios.filter((scenario) => scenario.status === "pass").length,
		failed: scenarios.filter((scenario) => scenario.status === "fail").length,
		skipped: scenarios.filter((scenario) => scenario.status === "skip").length
	};
}
function withQaLabRunCounts(run) {
	return {
		...run,
		counts: countQaLabScenarioRun(run.scenarios)
	};
}
function injectKickoffMessage(params) {
	return params.state.addInboundMessage({
		conversation: {
			id: params.defaults.conversationId,
			kind: params.defaults.conversationKind,
			...params.defaults.conversationKind === "channel" ? { title: params.defaults.conversationId } : {}
		},
		senderId: params.defaults.senderId,
		senderName: params.defaults.senderName,
		text: params.kickoffTask
	});
}
async function readJson(req) {
	const chunks = [];
	for await (const chunk of req) chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
	const text = Buffer.concat(chunks).toString("utf8").trim();
	return text ? JSON.parse(text) : {};
}
function detectContentType(filePath) {
	if (filePath.endsWith(".css")) return "text/css; charset=utf-8";
	if (filePath.endsWith(".js")) return "text/javascript; charset=utf-8";
	if (filePath.endsWith(".json")) return "application/json; charset=utf-8";
	if (filePath.endsWith(".svg")) return "image/svg+xml";
	return "text/html; charset=utf-8";
}
function missingUiHtml() {
	return `<!doctype html>
<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <title>QA Lab UI Missing</title>
    <style>
      body { font-family: ui-sans-serif, system-ui, sans-serif; background: #0f1115; color: #f5f7fb; margin: 0; display: grid; place-items: center; min-height: 100vh; }
      main { max-width: 42rem; padding: 2rem; background: #171b22; border: 1px solid #283140; border-radius: 18px; box-shadow: 0 30px 80px rgba(0,0,0,.35); }
      code { font-family: ui-monospace, SFMono-Regular, Menlo, monospace; color: #9ee8d8; }
      h1 { margin-top: 0; }
    </style>
  </head>
  <body>
    <main>
      <h1>QA Lab UI not built</h1>
      <p>Build the private debugger bundle, then reload this page.</p>
      <p><code>pnpm qa:lab:build</code></p>
    </main>
  </body>
</html>`;
}
function resolveUiDistDir(overrideDir, repoRoot = process.cwd()) {
	if (overrideDir?.trim()) return overrideDir;
	const candidates = [
		path.resolve(repoRoot, "extensions/qa-lab/web/dist"),
		path.resolve(repoRoot, "dist/extensions/qa-lab/web/dist"),
		fileURLToPath(new URL("../web/dist", import.meta.url))
	];
	return candidates.find((candidate) => {
		if (!fs.existsSync(candidate)) return false;
		const indexPath = path.join(candidate, "index.html");
		return fs.existsSync(indexPath) && fs.statSync(indexPath).isFile();
	}) ?? candidates[0];
}
function listUiAssetFiles(rootDir, currentDir = rootDir) {
	const entries = fs.readdirSync(currentDir, { withFileTypes: true }).toSorted((left, right) => left.name.localeCompare(right.name));
	const files = [];
	for (const entry of entries) {
		const resolved = path.join(currentDir, entry.name);
		if (entry.isDirectory()) {
			files.push(...listUiAssetFiles(rootDir, resolved));
			continue;
		}
		if (!entry.isFile()) continue;
		files.push(path.relative(rootDir, resolved));
	}
	return files;
}
function resolveUiAssetVersion(overrideDir) {
	try {
		const distDir = resolveUiDistDir(overrideDir);
		const indexPath = path.join(distDir, "index.html");
		if (!fs.existsSync(indexPath) || !fs.statSync(indexPath).isFile()) return null;
		const hash = createHash("sha1");
		for (const relativeFile of listUiAssetFiles(distDir)) {
			hash.update(relativeFile);
			hash.update("\0");
			hash.update(fs.readFileSync(path.join(distDir, relativeFile)));
			hash.update("\0");
		}
		return hash.digest("hex").slice(0, 12);
	} catch {
		return null;
	}
}
function resolveAdvertisedBaseUrl(params) {
	return `http://${params.advertiseHost?.trim() || (params.bindHost && params.bindHost !== "0.0.0.0" ? params.bindHost : "127.0.0.1")}:${typeof params.advertisePort === "number" && Number.isFinite(params.advertisePort) ? params.advertisePort : params.bindPort}`;
}
function createBootstrapDefaults(autoKickoffTarget) {
	if (autoKickoffTarget === "channel") return {
		conversationKind: "channel",
		conversationId: "qa-lab",
		senderId: "qa-operator",
		senderName: "QA Operator"
	};
	return {
		conversationKind: "direct",
		conversationId: "qa-operator",
		senderId: "qa-operator",
		senderName: "QA Operator"
	};
}
function isControlUiProxyPath(pathname) {
	return pathname === "/control-ui" || pathname.startsWith("/control-ui/");
}
function rewriteControlUiProxyPath(pathname, search) {
	return `${pathname === "/control-ui" ? "/" : pathname.slice(11) || "/"}${search}`;
}
function rewriteEmbeddedControlUiHeaders(headers) {
	const rewritten = { ...headers };
	delete rewritten["x-frame-options"];
	const csp = headers["content-security-policy"];
	if (typeof csp === "string") rewritten["content-security-policy"] = csp.includes("frame-ancestors") ? csp.replace(/frame-ancestors\s+[^;]+/i, "frame-ancestors 'self'") : `${csp}; frame-ancestors 'self'`;
	return rewritten;
}
async function proxyHttpRequest(params) {
	const upstreamReq = (params.target.protocol === "https:" ? request$1 : request)({
		protocol: params.target.protocol,
		hostname: params.target.hostname,
		port: params.target.port || (params.target.protocol === "https:" ? 443 : 80),
		method: params.req.method,
		path: rewriteControlUiProxyPath(params.pathname, params.search),
		headers: {
			...params.req.headers,
			host: params.target.host
		}
	}, (upstreamRes) => {
		params.res.writeHead(upstreamRes.statusCode ?? 502, rewriteEmbeddedControlUiHeaders(upstreamRes.headers));
		upstreamRes.pipe(params.res);
	});
	upstreamReq.on("error", (error) => {
		if (!params.res.headersSent) {
			writeError(params.res, 502, error);
			return;
		}
		params.res.destroy(error);
	});
	if (params.req.method === "GET" || params.req.method === "HEAD") {
		upstreamReq.end();
		return;
	}
	params.req.pipe(upstreamReq);
}
function proxyUpgradeRequest(params) {
	const requestUrl = new URL(params.req.url ?? "/", "http://127.0.0.1");
	const port = Number(params.target.port || (params.target.protocol === "https:" ? 443 : 80));
	const upstream = params.target.protocol === "https:" ? tls.connect({
		host: params.target.hostname,
		port,
		servername: params.target.hostname
	}) : net.connect({
		host: params.target.hostname,
		port
	});
	const headerLines = [];
	for (let index = 0; index < params.req.rawHeaders.length; index += 2) {
		const name = params.req.rawHeaders[index];
		const value = params.req.rawHeaders[index + 1] ?? "";
		if (normalizeLowercaseStringOrEmpty(name) === "host") continue;
		headerLines.push(`${name}: ${value}`);
	}
	upstream.once("connect", () => {
		const requestText = [
			`${params.req.method ?? "GET"} ${rewriteControlUiProxyPath(requestUrl.pathname, requestUrl.search)} HTTP/${params.req.httpVersion}`,
			`Host: ${params.target.host}`,
			...headerLines,
			"",
			""
		].join("\r\n");
		upstream.write(requestText);
		if (params.head.length > 0) upstream.write(params.head);
		upstream.pipe(params.socket);
		params.socket.pipe(upstream);
	});
	const closeBoth = () => {
		if (!params.socket.destroyed) params.socket.destroy();
		if (!upstream.destroyed) upstream.destroy();
	};
	upstream.on("error", () => {
		if (!params.socket.destroyed) params.socket.write("HTTP/1.1 502 Bad Gateway\r\nConnection: close\r\n\r\n");
		closeBoth();
	});
	params.socket.on("error", closeBoth);
	params.socket.on("close", closeBoth);
}
function tryResolveUiAsset(pathname, overrideDir, repoRoot = process.cwd()) {
	const distDir = resolveUiDistDir(overrideDir, repoRoot);
	if (!fs.existsSync(distDir)) return null;
	const decoded = decodeURIComponent(pathname === "/" ? "/index.html" : pathname);
	const candidate = path.resolve(distDir, `.${decoded.startsWith("/") ? decoded : `/${decoded}`}`);
	const relative = path.relative(distDir, candidate);
	if (relative.startsWith("..") || path.isAbsolute(relative)) return null;
	if (fs.existsSync(candidate) && fs.statSync(candidate).isFile()) return candidate;
	const fallback = path.join(distDir, "index.html");
	return fs.existsSync(fallback) ? fallback : null;
}
function createQaLabConfig(baseUrl) {
	return { channels: { "qa-channel": {
		enabled: true,
		baseUrl,
		botUserId: "openclaw",
		botDisplayName: "OpenClaw QA",
		allowFrom: ["*"]
	} } };
}
async function startQaGatewayLoop(params) {
	setQaChannelRuntime(createQaRunnerRuntime());
	const cfg = createQaLabConfig(params.baseUrl);
	const account = qaChannelPlugin.config.resolveAccount(cfg, "default");
	const abort = new AbortController();
	const task = qaChannelPlugin.gateway?.startAccount?.({
		accountId: account.accountId,
		account,
		cfg,
		runtime: {
			log: () => void 0,
			error: () => void 0,
			exit: () => void 0
		},
		abortSignal: abort.signal,
		log: {
			info: () => void 0,
			warn: () => void 0,
			error: () => void 0,
			debug: () => void 0
		},
		getStatus: () => ({
			accountId: account.accountId,
			configured: true,
			enabled: true,
			running: true
		}),
		setStatus: () => void 0
	});
	return {
		cfg,
		async stop() {
			abort.abort();
			await task;
		}
	};
}
async function startQaLabServer(params) {
	const repoRoot = path.resolve(params?.repoRoot ?? process.cwd());
	const captureSettings = resolveDebugProxySettings();
	const captureStore = getDebugProxyCaptureStore(captureSettings.dbPath, captureSettings.blobDir);
	const state = createQaBusState();
	let latestReport = null;
	let latestScenarioRun = null;
	const scenarioCatalog = readQaBootstrapScenarioCatalog();
	const bootstrapDefaults = createBootstrapDefaults(params?.autoKickoffTarget);
	let runnerModelOptions = [];
	let runnerModelCatalogStatus = "loading";
	let runnerSnapshot = createIdleQaRunnerSnapshot(scenarioCatalog.scenarios);
	let activeSuiteRun = null;
	let controlUiProxyTarget = params?.controlUiProxyTarget?.trim() ? new URL(params.controlUiProxyTarget) : null;
	let controlUiUrl = params?.controlUiUrl?.trim() || null;
	let controlUiToken = params?.controlUiToken?.trim() || null;
	let gateway;
	const embeddedGatewayEnabled = params?.embeddedGateway !== "disabled";
	let labHandle = null;
	let publicBaseUrl = "";
	let runnerModelCatalogPromise = null;
	let runnerModelCatalogAbort = null;
	const ensureRunnerModelCatalog = () => {
		if (runnerModelCatalogPromise) return runnerModelCatalogPromise;
		runnerModelCatalogAbort = new AbortController();
		runnerModelCatalogPromise = (async () => {
			try {
				const { loadQaRunnerModelOptions } = await import("./model-catalog.runtime-lRR01ti-.js");
				runnerModelOptions = await loadQaRunnerModelOptions({
					repoRoot,
					signal: runnerModelCatalogAbort?.signal
				});
				runnerModelCatalogStatus = "ready";
			} catch {
				runnerModelOptions = [];
				runnerModelCatalogStatus = "failed";
			}
		})().finally(() => {
			runnerModelCatalogAbort = null;
		});
		return runnerModelCatalogPromise;
	};
	async function runSelfCheck() {
		latestScenarioRun = withQaLabRunCounts({
			kind: "self-check",
			status: "running",
			startedAt: (/* @__PURE__ */ new Date()).toISOString(),
			scenarios: [{
				id: "qa-self-check",
				name: "Synthetic Slack-class roundtrip",
				status: "running"
			}]
		});
		const result = await runQaSelfCheckAgainstState({
			state,
			cfg: gateway?.cfg ?? createQaLabConfig(listenUrl),
			outputPath: params?.outputPath,
			repoRoot
		});
		latestScenarioRun = withQaLabRunCounts({
			kind: "self-check",
			status: "completed",
			startedAt: latestScenarioRun.startedAt,
			finishedAt: (/* @__PURE__ */ new Date()).toISOString(),
			scenarios: [{
				id: "qa-self-check",
				name: result.scenarioResult.name,
				status: result.scenarioResult.status,
				details: result.scenarioResult.details,
				steps: result.scenarioResult.steps
			}]
		});
		latestReport = {
			outputPath: result.outputPath,
			markdown: result.report,
			generatedAt: (/* @__PURE__ */ new Date()).toISOString()
		};
		return result;
	}
	const server = createServer$1(async (req, res) => {
		const url = new URL(req.url ?? "/", "http://127.0.0.1");
		if (await handleQaBusRequest({
			req,
			res,
			state
		})) return;
		try {
			if (controlUiProxyTarget && isControlUiProxyPath(url.pathname)) {
				await proxyHttpRequest({
					req,
					res,
					target: controlUiProxyTarget,
					pathname: url.pathname,
					search: url.search
				});
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/bootstrap") {
				ensureRunnerModelCatalog();
				const resolvedControlUiUrl = controlUiProxyTarget ? `${publicBaseUrl}/control-ui/` : controlUiUrl;
				const controlUiEmbeddedUrl = resolvedControlUiUrl && controlUiToken ? `${resolvedControlUiUrl.replace(/\/?$/, "/")}#token=${encodeURIComponent(controlUiToken)}` : resolvedControlUiUrl;
				writeJson(res, 200, {
					baseUrl: publicBaseUrl,
					latestReport,
					controlUiUrl: resolvedControlUiUrl,
					controlUiEmbeddedUrl,
					kickoffTask: scenarioCatalog.kickoffTask,
					scenarios: scenarioCatalog.scenarios,
					defaults: bootstrapDefaults,
					runner: runnerSnapshot,
					runnerCatalog: {
						status: runnerModelCatalogStatus,
						real: runnerModelOptions
					}
				});
				return;
			}
			if (req.method === "GET" && (url.pathname === "/healthz" || url.pathname === "/readyz")) {
				writeJson(res, 200, {
					ok: true,
					status: "live"
				});
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/state") {
				writeJson(res, 200, state.getSnapshot());
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/report") {
				writeJson(res, 200, { report: latestReport });
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/ui-version") {
				res.writeHead(200, {
					"content-type": "application/json; charset=utf-8",
					"cache-control": "no-store"
				});
				res.end(JSON.stringify({ version: resolveUiAssetVersion(params?.uiDistDir) }));
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/outcomes") {
				writeJson(res, 200, { run: latestScenarioRun });
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/capture/sessions") {
				writeJson(res, 200, { sessions: captureStore.listSessions(50) });
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/capture/startup-status") {
				const proxyUrl = captureSettings.proxyUrl || "http://127.0.0.1:7799";
				const gatewayUrl = controlUiUrl || "http://127.0.0.1:18789/";
				const [proxy, gateway] = await Promise.all([probeTcpReachability(proxyUrl), probeTcpReachability(gatewayUrl)]);
				writeJson(res, 200, { status: {
					proxy: {
						...proxy,
						label: "Proxy"
					},
					gateway: {
						...gateway,
						label: "Gateway"
					},
					qaLab: {
						label: "QA Lab",
						url: publicBaseUrl,
						ok: true
					}
				} });
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/capture/events") {
				const sessionId = url.searchParams.get("sessionId")?.trim();
				writeJson(res, 200, { events: sessionId ? captureStore.getSessionEvents(sessionId, 200).map(mapCaptureEventForQa) : [] });
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/capture/coverage") {
				const sessionId = url.searchParams.get("sessionId")?.trim();
				if (!sessionId) {
					writeError(res, 400, "Missing sessionId");
					return;
				}
				writeJson(res, 200, { coverage: captureStore.summarizeSessionCoverage(sessionId) });
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/capture/query") {
				const preset = url.searchParams.get("preset")?.trim();
				const sessionId = url.searchParams.get("sessionId")?.trim() || void 0;
				if (!preset) {
					writeError(res, 400, "Missing preset");
					return;
				}
				if (!isCaptureQueryPreset(preset)) {
					writeError(res, 400, "Unknown preset");
					return;
				}
				writeJson(res, 200, { rows: captureStore.queryPreset(preset, sessionId) });
				return;
			}
			if (req.method === "GET" && url.pathname === "/api/capture/blob") {
				const blobId = url.searchParams.get("id")?.trim();
				if (!blobId) {
					writeError(res, 400, "Missing blob id");
					return;
				}
				const content = captureStore.readBlob(blobId);
				if (content == null) {
					writeError(res, 404, "Blob not found");
					return;
				}
				writeJson(res, 200, {
					id: blobId,
					content
				});
				return;
			}
			if (req.method === "POST" && url.pathname === "/api/capture/delete-sessions") {
				const body = await readJson(req);
				const sessionIds = Array.isArray(body.sessionIds) ? body.sessionIds.filter((value) => typeof value === "string") : [];
				writeJson(res, 200, { result: captureStore.deleteSessions(sessionIds) });
				return;
			}
			if (req.method === "POST" && url.pathname === "/api/capture/purge") {
				writeJson(res, 200, { result: captureStore.purgeAll() });
				return;
			}
			if (req.method === "POST" && url.pathname === "/api/reset") {
				if (activeSuiteRun) {
					writeError(res, 409, "QA suite run already in progress");
					return;
				}
				state.reset();
				latestReport = null;
				latestScenarioRun = null;
				runnerSnapshot = {
					...runnerSnapshot,
					status: "idle",
					artifacts: null,
					error: null,
					startedAt: void 0,
					finishedAt: void 0
				};
				writeJson(res, 200, { ok: true });
				return;
			}
			if (req.method === "POST" && url.pathname === "/api/inbound/message") {
				const body = await readJson(req);
				writeJson(res, 200, { message: state.addInboundMessage(body) });
				return;
			}
			if (req.method === "POST" && url.pathname === "/api/kickoff") {
				writeJson(res, 200, { message: injectKickoffMessage({
					state,
					defaults: bootstrapDefaults,
					kickoffTask: scenarioCatalog.kickoffTask
				}) });
				return;
			}
			if (req.method === "POST" && url.pathname === "/api/scenario/self-check") {
				if (activeSuiteRun) {
					writeError(res, 409, "QA suite run already in progress");
					return;
				}
				writeJson(res, 200, serializeSelfCheck(await runSelfCheck()));
				return;
			}
			if (req.method === "POST" && url.pathname === "/api/scenario/suite") {
				if (activeSuiteRun) {
					writeError(res, 409, "QA suite run already in progress");
					return;
				}
				const selection = normalizeQaRunSelection(await readJson(req), scenarioCatalog.scenarios);
				state.reset();
				latestReport = null;
				latestScenarioRun = null;
				const startedAt = (/* @__PURE__ */ new Date()).toISOString();
				runnerSnapshot = {
					status: "running",
					selection,
					startedAt,
					finishedAt: void 0,
					artifacts: null,
					error: null
				};
				activeSuiteRun = (async () => {
					try {
						const { runQaSuite } = await import("./suite-DnvG1Tyv.js");
						const result = await runQaSuite({
							lab: labHandle ?? void 0,
							outputDir: createQaRunOutputDir(repoRoot),
							providerMode: selection.providerMode,
							primaryModel: selection.primaryModel,
							alternateModel: selection.alternateModel,
							scenarioIds: selection.scenarioIds
						});
						runnerSnapshot = {
							status: "completed",
							selection,
							startedAt,
							finishedAt: (/* @__PURE__ */ new Date()).toISOString(),
							artifacts: {
								outputDir: result.outputDir,
								reportPath: result.reportPath,
								summaryPath: result.summaryPath,
								watchUrl: result.watchUrl
							},
							error: null
						};
					} catch (error) {
						runnerSnapshot = {
							status: "failed",
							selection,
							startedAt,
							finishedAt: (/* @__PURE__ */ new Date()).toISOString(),
							artifacts: null,
							error: formatErrorMessage(error)
						};
					} finally {
						activeSuiteRun = null;
					}
				})();
				writeJson(res, 202, {
					ok: true,
					runner: runnerSnapshot
				});
				return;
			}
			if (req.method !== "GET" && req.method !== "HEAD") {
				writeError(res, 404, "not found");
				return;
			}
			const asset = tryResolveUiAsset(url.pathname, params?.uiDistDir, repoRoot);
			if (!asset) {
				const html = missingUiHtml();
				res.writeHead(200, {
					"content-type": "text/html; charset=utf-8",
					"content-length": Buffer.byteLength(html)
				});
				if (req.method === "HEAD") {
					res.end();
					return;
				}
				res.end(html);
				return;
			}
			const body = fs.readFileSync(asset);
			res.writeHead(200, {
				"content-type": detectContentType(asset),
				"content-length": body.byteLength
			});
			if (req.method === "HEAD") {
				res.end();
				return;
			}
			res.end(body);
		} catch (error) {
			writeError(res, 500, error);
		}
	});
	await new Promise((resolve, reject) => {
		server.once("error", reject);
		server.listen(params?.port ?? 0, params?.host ?? "127.0.0.1", () => resolve());
	});
	const address = server.address();
	if (!address || typeof address === "string") throw new Error("qa-lab failed to bind");
	const listenUrl = resolveAdvertisedBaseUrl({
		bindHost: params?.host ?? "127.0.0.1",
		bindPort: address.port
	});
	publicBaseUrl = resolveAdvertisedBaseUrl({
		bindHost: params?.host ?? "127.0.0.1",
		bindPort: address.port,
		advertiseHost: params?.advertiseHost,
		advertisePort: params?.advertisePort
	});
	if (embeddedGatewayEnabled) gateway = await startQaGatewayLoop({
		state,
		baseUrl: listenUrl
	});
	if (params?.sendKickoffOnStart) injectKickoffMessage({
		state,
		defaults: bootstrapDefaults,
		kickoffTask: scenarioCatalog.kickoffTask
	});
	server.on("upgrade", (req, socket, head) => {
		const url = new URL(req.url ?? "/", "http://127.0.0.1");
		if (!controlUiProxyTarget || !isControlUiProxyPath(url.pathname)) {
			socket.destroy();
			return;
		}
		proxyUpgradeRequest({
			req,
			socket,
			head,
			target: controlUiProxyTarget
		});
	});
	const lab = {
		baseUrl: publicBaseUrl,
		listenUrl,
		state,
		setControlUi(next) {
			controlUiUrl = next.controlUiUrl?.trim() || null;
			controlUiToken = next.controlUiToken?.trim() || null;
			controlUiProxyTarget = next.controlUiProxyTarget?.trim() ? new URL(next.controlUiProxyTarget) : null;
		},
		setScenarioRun(next) {
			latestScenarioRun = next ? withQaLabRunCounts(next) : null;
		},
		setLatestReport(next) {
			latestReport = next;
		},
		runSelfCheck,
		async stop() {
			runnerModelCatalogAbort?.abort();
			await runnerModelCatalogPromise?.catch(() => void 0);
			await gateway?.stop();
			await closeQaHttpServer(server);
		}
	};
	labHandle = lab;
	return lab;
}
function serializeSelfCheck(result) {
	return {
		outputPath: result.outputPath,
		report: result.report,
		checks: result.checks,
		scenario: result.scenarioResult
	};
}
//#endregion
export { readQaBusMessage as _, runQaScenario as a, DEFAULT_WAIT_TIMEOUT_MS as c, buildQaBusSnapshot as d, cloneEvent as f, pollQaBusEvents as g, normalizeConversationFromTarget as h, createQaSelfCheckScenario as i, createQaBusWaiterStore as l, normalizeAccountId as m, resolveQaSelfCheckOutputPath as n, createQaRunnerRuntime as o, cloneMessage as p, runQaSelfCheckAgainstState as r, createQaBusState as s, startQaLabServer as t, DEFAULT_ACCOUNT_ID as u, searchQaBusMessages as v };
