// openclaw/dist/library-y52fQOhw.js
import { u as normalizeE164 } from "./utils-D5DtWkEu.js";
import { a as loadConfig } from "./io-B-YPX79V.js";
import "./config-M6t0-B38.js";
import { u as resolveStorePath } from "./paths-CrHqfnMp.js";
import { i as handlePortError, n as describePortOwner, r as ensurePortAvailable, t as PortInUseError } from "./ports-CHYP3VY0.js";
import { o as saveSessionStore } from "./store-Cw5RrVCw.js";
import { n as resolveSessionKey, t as deriveSessionKey } from "./session-key-CLSm5qo5.js";
import { t as loadSessionStore } from "./store-load-DsLWM8wp.js";
import { t as applyTemplate } from "./templating-De0Tbhb5.js";
import { t as createDefaultDeps } from "./deps-BFyrfgWM.js";
import { t as waitForever } from "./wait-DuJpzD0a.js";
//#region src/library.ts
let replyRuntimePromise = null;
let promptRuntimePromise = null;
let binariesRuntimePromise = null;
let execRuntimePromise = null;
let webChannelRuntimePromise = null;
function loadReplyRuntime() {
	replyRuntimePromise ??= import("./reply.runtime-DFTH1Skv.js");
	return replyRuntimePromise;
}
function loadPromptRuntime() {
	promptRuntimePromise ??= import("./prompt-C1TF3QEe.js");
	return promptRuntimePromise;
}
function loadBinariesRuntime() {
	binariesRuntimePromise ??= import("./binaries-CxAEQyTj.js");
	return binariesRuntimePromise;
}
function loadExecRuntime() {
	execRuntimePromise ??= import("./exec-DvAkQvu6.js");
	return execRuntimePromise;
}
function loadWebChannelRuntime() {
	webChannelRuntimePromise ??= import("./runtime-web-channel-plugin-DzPjUCX9.js");
	return webChannelRuntimePromise;
}
const getReplyFromConfig = async (...args) => (await loadReplyRuntime()).getReplyFromConfig(...args);
const promptYesNo = async (...args) => (await loadPromptRuntime()).promptYesNo(...args);
const ensureBinary = async (...args) => (await loadBinariesRuntime()).ensureBinary(...args);
const runExec = async (...args) => (await loadExecRuntime()).runExec(...args);
const runCommandWithTimeout = async (...args) => (await loadExecRuntime()).runCommandWithTimeout(...args);
const monitorWebChannel = async (...args) => (await loadWebChannelRuntime()).monitorWebChannel(...args);
//#endregion
export { PortInUseError, applyTemplate, createDefaultDeps, deriveSessionKey, describePortOwner, ensureBinary, ensurePortAvailable, getReplyFromConfig, handlePortError, loadConfig, loadSessionStore, monitorWebChannel, normalizeE164, promptYesNo, resolveSessionKey, resolveStorePath, runCommandWithTimeout, runExec, saveSessionStore, waitForever };
