// openclaw/dist/live-transport-scenarios-Bb9v3b4q.js
import { i as formatErrorMessage } from "./errors-D8p6rxH8.js";
import "./error-runtime-Da0X1TOQ.js";
import { i as normalizeQaProviderMode } from "./run-config-B359hkWD.js";
import { o as startQaGatewayChild, s as startQaMockOpenAiServer } from "./gateway-child-Du_AnSFD.js";
import { t as resolveRepoRelativeOutputDir } from "./cli-paths-C0Fbq3dW.js";
import path from "node:path";
//#region extensions/qa-lab/src/live-transports/shared/live-transport-cli.runtime.ts
function resolveLiveTransportQaRunOptions(opts) {
	return {
		repoRoot: path.resolve(opts.repoRoot ?? process.cwd()),
		outputDir: resolveRepoRelativeOutputDir(path.resolve(opts.repoRoot ?? process.cwd()), opts.outputDir),
		providerMode: opts.providerMode === void 0 ? "live-frontier" : normalizeQaProviderMode(opts.providerMode),
		primaryModel: opts.primaryModel,
		alternateModel: opts.alternateModel,
		fastMode: opts.fastMode,
		scenarioIds: opts.scenarioIds,
		sutAccountId: opts.sutAccountId
	};
}
function printLiveTransportQaArtifacts(laneLabel, artifacts) {
	for (const [label, filePath] of Object.entries(artifacts)) process.stdout.write(`${laneLabel} ${label}: ${filePath}\n`);
}
//#endregion
//#region extensions/qa-lab/src/live-transports/shared/live-lane-helpers.ts
function appendLiveLaneIssue(issues, label, error) {
	issues.push(`${label}: ${formatErrorMessage(error)}`);
}
function buildLiveLaneArtifactsError(params) {
	return [
		params.heading,
		...params.details ?? [],
		"Artifacts:",
		...Object.entries(params.artifacts).map(([label, filePath]) => `- ${label}: ${filePath}`)
	].join("\n");
}
//#endregion
//#region extensions/qa-lab/src/live-transports/shared/live-gateway.runtime.ts
async function stopQaLiveLaneResources(resources) {
	const errors = [];
	try {
		await resources.gateway.stop();
	} catch (error) {
		appendLiveLaneIssue(errors, "gateway stop failed", error);
	}
	if (resources.mock) try {
		await resources.mock.stop();
	} catch (error) {
		appendLiveLaneIssue(errors, "mock provider stop failed", error);
	}
	if (errors.length > 0) throw new Error(`failed to stop QA live lane resources:\n${errors.join("\n")}`);
}
async function startQaLiveLaneGateway(params) {
	const mock = params.providerMode === "mock-openai" ? await startQaMockOpenAiServer({
		host: "127.0.0.1",
		port: 0
	}) : null;
	try {
		const gateway = await startQaGatewayChild({
			repoRoot: params.repoRoot,
			providerBaseUrl: mock ? `${mock.baseUrl}/v1` : void 0,
			qaBusBaseUrl: params.qaBusBaseUrl,
			includeQaChannel: false,
			controlUiAllowedOrigins: params.controlUiAllowedOrigins,
			providerMode: params.providerMode,
			primaryModel: params.primaryModel,
			alternateModel: params.alternateModel,
			fastMode: params.fastMode,
			thinkingDefault: params.thinkingDefault,
			claudeCliAuthMode: params.claudeCliAuthMode,
			controlUiEnabled: params.controlUiEnabled,
			mutateConfig: params.mutateConfig
		});
		return {
			gateway,
			mock,
			async stop() {
				await stopQaLiveLaneResources({
					gateway,
					mock
				});
			}
		};
	} catch (error) {
		await mock?.stop().catch(() => {});
		throw error;
	}
}
const LIVE_TRANSPORT_STANDARD_SCENARIO_ID_SET = new Set([
	{
		id: "canary",
		title: "Transport canary",
		description: "The lane can trigger one known-good reply on the real transport."
	},
	{
		id: "mention-gating",
		title: "Mention gating",
		description: "Messages without the required mention do not trigger a reply."
	},
	{
		id: "allowlist-block",
		title: "Sender allowlist block",
		description: "Non-allowlisted senders do not trigger a reply."
	},
	{
		id: "top-level-reply-shape",
		title: "Top-level reply shape",
		description: "Top-level replies stay top-level when the lane is configured that way."
	},
	{
		id: "restart-resume",
		title: "Restart resume",
		description: "The lane still responds after a gateway restart."
	},
	{
		id: "thread-follow-up",
		title: "Thread follow-up",
		description: "Threaded prompts receive threaded replies with the expected relation metadata."
	},
	{
		id: "thread-isolation",
		title: "Thread isolation",
		description: "Fresh top-level prompts stay out of prior threads."
	},
	{
		id: "reaction-observation",
		title: "Reaction observation",
		description: "Reaction events are observed and normalized correctly."
	},
	{
		id: "help-command",
		title: "Help command",
		description: "The transport-specific help command path replies successfully."
	}
].map((scenario) => scenario.id));
function assertKnownStandardScenarioIds(ids) {
	for (const id of ids) if (!LIVE_TRANSPORT_STANDARD_SCENARIO_ID_SET.has(id)) throw new Error(`unknown live transport standard scenario id: ${id}`);
}
function selectLiveTransportScenarios(params) {
	if (!params.ids || params.ids.length === 0) return [...params.scenarios];
	const requested = new Set(params.ids);
	const selected = params.scenarios.filter((scenario) => params.ids?.includes(scenario.id));
	const missingIds = [...requested].filter((id) => !selected.some((scenario) => scenario.id === id));
	if (missingIds.length > 0) throw new Error(`unknown ${params.laneLabel} QA scenario id(s): ${missingIds.join(", ")}`);
	return selected;
}
function collectLiveTransportStandardScenarioCoverage(params) {
	const coverage = [];
	const seen = /* @__PURE__ */ new Set();
	const append = (id) => {
		if (!id || seen.has(id)) return;
		seen.add(id);
		coverage.push(id);
	};
	assertKnownStandardScenarioIds(params.alwaysOnStandardScenarioIds ?? []);
	for (const id of params.alwaysOnStandardScenarioIds ?? []) append(id);
	for (const scenario of params.scenarios) {
		if (scenario.standardId) assertKnownStandardScenarioIds([scenario.standardId]);
		append(scenario.standardId);
	}
	return coverage;
}
//#endregion
export { buildLiveLaneArtifactsError as a, appendLiveLaneIssue as i, selectLiveTransportScenarios as n, printLiveTransportQaArtifacts as o, startQaLiveLaneGateway as r, resolveLiveTransportQaRunOptions as s, collectLiveTransportStandardScenarioCoverage as t };
