// openclaw/dist/memory-core-bundled-runtime-BcwkfmWM.js
import { i as loadBundledPluginPublicSurfaceModuleSync } from "./facade-loader-kYdvnr1U.js";
//#region src/plugin-sdk/memory-core-bundled-runtime.ts
function loadApiFacadeModule() {
	return loadBundledPluginPublicSurfaceModuleSync({
		dirName: "memory-core",
		artifactBasename: "api.js"
	});
}
function loadRuntimeFacadeModule() {
	return loadBundledPluginPublicSurfaceModuleSync({
		dirName: "memory-core",
		artifactBasename: "runtime-api.js"
	});
}
const createEmbeddingProvider = ((...args) => loadRuntimeFacadeModule().createEmbeddingProvider(...args));
const registerBuiltInMemoryEmbeddingProviders = ((...args) => loadRuntimeFacadeModule().registerBuiltInMemoryEmbeddingProviders(...args));
const removeGroundedShortTermCandidates = ((...args) => loadRuntimeFacadeModule().removeGroundedShortTermCandidates(...args));
const previewGroundedRemMarkdown = ((...args) => loadApiFacadeModule().previewGroundedRemMarkdown(...args));
const writeBackfillDiaryEntries = ((...args) => loadApiFacadeModule().writeBackfillDiaryEntries(...args));
const removeBackfillDiaryEntries = ((...args) => loadApiFacadeModule().removeBackfillDiaryEntries(...args));
//#endregion
export { removeGroundedShortTermCandidates as a, removeBackfillDiaryEntries as i, previewGroundedRemMarkdown as n, writeBackfillDiaryEntries as o, registerBuiltInMemoryEmbeddingProviders as r, createEmbeddingProvider as t };
