// openclaw/dist/provider-auth-CzzkrCf0.js
import { n as ensureAuthProfileStore } from "./store-C5g_fGOG.js";
import "./types.secrets-B4TpMTxm.js";
import "./ref-contract-CG9Ki9BF.js";
import "./provider-env-vars-CSnj2UxZ.js";
import { n as listProfilesForProvider } from "./profiles-DT2g1T-s.js";
import "./repair-BAgVLJtj.js";
import "./model-auth-markers-aQrg9OWc.js";
import { t as resolveEnvApiKey } from "./model-auth-env-CUN3E4LN.js";
import "./provider-auth-helpers-BYWHtohY.js";
import "./provider-auth-input-BpwrQDFP.js";
import "./provider-api-key-auth-Bl3Ex8d2.js";
import { createHash, randomBytes } from "node:crypto";
//#region src/plugin-sdk/oauth-utils.ts
/** Encode a flat object as application/x-www-form-urlencoded form data. */
function toFormUrlEncoded(data) {
	return Object.entries(data).map(([key, value]) => `${encodeURIComponent(key)}=${encodeURIComponent(value)}`).join("&");
}
/** Generate a PKCE verifier/challenge pair suitable for OAuth authorization flows. */
function generatePkceVerifierChallenge() {
	const verifier = randomBytes(32).toString("base64url");
	return {
		verifier,
		challenge: createHash("sha256").update(verifier).digest("base64url")
	};
}
//#endregion
//#region src/plugin-sdk/provider-auth.ts
function isProviderApiKeyConfigured(params) {
	if (resolveEnvApiKey(params.provider)?.apiKey) return true;
	const agentDir = params.agentDir?.trim();
	if (!agentDir) return false;
	return listProfilesForProvider(ensureAuthProfileStore(agentDir, { allowKeychainPrompt: false }), params.provider).length > 0;
}
//#endregion
export { generatePkceVerifierChallenge as n, toFormUrlEncoded as r, isProviderApiKeyConfigured as t };
