// openclaw/dist/reply-failure-CNRsmk_Z.js
import { i as normalizeLowercaseStringOrEmpty } from "./string-coerce-BUSzWgUA.js";
import "./text-runtime-3VPC1sGh.js";
//#region extensions/qa-lab/src/live-timeout.ts
function isAnthropicModel(modelRef) {
	return modelRef.startsWith("anthropic/");
}
function isOpenAiModel(modelRef) {
	return modelRef.startsWith("openai/");
}
function isGptFiveModel(modelRef) {
	return isOpenAiModel(modelRef) && modelRef.slice(7).startsWith("gpt-5");
}
function isClaudeOpusModel(modelRef) {
	return isAnthropicModel(modelRef) && modelRef.includes("claude-opus");
}
function resolveQaLiveTurnTimeoutMs(profile, fallbackMs, modelRef = profile.primaryModel) {
	if (profile.providerMode === "mock-openai") return fallbackMs;
	if (isClaudeOpusModel(modelRef)) return Math.max(fallbackMs, 24e4);
	if (isAnthropicModel(modelRef)) return Math.max(fallbackMs, 18e4);
	if (isGptFiveModel(modelRef)) return Math.max(fallbackMs, 36e4);
	return Math.max(fallbackMs, 12e4);
}
//#endregion
//#region extensions/qa-lab/src/reply-failure.ts
const FAILURE_REPLY_PREFIXES = [
	"⚠️ something went wrong while processing your request.",
	"⚠️ session history got out of sync.",
	"⚠️ session history was corrupted.",
	"⚠️ context overflow",
	"⚠️ message ordering conflict.",
	"⚠️ model login expired on the gateway",
	"⚠️ model login failed on the gateway",
	"⚠️ agent failed before reply:",
	"⚠️ ✉️ message failed",
	"⚠️ no api key found for provider ",
	"⚠️ missing api key for "
];
const VISIBLE_REPLY_LEAK_PATTERNS = [
	/\bchecking thread context\b/i,
	/\bthread context thin\b/i,
	/\bpost a tight progress reply here\b/i,
	/\bposting a coordination nudge\b/i,
	/\bposted a short coordination reply\b/i,
	/\bnot inventing status\b/i
];
function extractQaVisibleReplyLeakText(text) {
	const trimmed = text.trim();
	if (!trimmed) return;
	if (VISIBLE_REPLY_LEAK_PATTERNS.some((pattern) => pattern.test(trimmed))) return trimmed;
}
function extractQaFailureReplyText(text) {
	const trimmed = text.trim();
	if (!trimmed) return;
	const lower = normalizeLowercaseStringOrEmpty(trimmed);
	if (FAILURE_REPLY_PREFIXES.some((prefix) => lower.startsWith(prefix))) return trimmed;
	const visibleReplyLeak = extractQaVisibleReplyLeakText(trimmed);
	if (visibleReplyLeak) return visibleReplyLeak;
}
//#endregion
export { extractQaVisibleReplyLeakText as n, resolveQaLiveTurnTimeoutMs as r, extractQaFailureReplyText as t };
