// openclaw/dist/run-config-B359hkWD.js
import { i as normalizeQaProviderMode$1, t as defaultQaModelForMode$1 } from "./model-selection-ClMg2tje.js";
import path from "node:path";
//#region extensions/qa-lab/src/run-config.ts
function defaultQaModelForMode(mode, alternate = false) {
	return defaultQaModelForMode$1(mode, alternate ? { alternate: true } : void 0);
}
function createDefaultQaRunSelection(scenarios) {
	const providerMode = "mock-openai";
	return {
		providerMode,
		primaryModel: defaultQaModelForMode(providerMode),
		alternateModel: defaultQaModelForMode(providerMode, true),
		fastMode: false,
		scenarioIds: scenarios.map((scenario) => scenario.id)
	};
}
function normalizeQaProviderMode(input) {
	return normalizeQaProviderMode$1(input === "live-frontier" || input === "live-openai" ? input : "mock-openai");
}
function normalizeModel(input, fallback) {
	return (typeof input === "string" ? input.trim() : "") || fallback;
}
function normalizeScenarioIds(input, scenarios) {
	const availableIds = new Set(scenarios.map((scenario) => scenario.id));
	const requestedIds = Array.isArray(input) ? input.map((value) => typeof value === "string" ? value.trim() : "").filter((value) => value.length > 0) : [];
	const selectedIds = requestedIds.filter((id, index) => {
		return availableIds.has(id) && requestedIds.indexOf(id) === index;
	});
	return selectedIds.length > 0 ? selectedIds : scenarios.map((scenario) => scenario.id);
}
function normalizeQaRunSelection(input, scenarios) {
	const payload = input && typeof input === "object" ? input : {};
	const providerMode = normalizeQaProviderMode(payload.providerMode);
	return {
		providerMode,
		primaryModel: normalizeModel(payload.primaryModel, defaultQaModelForMode(providerMode)),
		alternateModel: normalizeModel(payload.alternateModel, defaultQaModelForMode(providerMode, true)),
		fastMode: providerMode === "live-frontier" || payload.fastMode === true,
		scenarioIds: normalizeScenarioIds(payload.scenarioIds, scenarios)
	};
}
function createIdleQaRunnerSnapshot(scenarios) {
	return {
		status: "idle",
		selection: createDefaultQaRunSelection(scenarios),
		artifacts: null,
		error: null
	};
}
function createQaRunOutputDir(baseDir = process.cwd()) {
	const stamp = (/* @__PURE__ */ new Date()).toISOString().replaceAll(":", "").replaceAll(".", "").replace("T", "-");
	return path.join(baseDir, ".artifacts", "qa-e2e", `lab-${stamp}`);
}
//#endregion
export { normalizeQaRunSelection as a, normalizeQaProviderMode as i, createQaRunOutputDir as n, defaultQaModelForMode as r, createIdleQaRunnerSnapshot as t };
