// openclaw/dist/sessions-BWz-crFz.js
import { a as loadConfig } from "./io-B-YPX79V.js";
import { u as resolveStorePath } from "./paths-CrHqfnMp.js";
import "./store-Cw5RrVCw.js";
import { i as resolveMainSessionKey } from "./main-session-D3efLc-S.js";
import "./reset-CSt9R-yj.js";
import "./session-key-CLSm5qo5.js";
import { t as deliveryContextFromSession } from "./delivery-context.shared-OqD5syUc.js";
import { t as loadSessionStore } from "./store-load-DsLWM8wp.js";
import "./session-file-C9xMnbq2.js";
import { t as parseSessionThreadInfo } from "./thread-info-BFSP_iUR.js";
import "./transcript-DcMS3DOB.js";
import "./targets-X-Vp8dFb.js";
//#region src/config/sessions/main-session.runtime.ts
function resolveMainSessionKeyFromConfig() {
	return resolveMainSessionKey(loadConfig());
}
//#endregion
//#region src/config/sessions/delivery-info.ts
function extractDeliveryInfo(sessionKey) {
	const hasRoutableDeliveryContext = (context) => Boolean(context?.channel && context?.to);
	const { baseSessionKey, threadId } = parseSessionThreadInfo(sessionKey);
	if (!sessionKey || !baseSessionKey) return {
		deliveryContext: void 0,
		threadId
	};
	let deliveryContext;
	try {
		const store = loadSessionStore(resolveStorePath(loadConfig().session?.store));
		let entry = store[sessionKey];
		let storedDeliveryContext = deliveryContextFromSession(entry);
		if (!hasRoutableDeliveryContext(storedDeliveryContext) && baseSessionKey !== sessionKey) {
			entry = store[baseSessionKey];
			storedDeliveryContext = deliveryContextFromSession(entry);
		}
		if (hasRoutableDeliveryContext(storedDeliveryContext)) deliveryContext = {
			channel: storedDeliveryContext.channel,
			to: storedDeliveryContext.to,
			accountId: storedDeliveryContext.accountId,
			threadId: storedDeliveryContext.threadId != null ? String(storedDeliveryContext.threadId) : void 0
		};
	} catch {}
	return {
		deliveryContext,
		threadId
	};
}
//#endregion
export { resolveMainSessionKeyFromConfig as n, extractDeliveryInfo as t };
