// openclaw/dist/slash-dispatch.runtime-zFnWGl3W.js
import { t as resolveConversationLabel$1 } from "./conversation-label-gBggcQtH.js";
import { s as resolveChunkMode$1 } from "./chunk-BJOSq3sp.js";
import { i as resolveAgentRoute$1 } from "./resolve-route-C1CfL-zm.js";
import "./routing-Ewkev_pm.js";
import { t as finalizeInboundContext$1 } from "./inbound-context-BM3S6N92.js";
import { n as dispatchReplyWithDispatcher$1 } from "./provider-dispatcher-Bq4ctOM5.js";
import { t as resolveMarkdownTableMode$1 } from "./markdown-tables-BYk02T3e.js";
import "./config-runtime-Hrpwdk2u.js";
import "./reply-runtime-B2AVTFFN.js";
import { t as recordInboundSessionMetaSafe$1 } from "./conversation-runtime-Du8VNJpk.js";
import { r as deliverSlackSlashReplies$1 } from "./replies-CVaKPoqP.js";
//#region extensions/slack/src/monitor/slash-dispatch.runtime.ts
function resolveChunkMode(...args) {
	return resolveChunkMode$1(...args);
}
function finalizeInboundContext(...args) {
	return finalizeInboundContext$1(...args);
}
function dispatchReplyWithDispatcher(...args) {
	return dispatchReplyWithDispatcher$1(...args);
}
function resolveConversationLabel(...args) {
	return resolveConversationLabel$1(...args);
}
function recordInboundSessionMetaSafe(...args) {
	return recordInboundSessionMetaSafe$1(...args);
}
function resolveMarkdownTableMode(...args) {
	return resolveMarkdownTableMode$1(...args);
}
function resolveAgentRoute(...args) {
	return resolveAgentRoute$1(...args);
}
function deliverSlackSlashReplies(...args) {
	return deliverSlackSlashReplies$1(...args);
}
//#endregion
export { deliverSlackSlashReplies, dispatchReplyWithDispatcher, finalizeInboundContext, recordInboundSessionMetaSafe, resolveAgentRoute, resolveChunkMode, resolveConversationLabel, resolveMarkdownTableMode };
