// openclaw/dist/target-parsing-DjBM2G27.js
import { l as normalizeOptionalThreadValue, s as normalizeOptionalString } from "./string-coerce-BUSzWgUA.js";
import { r as normalizeChatChannelId } from "./ids-ZUx9SnUD.js";
import "./registry-DR0Inr-7.js";
import { i as normalizeChannelId, n as getLoadedChannelPlugin, t as getChannelPlugin } from "./registry-CAnQx2oU.js";
import "./plugins-DB_0wnJl.js";
//#region src/channels/plugins/target-parsing.ts
function parseWithPlugin(getPlugin, rawChannel, rawTarget) {
	const channel = normalizeChatChannelId(rawChannel) ?? normalizeChannelId(rawChannel);
	if (!channel) return null;
	return getPlugin(channel)?.messaging?.parseExplicitTarget?.({ raw: rawTarget }) ?? null;
}
function parseExplicitTargetForChannel(channel, rawTarget) {
	return parseWithPlugin(getChannelPlugin, channel, rawTarget);
}
function parseExplicitTargetForLoadedChannel(channel, rawTarget) {
	return parseWithPlugin(getLoadedChannelPlugin, channel, rawTarget);
}
function resolveComparableTargetForLoadedChannel(params) {
	const rawTo = normalizeOptionalString(params.rawTarget);
	if (!rawTo) return null;
	const parsed = parseExplicitTargetForLoadedChannel(params.channel, rawTo);
	const fallbackThreadId = normalizeOptionalThreadValue(params.fallbackThreadId);
	return {
		rawTo,
		to: parsed?.to ?? rawTo,
		threadId: normalizeOptionalThreadValue(parsed?.threadId ?? fallbackThreadId),
		chatType: parsed?.chatType
	};
}
function comparableChannelTargetsShareRoute(params) {
	const left = params.left;
	const right = params.right;
	if (!left || !right) return false;
	if (left.to !== right.to) return false;
	if (left.threadId == null || right.threadId == null) return true;
	return left.threadId === right.threadId;
}
//#endregion
export { resolveComparableTargetForLoadedChannel as i, parseExplicitTargetForChannel as n, parseExplicitTargetForLoadedChannel as r, comparableChannelTargetsShareRoute as t };
