// openclaw/dist/extensions/slack/runtime-api.js
import { t as DEFAULT_ACCOUNT_ID } from "../../account-id-DvNGLnmF.js";
import { a as resolveSlackAccount, c as resolveSlackBotToken, i as resolveDefaultSlackAccountId, n as listSlackAccountIds, s as resolveSlackAppToken, t as listEnabledSlackAccounts } from "../../accounts-K85qz17O.js";
import { a as resolveSlackGroupToolPolicy, i as resolveSlackGroupRequireMention, r as setSlackRuntime } from "../../runtime-BwzdOvpF.js";
import { t as resolveSlackChannelAllowlist } from "../../resolve-channels-CGivUzK2.js";
import { t as resolveSlackUserAllowlist } from "../../resolve-users-DL_ibepC.js";
import { t as sendMessageSlack } from "../../send-DfKFN4bS.js";
import { a as listSlackEmojis, c as pinSlackMessage, d as removeOwnSlackReactions, f as removeSlackReaction, i as getSlackMemberInfo, l as reactSlackMessage, m as unpinSlackMessage, o as listSlackPins, p as sendSlackMessage, r as editSlackMessage, s as listSlackReactions, t as deleteSlackMessage, u as readSlackMessages } from "../../actions-bBSmAz-_.js";
import { r as normalizeSlackWebhookPath } from "../../registry-BxCNHkz6.js";
import { t as probeSlack } from "../../probe-CtaGAoiD.js";
import { t as monitorSlackProvider } from "../../provider-DViZzyeX.js";
import { n as slackActionRuntime, t as handleSlackAction } from "../../action-runtime-BvxHwVbD.js";
import { n as listSlackDirectoryPeersLive, t as listSlackDirectoryGroupsLive } from "../../directory-live-CFtLw0c2.js";
import "../../monitor-CrFEcxid.js";
//#region extensions/slack/src/http/plugin-routes.ts
let slackHttpHandlerRuntimePromise = null;
async function loadSlackHttpHandlerRuntime() {
	slackHttpHandlerRuntimePromise ??= import("../../handler.runtime-DRvSXQ-E.js");
	return await slackHttpHandlerRuntimePromise;
}
function registerSlackPluginHttpRoutes(api) {
	const accountIds = new Set([DEFAULT_ACCOUNT_ID, ...listSlackAccountIds(api.config)]);
	const registeredPaths = /* @__PURE__ */ new Set();
	for (const accountId of accountIds) {
		const account = resolveSlackAccount({
			cfg: api.config,
			accountId
		});
		registeredPaths.add(normalizeSlackWebhookPath(account.config.webhookPath));
	}
	if (registeredPaths.size === 0) registeredPaths.add(normalizeSlackWebhookPath());
	for (const path of registeredPaths) api.registerHttpRoute({
		path,
		auth: "plugin",
		handler: async (req, res) => await (await loadSlackHttpHandlerRuntime()).handleSlackHttpRequest(req, res)
	});
}
//#endregion
export { deleteSlackMessage, editSlackMessage, getSlackMemberInfo, handleSlackAction, listEnabledSlackAccounts, listSlackAccountIds, listSlackDirectoryGroupsLive, listSlackDirectoryPeersLive, listSlackEmojis, listSlackPins, listSlackReactions, monitorSlackProvider, pinSlackMessage, probeSlack, reactSlackMessage, readSlackMessages, registerSlackPluginHttpRoutes, removeOwnSlackReactions, removeSlackReaction, resolveDefaultSlackAccountId, resolveSlackAccount, resolveSlackAppToken, resolveSlackBotToken, resolveSlackChannelAllowlist, resolveSlackGroupRequireMention, resolveSlackGroupToolPolicy, resolveSlackUserAllowlist, sendMessageSlack, sendSlackMessage, setSlackRuntime, slackActionRuntime, unpinSlackMessage };
