// openclaw/dist/extensions/openai/index.js
import { a as buildProviderToolCompatFamilyHooks } from "../../provider-tools-DnkYXOur.js";
import { t as definePluginEntry } from "../../plugin-entry-BFdKnvae.js";
import { t as buildOpenAICodexCliBackend } from "../../cli-backend-Bzchu0pV.js";
import { t as buildOpenAIImageGenerationProvider } from "../../image-generation-provider-CIvcfdVO.js";
import { n as openaiCodexMediaUnderstandingProvider, r as openaiMediaUnderstandingProvider } from "../../media-understanding-provider-C2OGreYZ.js";
import { t as buildOpenAICodexProviderPlugin } from "../../openai-codex-provider-DdXYIIom.js";
import { t as buildOpenAIProvider } from "../../openai-provider-D8FnWoQS.js";
import { a as resolveOpenAISystemPromptContribution, i as resolveOpenAIPromptOverlayMode } from "../../prompt-overlay-BVZ1irFI.js";
import { t as buildOpenAIRealtimeTranscriptionProvider } from "../../realtime-transcription-provider-HgYPduSJ.js";
import { t as buildOpenAIRealtimeVoiceProvider } from "../../realtime-voice-provider-DiLdCfh-.js";
import { t as buildOpenAISpeechProvider } from "../../speech-provider-DdudsLh2.js";
import { t as buildOpenAIVideoGenerationProvider } from "../../video-generation-provider-CEMZJBaz.js";
//#region extensions/openai/index.ts
var openai_default = definePluginEntry({
	id: "openai",
	name: "OpenAI Provider",
	description: "Bundled OpenAI provider plugins",
	register(api) {
		const promptOverlayMode = resolveOpenAIPromptOverlayMode(api.pluginConfig);
		const openAIToolCompatHooks = buildProviderToolCompatFamilyHooks("openai");
		const buildProviderWithPromptContribution = (provider) => ({
			...provider,
			...openAIToolCompatHooks,
			resolveSystemPromptContribution: (ctx) => resolveOpenAISystemPromptContribution({
				mode: promptOverlayMode,
				modelProviderId: provider.id,
				modelId: ctx.modelId
			})
		});
		api.registerCliBackend(buildOpenAICodexCliBackend());
		api.registerProvider(buildProviderWithPromptContribution(buildOpenAIProvider()));
		api.registerProvider(buildProviderWithPromptContribution(buildOpenAICodexProviderPlugin()));
		api.registerImageGenerationProvider(buildOpenAIImageGenerationProvider());
		api.registerRealtimeTranscriptionProvider(buildOpenAIRealtimeTranscriptionProvider());
		api.registerRealtimeVoiceProvider(buildOpenAIRealtimeVoiceProvider());
		api.registerSpeechProvider(buildOpenAISpeechProvider());
		api.registerMediaUnderstandingProvider(openaiMediaUnderstandingProvider);
		api.registerMediaUnderstandingProvider(openaiCodexMediaUnderstandingProvider);
		api.registerVideoGenerationProvider(buildOpenAIVideoGenerationProvider());
	}
});
//#endregion
export { openai_default as default };
