// openclaw/dist/extensions/github-copilot/register.runtime.js
import { n as ensureAuthProfileStore } from "../../store-C5g_fGOG.js";
import { i as coerceSecretRef } from "../../types.secrets-B4TpMTxm.js";
import { n as listProfilesForProvider } from "../../profiles-DT2g1T-s.js";
import { r as resolveCopilotApiToken, t as DEFAULT_COPILOT_API_BASE_URL } from "../../github-copilot-token-C25jjxCe.js";
import "../../provider-auth-CzzkrCf0.js";
import { t as githubCopilotLoginCommand } from "../../provider-auth-login-CVamMKM8.js";
import { n as resolveCopilotForwardCompatModel, t as PROVIDER_ID } from "../../models-BKd_f90A.js";
import { n as wrapCopilotProviderStream, t as wrapCopilotAnthropicStream } from "../../stream-DiBJBIL3.js";
import "../../token-C-4H-FOD.js";
import { t as fetchCopilotUsage } from "../../usage-Ht-7NZj1.js";
export { DEFAULT_COPILOT_API_BASE_URL, PROVIDER_ID, coerceSecretRef, ensureAuthProfileStore, fetchCopilotUsage, githubCopilotLoginCommand, listProfilesForProvider, resolveCopilotApiToken, resolveCopilotForwardCompatModel, wrapCopilotAnthropicStream, wrapCopilotProviderStream };
