// openclaw/dist/plugin-sdk/browser-node-runtime.js
import { n as defaultRuntime } from "../runtime-Df11Xt6R.js";
import { i as runExec } from "../exec-xi_1oMVJ.js";
import { a as isLoopbackHost } from "../net-BWcjOlSf.js";
import { o as resolveGatewayAuth } from "../auth-BbsY8q8P.js";
import { t as rawDataToString } from "../ws-BRmHzU5W.js";
import { nn as ErrorCodes, rn as errorShape } from "../protocol-CnJ5gHtr.js";
import { t as withTimeout } from "../with-timeout-XnflBimJ.js";
import { t as startLazyPluginServiceModule } from "../lazy-service-module-DkyXSGRp.js";
import { n as callGatewayFromCli, t as addGatewayClientOptions } from "../gateway-rpc-BGzygo8Y.js";
import { n as runCommandWithRuntime } from "../cli-utils-DTpzY2FY.js";
import { i as resolveNodeCommandAllowlist, n as isNodeCommandAllowed } from "../node-command-policy-C2zbAtAA.js";
import { a as respondUnavailableOnNodeInvokeError, s as safeParseJson, t as ensureGatewayStartupAuth } from "../startup-auth-DfywKJPt.js";
import "../browser-node-runtime-BCRR_cE6.js";
export { ErrorCodes, addGatewayClientOptions, callGatewayFromCli, defaultRuntime, ensureGatewayStartupAuth, errorShape, isLoopbackHost, isNodeCommandAllowed, rawDataToString, resolveGatewayAuth, resolveNodeCommandAllowlist, respondUnavailableOnNodeInvokeError, runCommandWithRuntime, runExec, safeParseJson, startLazyPluginServiceModule, withTimeout };
