// openclaw/dist/plugin-sdk/memory-core-host-runtime-cli.js
import { i as formatErrorMessage } from "../errors-D8p6rxH8.js";
import { _ as shortenHomePath, g as shortenHomeInString } from "../utils-D5DtWkEu.js";
import { r as setVerbose, t as isVerbose } from "../global-state-CKgA_HLe.js";
import { n as defaultRuntime } from "../runtime-Df11Xt6R.js";
import { t as formatDocsLink } from "../links-B6av-_2s.js";
import { n as isRich, r as theme, t as colorize } from "../theme-Clp64kpu.js";
import { t as resolveCommandSecretRefsViaGateway } from "../command-secret-gateway-CQYnY_2w.js";
import { r as withManager } from "../cli-utils-DTpzY2FY.js";
import { t as formatHelpExamples } from "../help-format-7aB74-Nx.js";
import { n as withProgress, r as withProgressTotals } from "../progress-BtlZfiBc.js";
import "../memory-core-host-runtime-cli-9svuKL2G.js";
export { colorize, defaultRuntime, formatDocsLink, formatErrorMessage, formatHelpExamples, isRich, isVerbose, resolveCommandSecretRefsViaGateway, setVerbose, shortenHomeInString, shortenHomePath, theme, withManager, withProgress, withProgressTotals };
