// openclaw/dist/plugin-sdk/tlon.js
import { t as formatDocsLink } from "../links-B6av-_2s.js";
import { r as buildChannelConfigSchema } from "../config-schema-CZtpPFHc.js";
import { n as normalizeAccountId, t as DEFAULT_ACCOUNT_ID } from "../account-id-DvNGLnmF.js";
import { t as createDedupeCache } from "../dedupe-DvG9Uaxi.js";
import { n as fetchWithSsrFGuard } from "../fetch-guard-DKe4b6qW.js";
import { s as isBlockedHostnameOrIp, t as SsrFBlockedError } from "../ssrf-tCiyDHOt.js";
import { n as emptyPluginConfigSchema } from "../config-schema-66-MD4FQ.js";
import { l as patchScopedAccountConfig, t as applyAccountNameToChannelSection } from "../setup-helpers-B_T9A0B8.js";
import { t as createChannelReplyPipeline } from "../channel-reply-pipeline-CJZpAxBU.js";
import { r as buildComputedAccountStatusSnapshot } from "../status-helpers-Csb7hAJZ.js";
import { t as createLoggerBackedRuntime } from "../runtime-logger-DFBngc5V.js";
import "../runtime-CfzB51VF.js";
import { t as createOptionalChannelSetupSurface } from "../channel-setup-DTFU7XjL.js";
//#region src/plugin-sdk/tlon.ts
const tlonSetup = createOptionalChannelSetupSurface({
	channel: "tlon",
	label: "Tlon",
	npmSpec: "@openclaw/tlon",
	docsPath: "/channels/tlon"
});
const tlonSetupAdapter = tlonSetup.setupAdapter;
const tlonSetupWizard = tlonSetup.setupWizard;
//#endregion
export { DEFAULT_ACCOUNT_ID, SsrFBlockedError, applyAccountNameToChannelSection, buildChannelConfigSchema, buildComputedAccountStatusSnapshot, createChannelReplyPipeline, createDedupeCache, createLoggerBackedRuntime, emptyPluginConfigSchema, fetchWithSsrFGuard, formatDocsLink, isBlockedHostnameOrIp, normalizeAccountId, patchScopedAccountConfig, tlonSetupAdapter, tlonSetupWizard };
