// openclaw/dist/extensions/whatsapp/test-api.js
import { a as shouldLogVerbose } from "../../globals-DQvpEApL.js";
import { b as sendTextMediaPayload, p as resolveSendableOutboundReplyParts } from "../../reply-payload-C-MyOxXD.js";
import { a as chunkText } from "../../chunk-BJOSq3sp.js";
import { t as resolveOutboundSendDep } from "../../send-deps-BtbJzzlt.js";
import "../../runtime-env-DLaqJQah.js";
import "../../reply-runtime-B2AVTFFN.js";
import { t as sanitizeForPlainText } from "../../sanitize-text-DbBlAB30.js";
import "../../outbound-runtime--HEvGZCa.js";
import { a as createEmptyChannelResult, i as createAttachedChannelResultAdapter } from "../../channel-send-result-eblsyUhk.js";
import { t as resolveWhatsAppOutboundTarget } from "../../resolve-outbound-target-BsVfR3a3.js";
import { n as sendPollWhatsApp } from "../../send-CDDCS5xn.js";
import { r as WHATSAPP_LEGACY_OUTBOUND_SEND_DEP_KEYS } from "../../heartbeat-recipients-DUr-nmKF.js";
import "../../runtime-api-D3jexpeh.js";
import { t as resolveWhatsAppRuntimeGroupPolicy } from "../../runtime-group-policy-bd8RDy2e.js";
//#region extensions/whatsapp/src/outbound-adapter.ts
function trimLeadingWhitespace(text) {
	return text?.trimStart() ?? "";
}
const whatsappOutbound = {
	deliveryMode: "gateway",
	chunker: chunkText,
	chunkerMode: "text",
	textChunkLimit: 4e3,
	sanitizeText: ({ text }) => sanitizeForPlainText(text),
	pollMaxOptions: 12,
	resolveTarget: ({ to, allowFrom, mode }) => resolveWhatsAppOutboundTarget({
		to,
		allowFrom,
		mode
	}),
	sendPayload: async (ctx) => {
		const text = trimLeadingWhitespace(ctx.payload.text);
		const hasMedia = resolveSendableOutboundReplyParts(ctx.payload).hasMedia;
		if (!text && !hasMedia) return createEmptyChannelResult("whatsapp");
		return await sendTextMediaPayload({
			channel: "whatsapp",
			ctx: {
				...ctx,
				payload: {
					...ctx.payload,
					text
				}
			},
			adapter: whatsappOutbound
		});
	},
	...createAttachedChannelResultAdapter({
		channel: "whatsapp",
		sendText: async ({ cfg, to, text, accountId, deps, gifPlayback }) => {
			const normalizedText = trimLeadingWhitespace(text);
			if (!normalizedText) return createEmptyChannelResult("whatsapp");
			return await (resolveOutboundSendDep(deps, "whatsapp", { legacyKeys: WHATSAPP_LEGACY_OUTBOUND_SEND_DEP_KEYS }) ?? (await import("../../send-BogKJ7Nd.js")).sendMessageWhatsApp)(to, normalizedText, {
				verbose: false,
				cfg,
				accountId: accountId ?? void 0,
				gifPlayback
			});
		},
		sendMedia: async ({ cfg, to, text, mediaUrl, mediaLocalRoots, mediaReadFile, accountId, deps, gifPlayback }) => {
			const normalizedText = trimLeadingWhitespace(text);
			return await (resolveOutboundSendDep(deps, "whatsapp", { legacyKeys: WHATSAPP_LEGACY_OUTBOUND_SEND_DEP_KEYS }) ?? (await import("../../send-BogKJ7Nd.js")).sendMessageWhatsApp)(to, normalizedText, {
				verbose: false,
				cfg,
				mediaUrl,
				mediaLocalRoots,
				mediaReadFile,
				accountId: accountId ?? void 0,
				gifPlayback
			});
		},
		sendPoll: async ({ cfg, to, poll, accountId }) => await sendPollWhatsApp(to, poll, {
			verbose: shouldLogVerbose(),
			accountId: accountId ?? void 0,
			cfg
		})
	})
};
//#endregion
export { resolveWhatsAppRuntimeGroupPolicy, whatsappOutbound };
